---
title: "Panel comparison analysis"
author: "Clalit Genomics"
date: "06/02/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

<style type="text/css">
  body{
  font-size: 14pt;
}
</style>

# TODO
- summarize the number of overlapping genes with the expert panels
- generate an automatic panel generator that takes in a few initial panels and returns a panel based on them

```{r render_report,eval=F,echo=F}
library(rmarkdown)
output_dir <- "/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/gene_panel_generator_project/output/reports"
render('/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/gene_panel_generator_project/src/prepare_gene_panel_data.Rmd', output_dir = output_dir,output_file = sprintf('gene_panel_prep_report.%s.html',Sys.Date()),quiet = F)
```

```{r setup, include=FALSE}
project_dir<-'/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/'
knitr::opts_knit$set(root.dir='/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/')
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()

analysis_name<-glue('panel_comparison_analysis_{Sys.Date()}')
analysis_dir<-glue('output/{analysis_name}')
if (!dir.exists(analysis_dir)){dir.create(analysis_dir)}
```

# Source statistics

First, we compared the panel and gene content of each source. 
Altogether, we reviewed `r length(unique(source_db$source))` panel sources: `r paste0(unique(source_db$source),collapse=', ')`


```{r panel_stats,include=T,eval=T,message=F}
source_stats<-get_source_stats(source_db)
# Create a gt table based on preprocessed
source_stats %>%
  gt() %>%
  tab_header(title = "Source comparison")%>%
  fmt_number(columns = c(`Genes per panel`),suffixing = TRUE)

plot_number_of_genes_per_panel_dist()

gene_num_cat_table<-generate_gene_panel_cat_number_table(panels_list)
gene_num_cat_table%>%
  gt(rowname_col = 'source') %>%
   tab_spanner(
    label = "Number of genes per panel",
    columns = c(2:8)
  )

```

Distance between panels was calculated using the Jaccard method. 

```{r binarize_and_calculate_distance,include=F,eval=T,cache=T}
binary_panel_gene_table<-create_binary_gene_table(fixed_source_db)
# filter out the hpo data
binary_panel_gene_table_no_hpo<-binary_panel_gene_table%>%filter(!grepl('hpo',row.names(binary_panel_gene_table)))
# generate distance matrix between all the panels
dm<-generate_distance_matrix(binary_panel_gene_table_no_hpo)
# convert dm to df
dm_df<-convert_dm_to_df(dm)
```

# Expert panel - panels combination

we will generate panel combinations for the expert panels (panelapp NHS and clingen) including only high condidence genes

```{r prepare_expert_panels}
expert_panels_panelapp<-panels_list%>%
  filter(source=='panelapp' & (high_conf_genes>1))

# get the panelapp NHS approved panels
panelapp_nhs_panels<-fixed_source_db%>%filter(panel_comment=='NHS panel')%>%pull(panel_name)%>%unique()
expert_panels_panelapp<-expert_panels_panelapp%>%filter(panel_name %in% panelapp_nhs_panels)

# now get the clingen panels
expert_panels_clingen<-panels_list%>%
  filter(source=='clingen' & (high_conf_genes>1))

clingen_panelapp_panels<-c(expert_panels_clingen%>%pull(panel_joined_name),
                           expert_panels_panelapp%>%pull(panel_joined_name))

```

```{r find_similar_clingen_panelapp_panels}

# now compare the two panels
expert_panels_dm_df<-dm_df%>%filter(grepl('clingen',row))%>%filter(col %in% clingen_panelapp_panels)
clingen_panels<-as.character(expert_panels_clingen%>%pull(panel_joined_name))
clingen_vs_panelapp<-get_top_similar_panels_table(joined_panel_names = clingen_panels,
                                                  source_db = fixed_source_db,
                                                  dm_df = expert_panels_dm_df,
                                                  number_of_top_panels = 5,
                                                  same_source = F)
write.table(file=glue('{analysis_dir}/clingen_vs_panelapp.csv'),clingen_vs_panelapp,sep='\t',row.names = F)
```

once the corresponding clingen / panelapp panels were found join them and collect only the high confidence genes

```{r generate_expert_panels}
clingen_vs_panelapp<-readr::read_delim('./data/accessory_files/clingen_vs_panelapp_manual_match.csv')
# use the NHS approved panelapp panels as reference and add the clingen high confidence genes 
expert_panels_list<-expert_panels_panelapp%>%mutate(source='expert_panels')
expert_panels_source<-fixed_source_db%>%
  mutate(source='expert_panel')%>%
  filter(panel_joined_name%in%expert_panels_list$panel_joined_name)%>% # take only the NHS panels
  filter(confidence_level==3) # take only the high confidence genes

# now add the matching clingen panel genes
for (clingen_panel_name in clingen_vs_panelapp$selected_panel_name){
  clingen_panel_source<-fixed_source_db%>%filter(panel_joined_name==clingen_panel_name)
  matching_panelapp_panel_joined_name<-clingen_vs_panelapp%>%filter(selected_panel_name==clingen_panel_name)%>%pull(similar_panel)
  matching_panelapp_panel<-expert_panels_source%>%filter(panel_joined_name==matching_panelapp_panel_joined_name)
  # now get the clingen genes that are not in the panelapp  panel
  clingen_genes_to_add<-clingen_panel_source%>%
    filter(confidence_level==3)%>%
    filter(!(gene_symbol %in% (matching_panelapp_panel%>%pull(gene_symbol))))
  message(glue('Will add {nrow(clingen_genes_to_add)} genes from {clingen_panel_name} to {matching_panelapp_panel_joined_name}'))
  expert_panels_source<-expert_panels_source%>%bind_rows(clingen_genes_to_add%>%mutate(source='expert_panels',
                                                                panel_joined_name=matching_panelapp_panel_joined_name,
                                                                panel_id=unique(matching_panelapp_panel%>%pull(panel_id)),
                                                                panel_name=unique(matching_panelapp_panel%>%pull(panel_name))))
}

# now add all the clingen panels without a similar panelapp panel
remaining_clingen_panels<-fixed_source_db%>%filter(source=='clingen')%>%
  filter(!(panel_joined_name%in%clingen_vs_panelapp$selected_panel_name))%>%
  filter(!(panel_name %in% c('General Gene Curation','UNC Biocuration Core','Syndromic Disorders','PTEN')))%>%
  filter(!(panel_name %in% (panels_list%>%filter(source=='clingen' & high_conf_genes==1)%>%pull(panel_name))))

expert_panels_source<-expert_panels_source%>%bind_rows(remaining_clingen_panels)
  
# now fix the source table 
joined_panel_names<-split_joined_panel_name(expert_panels_source$panel_joined_name)
joined_panel_names<-joined_panel_names%>%mutate(source='expert_panel',
                                                panel_joined_name=glue('{source}|{panel_id}|{panel_name}'))

expert_panels_source$source<-joined_panel_names$source
expert_panels_source$panel_joined_name<-joined_panel_names$panel_joined_name

```

```{r compare_expert_panels_against_the_rest}
expert_and_other_source<-expert_panels_source%>%
  bind_rows(fixed_source_db%>%filter(!(source %in% c('hpo','clingen','panelapp'))))

binary_expert_table<-create_binary_gene_table(expert_and_other_source)
# generate distance matrix between all the panels
expert_dm<-generate_distance_matrix(binary_expert_table)
# convert dm to df
expert_dm_df<-convert_dm_to_df(expert_dm)


expert_panels_similars_summary<-expert_panel_top_similars%>%summarize(n=n(),skimr::skim_without_charts(value))

expert_panels_comparison_top_similar_panels<-get_top_similar_panels_table(expert_panels_similars_summary$row,
                                                                          source_db = expert_and_other_source,
                                                                          dm_df = expert_dm_df,
                                                                          number_of_top_panels = 5,
                                                                          same_source = F)

expert_panels_comparison_top_similar_panels<-expert_panels_comparison_top_similar_panels%>%
  separate(similar_panel,sep='\\|',into = c('similar_source','similar_id','similar_name'))

expert_panels_colsest_panels<-expert_panels_comparison_top_similar_panels%>%group_by(selected_panel_name)%>%slice_min(n=1,order_by = distance)%>%ungroup()

# A table that summarizes the overlap between the expert panel genes and the similar panels
expert_panels_summary<-expert_panels_colsest_panels%>%
  mutate(percent_genes_overlap=common_genes/selected_num_o_genes)%>%select(selected_panel_name,selected_num_o_genes,percent_genes_overlap)

g<-ggplot(expert_panels_summary,aes(x=selected_num_o_genes,y=percent_genes_overlap))+
  geom_point()+
  geom_smooth(col='red',method='gam')+ 
  scale_x_log10()+
  theme_minimal()
g

# A table that summarizes per panle source how good it overlapped with the expert panels
closest_panels_summary<-expert_panels_colsest_panels%>%
  group_by(similar_source)%>%summarize(n=n(),skimr::skim_without_charts(distance))%>%
  left_join(panels_list%>%filter(number_of_genes>1)%>%group_by(source)%>%summarize(total_number_of_panels=n()),by=c('similar_source'='source'))%>%
  mutate(closest_panels_to_total_ratio=n/total_number_of_panels)

# Plot the distance of the similar panels as a function of the number of genes in the expert panel
g<-ggplot(expert_panels_comparison_top_similar_panels%>%group_by(selected_panel_name)%>%slice_min(n=3,order_by = distance),aes(x=selected_num_o_genes,y=distance,col=similar_source))+
  geom_point(alpha=0.5)+
  geom_smooth(se=F)+
  scale_x_log10()+
  theme_minimal()+
  theme(legend.position = 'top')+
  labs(x='Number of genes in expert panel',y='Distance from expert panel',col=NULL)+
  scale_color_nejm()
g


```

Panelapp is XXX 
Our aim is to both demonstrate the extent of variability between expert-panel based and commercial gene panels.
For this purpose we reviewed selected panels designed by panelapp, and compared them against other similar panels from all other sources. 

```{r select_panels}
panelapp_panels<-panels_list%>%filter(source=='panelapp')
selected_panels<-c('Familial hypercholesterolaemia - targeted panel',
                   'RASopathies',
                   'Hypertrophic cardiomyopathy - teen and adult',
                   'Hearing loss',
                   'Autism')
```

```{r}

# now for hypertrophic cardiomyopathy - get the most associated panels
selected_panel_name<-find_panel_in_dm_df(dm_df,source='panelapp',panel_name='Hypertrophic cardio')
# get similar panels
similar_panels_df<-get_similar_panels_from_dm_df(dm_df,selected_panel_name,top_most_similar = 0.005,maximal_distance = 0.99)

similar_panels_binary_gene_table<-subset_binary_gene_table(binary_panel_gene_table_no_hpo,similar_panels_df$col)
plot_panel_gene_heatmap(similar_panels_binary_gene_table,sort_by_similarity_to_selected_panel = T)


gbin_for_heatmap<-bin_for_heatmap%>%mutate(panel=row.names(bin_for_heatmap))%>%pivot_longer(-panel)%>%left_join(similar_panels_df,by=c('panel'='col'))
p <- ggplot(gbin_for_heatmap, aes(forcats::fct_reorder(panel, value.y), name, fill= value.x)) + 
  geom_tile()+coord_flip()+theme(axis.text.x = element_text(angle=90))
p

panels_to_plot<-as.character(unique(panelapp_hcm$col))
panels_num<-match(panels_to_plot,row.names(test_bin_data))
selected_bin_data<-test_bin_data[panels_num,]
selected_dm<-generate_distance_matrix(selected_bin_data)
plot_dm_as_dendogram(selected_dm,selected_panel)
# generate heatmap !!!

hcm_panels<-row.names(binary_panel_gene_table)[grepl('hypertroph',row.names(binary_panel_gene_table),ignore.case = T) & grepl('cardio',row.names(binary_panel_gene_table),ignore.case=T) & !(grepl('hpo',row.names(binary_panel_gene_table),ignore.case = T))]

```

```{r auto_generate_panel}

dm_with_hpo<-generate_distance_matrix(binary_panel_gene_table%>%filter(row.names(binary_panel_gene_table) %in% (panels_list%>%filter(number_of_genes>1)%>%pull(panel_joined_name))))
# convert dm to df
dm_df_with_hpo<-convert_dm_to_df(dm)

joined_panel_names<-c('invitae|02261|Invitae Hypertrophic Cardiomyopathy Panel','health_in_code|1_1|Hypertrophic Cardiomyopathy – Basic Panel',"blueprint|1_16|Hypertrophic Cardiomyopathy (HCM) Panel",'cegat|3_2|Cardiomyopathy, hypertrophic','health_in_code|1_2|Hypertrophic Cardiomyopathy – Extended Panel')
gen_source_db<-fixed_source_db%>%filter(!grepl('clingen|panelapp|hpo',source))
gen_dm_df<-dm_df%>%filter(!grepl('clingen|panelapp',row))%>%filter(!grepl('clingen|panelapp|hpo',col))

comp_panel<-expert_panels_source%>%filter(grepl('Hypertrophic',panel_joined_name))
#comp_panel<-fixed_source_db%>%filter(grepl('panelapp',source))%>%filter(grepl('Hypertrophic',panel_joined_name))


gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db = gen_source_db,gen_dm_df = gen_dm_df,max_distance = 0.85)

dist_by_thresh<-NULL
for (thresh in seq(0,1,0.001)){
  genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
  dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
  dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
}
dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g

comp_panel<-comp_panel%>%left_join(gene_rank)


g<-ggplot(dist_by_thresh,aes(x=thresh,y=similarity))+
  geom_point()+
  theme_minimal()
g


joined_panel_names<-c("health_in_code|3_64|Movement disorders comprehensive panel" ,"pronto|86|Movement Disorders")
comp_panel<-expert_panels_source%>%filter(panel_joined_name=='expert_panel|540|Adult onset movement disorder')

gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db = gen_source_db,gen_dm_df = gen_dm_df,max_distance = 0.9)

comp_panel<-comp_panel%>%left_join(gene_rank)

dist_by_thresh<-NULL
for (thresh in seq(0,1,0.001)){
  genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
  dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
  dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
}
dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g

```
