---
title: "Panel comparison analysis"
author: "Clalit Genomics"
date: "06/02/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

<style type="text/css">
  body{
  font-size: 14pt;
}
</style>

# TODO
- re-run gtr matrix builder (need to add the binary table to the save)
- calculate the average/median distance between the different panels in a given set - call it concordance
- show that concordance is stronger with the automatic search vs the naive search
- NEED TO SHOW THAT WE ARE BETTER THAN THE NAIVE SEARCH AND NOT PANELAPP


```{r render_report,eval=F,echo=F}
library(rmarkdown)
output_dir <- "/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/gene_panel_generator_project/output/reports"
render('/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/gene_panel_generator_project/src/prepare_gene_panel_data.Rmd', output_dir = output_dir,output_file = sprintf('gene_panel_prep_report.%s.html',Sys.Date()),quiet = F)
```

```{r setup, include=FALSE}
project_dir<-'/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/'
knitr::opts_knit$set(root.dir='/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/')
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()

analysis_name<-glue('panel_comparison_analysis_{Sys.Date()}')
analysis_dir<-glue('output/{analysis_name}')
if (!dir.exists(analysis_dir)){dir.create(analysis_dir)}

```


```{r definitions}
# Define the minimal number of genes in a panel - panels with less genes will be excluded from the analysis (under the assumption that they are not panels per-se)
minimal_number_of_genes_in_panel<-3
maximal_number_of_genes_in_panel<-4000
# first fix the manually collected panels
source_db_panels_with_enough_genes<-fixed_source_db%>%count(panel_joined_name)%>%filter(n>=minimal_number_of_genes_in_panel,n<=maximal_number_of_genes_in_panel)%>%pull(panel_joined_name)
fixed_source_db<-fixed_source_db%>%filter(panel_joined_name%in%source_db_panels_with_enough_genes)
# remove panelapp and clingen low conf genes
fixed_source_db<-fixed_source_db%>%filter(!(source%in%c('panelapp','clingen') & confidence_level<2))

gtr_db_panels_with_enough_genes<-gtr_db%>%count(panel_joined_name)%>%filter(n>=minimal_number_of_genes_in_panel)%>%pull(panel_joined_name)
gtr_db<-gtr_db%>%filter(panel_joined_name%in%gtr_db_panels_with_enough_genes)

```

# Source statistics

First, we compared the panel and gene content of each source. 
Altogether, we reviewed `r length(unique(source_db$source))` panel sources: `r paste0(unique(source_db$source),collapse=', ')`


```{r panel_stats,include=T,eval=T,message=F}
source_stats<-get_source_stats(gtr_db)
# Create a gt table based on preprocessed
source_stats %>% arrange(desc(Panels))%>%
  gt() %>%
  tab_header(title = "Source comparison")%>%
  fmt_number(columns = c(`Genes per panel`),suffixing = TRUE)

gene_num_cat_table<-generate_gtr_gene_panel_cat_number_table(gtr_panels_list)
gene_num_cat_table%>%
  gt(rowname_col = 'name_of_laboratory') %>%
   tab_spanner(
    label = "Number of genes per panel",
    columns = c(2:7)
  )

unique_sources<-gtr_db%>%pull(source)%>%unique()
unique_panels<-gtr_db%>%pull(panel_joined_name)%>%unique()
unique_panelapp_panels<-fixed_source_db%>%filter(source=='panelapp')%>%pull(panel_joined_name)%>%unique()
source_summary_text<-paste0(glue('After exclusion, a total of {length(unique_panels)} panels from {length(unique_sources)} different vendors were collected from the GTR and {length(panelapp_panels)} panels were collected from PanelApp.'))

source_summary_text
```

Distance between panels was calculated using the Jaccard method. 

# Create a binary matrix of the manually selected gene panels

```{r gtr_binarize_and_calculate_distance,include=F,eval=T,cache=T}
# binary_panel_gene_table<-create_binary_gene_table(fixed_source_db)
# # filter out the hpo data
# binary_panel_gene_table_no_hpo<-binary_panel_gene_table%>%filter(!grepl('hpo',row.names(binary_panel_gene_table)))
# # generate distance matrix between all the panels
# dm<-generate_distance_matrix(binary_panel_gene_table_no_hpo)
# # convert dm to df
# dm_df<-convert_dm_to_df(dm)
# 
# save(dm,dm_df,file='./data/pre_generated_data/manual_panels_data.Rdata')

# load dm data
load('./data/pre_generated_data/manual_panels_data.Rdata')
```

# Create a binary matrix of the GTR gene panels

```{r gtr_binarize_and_calculate_distance,include=F,eval=T,cache=T}

# gtr_with_expert_db<-gtr_db%>%mutate(panel_id=as.character(panel_id))%>%
#   bind_rows(fixed_source_db%>%filter(source%in%c('panelapp','clingen')))#%>%
#   #bind_rows(expert_panels_source)
# 
# 
# gtr_binary_panel_gene_table<-create_binary_gene_table(gtr_with_expert_db)
# # generate distance matrix between all the panels
# gtr_dm<-generate_distance_matrix(gtr_binary_panel_gene_table)
# # convert dm to df
# gtr_dm_df<-convert_dm_to_df(gtr_dm)
# 
# # save gtr matrix and dataframe
# save(gtr_dm,gtr_binary_panel_gene_table,gtr_with_expert_db,gtr_dm_df,file='./data/pre_generated_data/gtr_data.Rdata')

# load gtr data
load('./data/pre_generated_data/gtr_data.Rdata')
```

# The problem: Major panel discrepancy

## Example: Selected genes
```{r panels_discrepancy}
all_panel_names<-gtr_dm_df%>%pull(row)%>%unique()
panelapp_panels<-grep('panelapp',all_panel_names,ignore.case = T,value = T)


summary_text<-paste0(glue('In this study we focus on {length(selected_phenotypes)} phenotypes: {paste0(selected_phenotypes,collapse=", ")}'))

# panels discrepancy
# if all genes vs pubmed for the phenotype was already calculated, can use it and skip the pubmed search part
panels_discrepancy_analysis<-function(phenotype_name,
                                      phenotype_panels,
                                      gtr_with_expert_db,
                                      panel_to_compare,
                                      all_genes_vs_pubmed=NA){
  discrepancy_analysis_res<-list()
  discrepancy_analysis_res$relatedness<-get_panel_relatedness_stats(phenotype_panels,
                                                                    gtr_with_expert_db,
                                                                    panel_to_compare = panel_to_compare)
  
  # run pubmed search for each gene in the panels 
  if (is.na(all_genes_vs_pubmed)){
    discrepancy_analysis_res$panel_genes_pubmed_search<-pubmed_gene_list_vs_phenotype(phenotype_name,discrepancy_analysis_res$relatedness$gene_list)
  }else{
    discrepancy_analysis_res$panel_genes_pubmed_search<-all_genes_vs_pubmed%>%filter(gene_symbol%in%discrepancy_analysis_res$relatedness$gene_list)
  }
  discrepancy_analysis_res$panel_vs_pubmed <-
    summarize_panel_vs_phenotype_pubmed(
      selected_panels = phenotype_panels,
      gtr_with_expert_db = gtr_with_expert_db,
      panel_genes_pubmed_search = discrepancy_analysis_res$panel_genes_pubmed_search,
      panelapp_panel = panel_to_compare
    )
  
  # discrepancy_analysis_res$num_publications_per_gene_vs_in_panelapp_plot<-
  #   plot_num_of_publications_per_gene_vs_in_panelapp(discrepancy_analysis_res$panel_vs_pubmed)
  # discrepancy_analysis_res$num_of_publications_vs_rate_of_panels_plot<-
  #   plot_num_of_publications_vs_rate_of_panels(discrepancy_analysis_res$panel_vs_pubmed)
  # discrepancy_analysis_res$num_of_genes_with_publications_vs_rate_of_panels_plot<-
  #   plot_num_of_genes_with_publications_vs_rate_of_panels(discrepancy_analysis_res$panel_vs_pubmed)
  # discrepancy_analysis_res$num_of_genes_with_publications_vs_rate_of_panels_plot<-
  #   plot_num_of_genes_with_publications_vs_rate_of_panels_with_panelapp(discrepancy_analysis_res$panel_vs_pubmed)
  
  discrepancy_analysis_res$summary_text<-get_relatedness_summary_text(phenotype_name,
                                                             discrepancy_analysis_res$relatedness,
                                                             discrepancy_analysis_res$panel_vs_pubmed)
  return(discrepancy_analysis_res)
}


# HCM
phenotype_name<-'hypertrophic cardiomyopathy'
panelapp_panel<-"panelapp|49|Hypertrophic cardiomyopathy - teen and adult" 
hcm_panels<-grep('osteo',grep('hypertrophic|hcm',panel_names,ignore.case = T,value=T),ignore.case = T,invert = T,value=T)

# panels concordance
remotes::install_github("philliplab/yasss")
library(yasss)
hcm_binary<-gtr_binary_panel_gene_table%>%filter(rownames(gtr_binary_panel_gene_table)%in%hcm_panels)
# generate distance matrix between all the panels
hcm_dm<-generate_distance_matrix(hcm_binary)
# need to consider how to summarize this distance matrix... maybe just summarize the distances calculated

hcm_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                             hcm_panels,
                                             gtr_with_expert_db,
                                             panelapp_panel)

# Epilepsy
phenotype_name<-'epilepsy'
panelapp_panel<-"panelapp|402|Genetic epilepsy syndromes" 
epilepsy_panels<-grep('osteo|autis',grep('epileps',panel_names,ignore.case = T,value=T),ignore.case = T,invert = T,value=T)

epilepsy_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                             epilepsy_panels,
                                             gtr_with_expert_db,
                                             panelapp_panel)


# cholestasis
phenotype_name<-'cholestasis'
panelapp_panel<-"panelapp|544|Cholestasis" 
cholestasis_panels<-grep('osteo',grep('cholesta',panel_names,ignore.case = T,value=T),ignore.case = T,invert = T,value = T)
cholestasis_discrepacy<-panels_discrepancy_analysis(phenotype_name,
                                             cholestasis_panels,
                                             gtr_with_expert_db,
                                             panelapp_panel)

# Cleft
phenotype_name<-'cleft lip'
panelapp_panel<-"panelapp|81|Clefting" 
cleft_panels<-grep('osteo',grep('cleft',panel_names,ignore.case = T,value=T),ignore.case = T,invert = T,value = T)
cleft_discrepacy<-panels_discrepancy_analysis(phenotype_name,
                                             cleft_panels,
                                             gtr_with_expert_db,
                                             panelapp_panel)


# CTD
panelapp_panel<-'panelapp|53|Ehlers Danlos syndromes'
phenotype_name<-'connective tissue disorders'
ctd_panels<-grep('tissue gene tests|tissue laboratory',grep('connective tissue|ehler',panel_names,ignore.case = T,value=T),ignore.case = T,invert = T,value = T)

ctd_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                             ctd_panels,
                                             gtr_with_expert_db,
                                             panelapp_panel)

# Upset plot
names(hcm_panels)<-paste0('p',1:length(hcm_panels))
panels_mat<-panels_bin%>%t()%>%as.data.frame()
colnames(panels_mat)<-names(hcm_panels)
upset(panels_mat,nsets=length(hcm_panels),nintersects = 20,order.by = 'freq')


```

# HCM as an example of naive (text based) search VS a clustering based search VS a distance cutoff based search

```{r hcm_naive_vs_auto}

panelapp_panel<-"panelapp|49|Hypertrophic cardiomyopathy - teen and adult" 
phenotype_name<-'hypertrophic cardiomyopathy'

# first get the number of HCM publications for all the genes
# all_genes<-gtr_with_expert_db%>%pull(gene_symbol)%>%unique()
# all_genes_vs_hcm<-pubmed_gene_list_vs_phenotype(phenotype_name,all_genes)
# save(all_genes_vs_hcm,file='./data/pre_generated_data/all_genes_vs_hcm.RData')
load('./data/pre_generated_data/all_genes_vs_hcm.RData')
all_genes_with_hcm_pub<-all_genes_vs_hcm%>%filter(pub_num>0)%>%pull(gene_symbol)

# panelapp metrics
all_metrics<-NULL
metrics_panelapp<-calculate_sens_and_ppv_vs_pubmed(gtr_with_expert_db%>%filter(panel_joined_name==panelapp_panel)%>%pull(gene_symbol),all_genes_with_hcm_pub)
metrics_panelapp$group<-'panelapp'
all_metrics<-all_metrics%>%bind_rows(as.data.frame(metrics_panelapp))

# naive search
hcm_naive_panels<-grep('osteo',grep('hypertrophic|hcm',panel_names,ignore.case = T,value=T),ignore.case = T,invert = T,value=T)

hcm_naive_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                             hcm_naive_panels,
                                             gtr_with_expert_db,
                                             panelapp_panel,
                                             all_genes_vs_pubmed = all_genes_vs_hcm)

metrics_naive<-calculate_sens_and_ppv_vs_pubmed(hcm_naive_discrepancy$relatedness$gene_list,all_genes_with_hcm_pub)
metrics_naive$group='naive_search'

# cluster search
# hc <- hclust(gtr_dm,method='ward.D' )
# # cut tree
# cut_avg <- cutree(hc,k = 100)
# panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
# # get the selected panelapp panel cluster
# panelapp_cluster<-panel_clusters%>%filter(panel_name==panelapp_panel)%>%pull(cluster)
# hcm_cluster_panels<-panel_clusters%>%filter(cluster==panelapp_cluster)%>%pull(panel_name)
# 
# setdiff(hcm_naive_panels,hcm_cluster_panels)
# 
# hcm_cluster_discrepancy<-panels_discrepancy_analysis(phenotype_name,
#                                              hcm_cluster_panels,
#                                              gtr_with_expert_db,
#                                              panelapp_panel)
# 
# metrics_cluster<-calculate_sens_and_ppv_vs_pubmed(hcm_cluster_discrepancy$relatedness$gene_list,all_genes_with_hcm_pub)

# check a range of number of clusters
#all_metrics<-NULL
clust_methods=c('ward.D','ward.D2')
for (clust_method in clust_methods){
  hc <- hclust(gtr_dm,method=clust_method)
  for (cluster_size in c(50,75,100,150,200,300)){
      cut_avg <- cutree(hc,k = cluster_size)
      panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
      # get the selected panelapp panel cluster
      panelapp_cluster<-panel_clusters%>%filter(panel_name==panelapp_panel)%>%pull(cluster)
      hcm_cluster_panels<-panel_clusters%>%filter(cluster==panelapp_cluster)%>%pull(panel_name)
      
      setdiff(hcm_naive_panels,hcm_cluster_panels)
      
      hcm_cluster_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                                   hcm_cluster_panels,
                                                   gtr_with_expert_db,
                                                   panelapp_panel,
                                                   all_genes_vs_pubmed = all_genes_vs_hcm)
      
      metrics_cluster<-calculate_sens_and_ppv_vs_pubmed(hcm_cluster_discrepancy$relatedness$gene_list,all_genes_with_hcm_pub)
    
      metrics_cluster$group=glue('method={clust_method},cluster_size={cluster_size}')
    
      all_metrics<-all_metrics%>%bind_rows(as.data.frame(metrics_cluster))
  }
}


# distance search
max_dist<-0.85

hcm_dist_panels<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,selected_panel_name = panelapp_panel,top_most_similar = 1,maximal_distance = max_dist,names_only = T)

#setdiff(hcm_dist_panels,hcm_cluster_panels)

hcm_dist_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                             hcm_dist_panels,
                                             gtr_with_expert_db,
                                             panelapp_panel,
                                             all_genes_vs_pubmed = all_genes_vs_hcm)

metrics_dist<-calculate_sens_and_ppv_vs_pubmed(hcm_dist_discrepancy$relatedness$gene_list,all_genes_with_hcm_pub)

# check a range of distances
#all_metrics<-NULL
for (max_dist in c(0.65,0.7,0.75,0.8,0.85,0.9)){
    hcm_dist_panels<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,
                                                           selected_panel_name = panelapp_panel,top_most_similar = 1,
                                                           maximal_distance = max_dist,
                                                           names_only = T)
  
  #setdiff(hcm_dist_panels,hcm_cluster_panels)
  
  hcm_dist_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                               hcm_dist_panels,
                                               gtr_with_expert_db,
                                               panelapp_panel,
                                               all_genes_vs_pubmed = all_genes_vs_hcm)
  
  metrics_dist<-calculate_sens_and_ppv_vs_pubmed(hcm_dist_discrepancy$relatedness$gene_list,all_genes_with_hcm_pub)
  metrics_dist$group=glue('max_dist={max_dist}')
  
  all_metrics<-all_metrics%>%bind_rows(as.data.frame(metrics_dist))
}

all_metrics%>%ggplot(aes(x=sensitivity,y=ppv,color=group))+
  geom_point()+
  theme_minimal()
```

# ADD CONCORDANCE CALCULATION
consider: https://rdrr.io/github/philliplab/yasss/man/summarize_dmat.html
```{r check_concordance}
all_panel_names<-gtr_dm_df%>%pull(row)%>%unique()
panelapp_panels<-grep('panelapp',all_panel_names,ignore.case = T,value = T)
# for a given set of panels, calculate distance metrics between all panels
phenotype_name<-'hypertrophic cardiomyopathy'
panelapp_panel<-"panelapp|49|Hypertrophic cardiomyopathy - teen and adult" 
naive_panels<-grep('osteo',grep('hypertrophic|hcm',all_panel_names,ignore.case = T,value=T),ignore.case = T,invert = T,value=T)



panels_concordance<-NULL
panels_concordance<-panels_concordance%>%bind_rows(data.frame(group='naive',
                                                              subset_dm_df_by_gene_panel(naive_panels,gtr_dm_df)))

# now do the same by dist
max_dist<-0.85
dist_panels<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,selected_panel_name = panelapp_panel,top_most_similar = 1,maximal_distance = max_dist,names_only = T)
panels_concordance<-panels_concordance%>%bind_rows(data.frame(group='dist',
                                                              subset_dm_df_by_gene_panel(dist_panels,gtr_dm_df)))


panels_concordance%>%group_by(group)%>%select(value)%>%skimr::skim()

panels_concordance%>%ggplot(aes(x=value,fill=group))+
  geom_density(alpha=0.4)+
  theme_minimal()
```



```{r check_dist_distribution_for_panelapp_panels}


panelapp_panels_dist<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%group_by(row)%>%
  filter(col!=row)%>%
  summarize(mean_dist=mean(value),
            median_dist=median(value),
            min_dist=min(value))%>%
  left_join(gtr_with_expert_db%>%
              filter(source=='panelapp')%>%
              mutate(panel_joined_name=as.character(panel_joined_name))%>%
              group_by(panel_joined_name)%>%
              summarize(n_genes=n()),
            by=c('row'='panel_joined_name'))

panelapp_panels_dist%>%filter(n_genes<1000)%>%ggplot(aes(x=n_genes,y=min_dist))+
  geom_point()+
  theme_minimal()



```

# Naive vs auto

```{r naive_vs_auto}
all_panel_names<-gtr_dm_df%>%pull(row)%>%unique()
panelapp_panels<-grep('panelapp',all_panel_names,ignore.case = T,value = T)

naive_vs_auto_test<-function(panelapp_panel,
                             phenotype_name,
                             naive_search,
                             all_panel_names,
                             all_genes,
                             pre_generated_gene_publications_list){
  
  test_res<-list()
  
  all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = phenotype_name,
                                                          gene_list = all_genes,
                                                          pre_generated_genes_pubs = pre_generated_gene_publications_list)
  
  all_genes_with_pheno_pub<-all_genes_vs_phenotype%>%filter(pub_num>1)%>%pull(gene_symbol)
  message(glue('Out of {length(all_genes)} genes, {length(all_genes_with_pheno_pub)} had a publication mentioning {phenotype_name}'))
  
  # panelapp metrics
  test_res$all_metrics<-NULL
  metrics_panelapp<-calculate_sens_and_ppv_vs_pubmed(gtr_with_expert_db%>%filter(panel_joined_name==panelapp_panel)%>%pull(gene_symbol),all_genes_with_pheno_pub)
  metrics_panelapp$group<-panelapp_panel
  test_res$all_metrics<-test_res$all_metrics%>%bind_rows(as.data.frame(metrics_panelapp))
  
  # naive search - with panelapp
  pheno_naive_panels<-grep(naive_search,all_panel_names,ignore.case = T,value=T)
  
  if (length(pheno_naive_panels)==0){
    stop(glue('Error: could not find any panel corresponding to the submitted naive search: {naive_search}'))}
  
  # first calculate performance per panel
  for (pheno_panel in pheno_naive_panels){
    if (grepl('panelapp',pheno_panel)){next}#no need to calculate again
    metrics_pheno_panel<-calculate_sens_and_ppv_vs_pubmed(gtr_with_expert_db%>%
                                                            filter(panel_joined_name==pheno_panel)%>%
                                                            pull(gene_symbol),
                                                          all_genes_with_pheno_pub)
    metrics_pheno_panel$group<-pheno_panel
    test_res$all_metrics<-test_res$all_metrics%>%bind_rows(as.data.frame(metrics_pheno_panel))
  }
  
  # then claculate performance for all panels
  
  test_res$naive_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                               pheno_naive_panels,
                                               gtr_with_expert_db,
                                               panelapp_panel,
                                               all_genes_vs_pubmed = all_genes_vs_phenotype)
  
  
  metrics_naive<-calculate_sens_and_ppv_vs_pubmed(test_res$naive_discrepancy$relatedness$gene_list,
                                                  all_genes_with_pheno_pub)
  metrics_naive$group='naive_search - with panelapp'
  test_res$all_metrics<-test_res$all_metrics%>%bind_rows(as.data.frame(metrics_naive))
  
  # naive search - no panelapp
  pheno_naive_panels_no_panelapp<-grep('panelapp',pheno_naive_panels,invert = T,value=T)
  
  if (length(pheno_naive_panels_no_panelapp)==0){
    stop(glue('Error: could not find any panel corresponding to the submitted naive search: {naive_search}'))}
  
  # then claculate performance for all panels
  
  test_res$naive_no_panelapp_discrepancy<-panels_discrepancy_analysis(phenotype_name,
                                               pheno_naive_panels_no_panelapp,
                                               gtr_with_expert_db,
                                               panelapp_panel,
                                               all_genes_vs_pubmed = all_genes_vs_phenotype)
  
  
  metrics_naive<-calculate_sens_and_ppv_vs_pubmed(test_res$naive_no_panelapp_discrepancy$relatedness$gene_list,
                                                  all_genes_with_pheno_pub)
  metrics_naive$group='naive_search - no panelapp'
  test_res$all_metrics<-test_res$all_metrics%>%bind_rows(as.data.frame(metrics_naive))
  
  
  # cluster search
  # check a range of number of clusters
  # clust_methods=c('ward.D','ward.D2')
  # for (clust_method in clust_methods){
  #   hc <- hclust(gtr_dm,method=clust_method)
  #   for (cluster_size in c(50,75,100,150,200,300)){
  #       cut_avg <- cutree(hc,k = cluster_size)
  #       panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
  #       # get the selected panelapp panel cluster
  #       panelapp_cluster<-panel_clusters%>%filter(panel_name==panelapp_panel)%>%pull(cluster)
  #       cluster_panels<-panel_clusters%>%filter(cluster==panelapp_cluster)%>%pull(panel_name)
  # 
  #       #setdiff(naive_panels,cluster_panels)
  # 
  #       test_res$cluster_discrepancy<-panels_discrepancy_analysis(phenotype_name,
  #                                                    cluster_panels,
  #                                                    gtr_with_expert_db,
  #                                                    panelapp_panel,
  #                                                    all_genes_vs_pubmed = all_genes_vs_phenotype)
  # 
  #       metrics_cluster<-calculate_sens_and_ppv_vs_pubmed(test_res$cluster_discrepancy$relatedness$gene_list,all_genes_with_pheno_pub)
  # 
  #       metrics_cluster$group=glue('method={clust_method},cluster_size={cluster_size}')
  # 
  #       test_res$all_metrics<-test_res$all_metrics%>%bind_rows(as.data.frame(metrics_cluster))
  #   }
  # }
  
  # check a range of distances
  #all_metrics<-NULL
  for (max_dist in c(0.7,0.75,0.8,0.85,0.95)){
      dist_panels<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,
                                                             selected_panel_name = panelapp_panel,
                                                             top_most_similar = 1,
                                                             maximal_distance = max_dist,
                                                             names_only = T)
    
    #setdiff(dist_panels,cluster_panels)
    
    discrepancy_name<-glue('dist_discrepancy_{max_dist}')
    test_res[[discrepancy_name]]<-panels_discrepancy_analysis(phenotype_name,
                                                 dist_panels,
                                                 gtr_with_expert_db,
                                                 panelapp_panel,
                                                 all_genes_vs_pubmed = all_genes_vs_phenotype)
    
    metrics_dist<-calculate_sens_and_ppv_vs_pubmed(test_res[[discrepancy_name]]$relatedness$gene_list,all_genes_with_pheno_pub)
    metrics_dist$group=glue('max_dist={max_dist}')
    
    test_res$all_metrics<-test_res$all_metrics%>%bind_rows(as.data.frame(metrics_dist))
  }
  
  test_res$all_metrics%>%ggplot(aes(x=sensitivity,y=ppv,color=group))+
    geom_point()+
    theme_minimal()
  return(test_res)
}

```

# Naive vs auto - selected panels

```{r nva_selected}
all_panel_names<-gtr_dm_df%>%pull(row)%>%unique()
panelapp_panels<-grep('panelapp',all_panel_names,ignore.case = T,value = T)
all_genes<-gtr_with_expert_db%>%pull(gene_symbol)%>%unique()
# run once to pre-generate the per gene publications data
#pubmed_grab_all_gene_publications(all_genes)
# load the pregenerated all gene publications
load('./data/pre_generated_data/genes_pubmed_search.RData')


pancreatitis_papers<-load('./data/pre_generated_data/pancreatitis_pubmed_search.RData')

selected_phenotypes_panelapp<-c("panelapp|49|Hypertrophic cardiomyopathy - teen and adult",
                                'panelapp|544|Cholestasis',
                                'panelapp|149|Nephrocalcinosis or nephrolithiasis',
                                'panelapp|179|Hydrocephalus',
                                'panelapp|657|Autism',
                                'panelapp|126|Hearing loss',
                                'panelapp|294|Ocular coloboma',
                                'panelapp|309|Skeletal dysplasia',
                                'panelapp|386|Pancreatitis',
                                'panelapp|285|Intellectual disability',
                                'panelapp|943|Osteopetrosis',
                                'panelapp|553|Ectodermal dysplasia',
                                'panelapp|487|Cystic renal disease',
                                'panelapp|258|Arthrogryposis',
                                'panelapp|78|Holoprosencephaly')
selected_phenotypes<-c('hypertrophic cardiomyopathy',
                       'cholestasis',
                       'nephrolithiasis',
                       'hydrocephalus',
                       'autism',
                       'hearing loss',
                       'coloboma',
                       'skeletal dysplasia',
                       'pancreatitis',
                       'intellectual disability',
                       'osteopetrosis',
                       'ectodermal dysplasia',
                       'polycystic kidney',
                       'Arthrogryposis',
                       'Holoprosencephaly')

phenotype_queries<-c('hypertrophic cardio|hcm|cardio.+hypertrophic',
                     'cholestasis',
                     'nephrolithiasis|monogenic kidney stone',
                     'hydrocephalus',
                     'autism',
                     'hearing loss|deafness',
                     'coloboma',
                     'skeletal dysplasia',
                     'pancreatitis',
                     'intellectual disability|mental retard',
                     'osteopetrosis',
                     'ectodermal dysplasia|ectoderm',
                     'Cystic renal|polycys.*kidn',
                     'Arthrogryposis',
                     'Holoprosencephaly'
                     )


selected_phenotypes_df<-data.frame(phenotype_name=selected_phenotypes,
                                   panelapp_panel=selected_phenotypes_panelapp,
                                   query=phenotype_queries)

selected_phenotypes_summary_text<-paste0(glue('In order to demonstrate the discrepancy between available panels, and the utility of our method, {nrow(selected_phenotypes_df)} phenotypes were selected: {paste0(selected_phenotypes,collapse=", ")}.'),
                                         collapse=' ')
selected_phenotypes_summary_text                                  
# rentrez::set_entrez_key('be1775d2481f815d7d6f044bb0d0bca39f08')

selected_phenotypes_naive_vs_auto<-list()
for (selected_phenotype_name in selected_phenotypes){
  message(glue('Running comparison for {selected_phenotype_name}..'))
  panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
  naive_search<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(query)
  selected_phenotypes_naive_vs_auto[[selected_phenotype_name]] <-
    naive_vs_auto_test(
      panelapp_panel = panelapp_panel,
      phenotype_name = selected_phenotype_name,
      naive_search =  naive_search,
      all_genes = all_genes,
      all_panel_names = all_panel_names,
      pre_generated_gene_publications_list = entrez_pubmed_res)
}
save(selected_phenotypes_naive_vs_auto,file='./data/pre_generated_data/selected_phenotypes_naive_vs_auto.RData')
#load('./data/pre_generated_data/selected_phenotypes_naive_vs_auto.RData')
# now check the ranks for each 

selected_phenotypes_naive_vs_auto_df<-NULL
for (selected_phenotype in names(selected_phenotypes_naive_vs_auto)){
  selected_phenotypes_naive_vs_auto_df<-selected_phenotypes_naive_vs_auto_df%>%
    bind_rows(data.frame(phenotype=selected_phenotype,
                         selected_phenotypes_naive_vs_auto[[selected_phenotype]]$all_metrics%>%
                           mutate(rank_f1=rank(desc(f1),ties.method = 'min'),
                                  rank_sens=rank(desc(sensitivity),ties.method = 'min'),
                                  rank_ppv=rank(desc(ppv),ties.method = 'min'))))
}
z<-selected_phenotypes_naive_vs_auto_df%>%
  mutate(group=ifelse(grepl('^panelapp',group),'panelapp',group))%>%
  filter(grepl('naive|panelapp$|max_dist|^panelapp|cluster',group))%>%group_by(group)%>%
  select(-c(ngenes,phenotype))%>%
  pivot_longer(-c(group))%>%
  group_by(group,name)%>%
  summarize(skimr::skim_without_charts(value))
  
selected_phenotypes_naive_vs_auto_df%>%
  mutate(group=ifelse(grepl('^panelapp',group),'panelapp',group))%>%
  filter(grepl('naive|panelapp$|max_dist|^panelapp',group))%>%
  select(group,rank_f1,rank_sens,rank_ppv)%>%
  pivot_longer(-c(group))%>%
  filter(group=='max_dist=0.75')%>%
  ggplot(aes(x=value,fill=group))+
  geom_histogram(alpha=0.3)+facet_wrap(name~.)+
  theme_minimal()

auto_generator_panelapp_only<-list()
auto_generator_panelapp_with_similar<-list()
max_dists=c(0.75,0.8,0.85,0.9,0.95)
for (max_dist in max_dists){
  auto_generator_panelapp_only[[glue('max_dist={max_dist}')]]<-list()
  for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
    message(glue('Generating panel for {selected_phenotype_name} max dist = {max_dist}'))
    panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
    # similar_panels<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,
    #                                                       selected_panel_name = panelapp_panel,
    #                                                       top_most_similar = 1,maximal_distance = 0.7,
    #                                                       names_only = T)
    auto_generator_panelapp_only[[glue('max_dist={max_dist}')]][[selected_phenotype_name]] <-
      generate_panel_from_panels(
        joined_panel_names = panelapp_panel,
        gen_source_db = gtr_with_expert_db,
        gen_dm_df = gtr_dm_df,
        max_distance = max_dist,
        with_mitochondrial = F)
    # auto_generator_panelapp_with_similar[[glue('max_dist={max_dist}')]][[selected_phenotype_name]] <-
    #   generate_panel_from_panels(
    #     joined_panel_names = similar_panels,
    #     gen_source_db = gtr_with_expert_db,
    #     gen_dm_df = gtr_dm_df,
    #     max_distance = max_dist,
    #     with_mitochondrial = F)
  }
}

# now add whether the gene is positive or not
library(tidymodels)
all_rocs<-NULL
panelapp_roc<-NULL
naive_roc<-NULL
dist_roc<-NULL
min_num_of_pub_to_consider_positive<-2

for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
  panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
  for (max_dist in max_dists){
    
    all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype_name,
                                                            gene_list = all_genes,
                                                            pre_generated_genes_pubs = entrez_pubmed_res)
    all_genes_vs_phenotype<-all_genes_vs_phenotype%>%mutate(has_pub=factor(ifelse(pub_num>=min_num_of_pub_to_consider_positive,1,0)))
    table_for_roc_panelapp_only<-all_genes_vs_phenotype%>%
                  select(gene_symbol,has_pub)%>%
      left_join(auto_generator_panelapp_only[[glue('max_dist={max_dist}')]][[selected_phenotype_name]])%>%
      mutate(gene_score=ifelse(is.na(gene_score),0,gene_score),
             rank_percentile=ifelse(is.na(rank_percentile),0,adjusted_gene_score),
             num_o_panels=ifelse(is.na(num_o_panels),0,num_o_panels))%>%
      mutate(gene_score_modified=ifelse(!is.na(in_original_panels)&(in_original_panels),gene_score+10,gene_score))
    
    # table_for_roc_panelapp_with_similar<-all_genes_vs_phenotype%>%
    #               select(gene_symbol,has_pub)%>%
    #   left_join(auto_generator_panelapp_with_similar[[glue('max_dist={max_dist}')]][[selected_phenotype_name]])%>%
    #   mutate(gene_score=ifelse(is.na(gene_score),0,gene_score),
    #          rank_percentile=ifelse(is.na(rank_percentile),0,adjusted_gene_score),
    #          num_o_panels=ifelse(is.na(num_o_panels),0,num_o_panels))%>%
    #   mutate(gene_score_with_panelapp=ifelse(!is.na(number_of_panels_with_gene),gene_score+num_o_panels,gene_score))
    # 
    all_rocs<-all_rocs%>%
      bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                           group='gene_score',
                           max_dist=max_dist,
                           table_for_roc_panelapp_only%>%
                             pr_curve(truth=has_pub,estimate=gene_score,event_level = 'second')))%>%
      # bind_rows(data.frame(phenotype_name=selected_phenotype_name,
      #                      group='gene_score_similar_panels',
      #                      max_dist=max_dist,
      #                      table_for_roc_panelapp_with_similar%>%
      #                        pr_curve(truth=has_pub,estimate=gene_score,event_level = 'second')))
      # bind_rows(data.frame(phenotype_name=selected_phenotype_name,
      #                      group='rank_percentile',
      #                      max_dist=max_dist,
      #                      table_for_roc%>%
      #                        pr_curve(truth=has_pub,estimate=rank_percentile,event_level = 'second')))%>%
      # bind_rows(data.frame(phenotype_name=selected_phenotype_name,
      #                      group='num_o_panels',
      #                      max_dist=max_dist,
      #                      table_for_roc%>%
      #                        pr_curve(truth=has_pub,estimate=num_o_panels,event_level = 'second')))%>%
      bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                           group='gene_score_modified',
                           max_dist=max_dist,
                           table_for_roc_panelapp_only%>%
                             pr_curve(truth=has_pub,estimate=gene_score_modified,event_level = 'second')))
    
  }
  panelapp_genes<-gtr_with_expert_db%>%filter(panel_joined_name==panelapp_panel)%>%pull(gene_symbol)
  panelapp_roc<-panelapp_roc%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         recall=sum(panelapp_genes%in%(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)))/length(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)),
                         precision=sum(panelapp_genes%in%(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)))/length(panelapp_genes)))
  
  naive_genes<-selected_phenotypes_naive_vs_auto[[selected_phenotype_name]]$naive_discrepancy$relatedness$gene_list
  naive_roc<-naive_roc%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         recall=sum(naive_genes%in%(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)))/length(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)),
                         precision=sum(naive_genes%in%(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)))/length(naive_genes)))
  
  dist_genes<-selected_phenotypes_naive_vs_auto[[selected_phenotype_name]]$dist_discrepancy_0.75$relatedness$gene_list
  dist_roc<-dist_roc%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         recall=sum(dist_genes%in%(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)))/length(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)),
                         precision=sum(dist_genes%in%(all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol)))/length(dist_genes)))
}



all_rocs%>%filter(max_dist==0.85,grepl('gene_score$',group))%>%ggplot(aes(x=recall,y=precision,col=factor(group)))+
  geom_line()+facet_wrap(phenotype_name~.)+
  geom_abline(slope = -1,intercept = 1,linetype=2,alpha=0.5)+
  geom_point(data=panelapp_roc,inherit.aes = F,aes(x=recall,y=precision))+
  geom_point(data=naive_roc,inherit.aes = F,aes(x=recall,y=precision),col='red')+
  #geom_point(data=dist_roc,inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  theme_minimal()


```

```{r pca}
# Add PCA calculation
test_binary<-gtr_binary_panel_gene_table[1:1000,1:1000]
library(logisticPCA)
logsvd_model = logisticSVD(test_binary, k = 2)
plot(logsvd_model)

panels = rownames(test_binary)
plot(logsvd_model, type = "scores") + geom_point(aes()) + 
  ggtitle("Exponential Family PCA") + scale_colour_manual(values = c("blue", "red"))

library(FactoMineR)
test_binary<-gtr_binary_panel_gene_table[,]
res.mca <- FactoMineR::PCA(test_binary, 
              ncp=2,graph=F)
coords<-data.frame(panel=rownames(res.mca$ind$coord),res.mca$ind$coord)%>%
  mutate(source=stringr::str_replace(panel,'\\|.+',''))

to_plot<-coords%>%filter(panel%in%naive_panels)%>%mutate(group='naive_panels')%>%
  bind_rows(coords%>%filter(panel%in%dist_panels)%>%mutate(group='dist_panels'))

to_plot%>%
  #filter(Dim.1<50,Dim.2<20)%>%
  ggplot(aes(x=Dim.1,y=Dim.2,col=factor(source=='panelapp')))+
  geom_point(alpha=0.4)#+facet_wrap(group~.)+
  theme_minimal()

```


```{r check_dist_distribution_for_panelapp_panels}
panelapp_panels_dist<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%group_by(row)%>%
  filter(col!=row)%>%
  summarize(mean_dist=mean(value),
            median_dist=median(value),
            min_dist=min(value))%>%
  left_join(gtr_with_expert_db%>%
              filter(source=='panelapp')%>%
              mutate(panel_joined_name=as.character(panel_joined_name))%>%
              group_by(panel_joined_name)%>%
              summarize(n_genes=n()),
            by=c('row'='panel_joined_name'))

panelapp_panels_dist%>%filter(n_genes<1000)%>%ggplot(aes(x=n_genes,y=min_dist))+
  geom_point()+
  theme_minimal()

```



```{r cluster_panels}
# hierarchal clustering
hc <- hclust(gtr_dm,method='ward.D' )
# cut tree
cut_avg <- cutree(hc,k = 100)
panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
# this is not bad - it actually groups similar panels together.. but what is the optimal number of clusters??
# basically the purpose is to reduce false positives while maintaining high sensitivity - so lets see what happens when i divide the panels into X clusters and go from there
panel_clusters%>%dplyr::count(cluster)%>%arrange(n)

```



```{r analyze_distance_matrix}
# Analyze the ratio between the number of genes in a panel and the distance it has with all other panels with the same approximate size

dm_to_analyze<-gtr_dm_df%>%
  left_join(gtr_panels_list%>%select(panel_joined_name,number_of_genes_row=number_of_genes),by=c('row'='panel_joined_name'))%>%
  left_join(gtr_panels_list%>%select(panel_joined_name,number_of_genes_col=number_of_genes),by=c('col'='panel_joined_name'))

dm_to_analyze<-dm_to_analyze%>%mutate(ngenes_cat_row=cut(number_of_genes_row,breaks=c(0,10,30,50,100,200,500,1000,10000)),
                                      ngenes_cat_col=cut(number_of_genes_col,breaks=c(0,10,30,50,100,200,500,1000,10000)))


head(dm_to_analyze)
dm_to_analyze%>%slice_sample(n=100000)%>%ggplot(aes(x=number_of_genes_row,y=value))+
  geom_point()+facet_wrap(ngenes_cat_row~.,scales='free')
  

```

```{r compare_panelapp_against_the_rest}
# This code goes through all of panelapps panels and compares them against all the panels with at most max_dist distance. 
# it then calculates sensitivity and precision for all the panels and outputs how many panels had missing genes in them

panelapp_df<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%filter(!grepl('panelapp',col))
panelapp_df<-panelapp_df%>%
  left_join(panels_list%>%
              mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))

# Check whether how well the different panelapp panels are represented by looking for the sensitivity 
max_dist<-c(1,0.99,0.95,0.9,0.85)
panelapp_vs_all_other<-NULL
for (md in max_dist){
  message(glue('calculating panelapp similarity for {md} distance'))
  panelapp_comp<-panelapp_df%>%group_by(row)%>%filter(value<md)%>%ungroup()
  panelapp_comp<-panelapp_comp%>%group_by(row)%>%dplyr::summarize(num_o_comp_panels=n(),compare_panel_vs_many_panels(row,unique(col),gtr_with_expert_db,min_confidence_level = 2))%>%mutate(max_dist=md)
  
  panels_without_similar_panels<-setdiff(panelapp_df%>%pull(row)%>%as.character()%>%unique(),panelapp_comp%>%pull(row))
  if (length(panels_without_similar_panels)>0){
    panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                 num_o_comp_panels=0,
                                                 genes_in_b=0,
                                                 common_genes=0,
                                                 B_not_in_A=0,
                                                 Agenes_not_in_B='',
                                                 max_dist=md)
    panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
      genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
      A_not_in_B=genes_in_a
    )
    panelapp_comp<-panelapp_comp%>%bind_rows(panels_without_similar_panels_df)
  }
  
  panelapp_vs_all_other<-panelapp_vs_all_other%>%bind_rows(panelapp_comp)
}

panelapp_vs_all_other<-panelapp_vs_all_other%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)

panelapp_vs_all_other%>%group_by(max_dist)%>%dplyr::summarize(mean_sens=mean(sensitivity),
                                              mean_prec=mean(precision),
                                              num_o_panels_with_missing_genes=sum(A_not_in_B>0))



```


```{r compare_panelapp_against_the_rest_cluster_analysis}
# This code goes through all of panelapps panels and compares them against all the panels with at most max_dist distance. 
# it then calculates sensitivity and precision for all the panels and outputs how many panels had missing genes in them



panelapp_vs_all_other_cluster<-NULL
for (method in c('ward.D2','ward.D')){
  print(method)
  for (cluster_size in c(1,75,150)){
    print(cluster_size)
    # hierarchal clustering
    hc <- hclust(gtr_dm,method=method )
    
    # library(cluster)
    # z<-agnes(gtr_dm)
    # cut tree
    cut_avg <- cutree(hc,k = cluster_size)
    panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
    # this is not bad - it actually groups similar panels together.. but what is the optimal number of clusters??
    # basically the purpose is to reduce false positives while maintaining high sensitivity - so i lets see what happens when i divide the panels into X clusters and go from there
    panel_clusters%>%dplyr::count(cluster)%>%arrange(n)
    
    
    panelapp_df<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%filter(!grepl('panelapp',col))
    panelapp_df<-panelapp_df%>%
      left_join(panels_list%>%
                  mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
      left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
      left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)
    
    # now do the same but only within the panel cluster
    # Check whether how well the different panelapp panels are represented by looking for the sensitivity 
    max_dist<-c(1,0.99,0.95,0.9)
    
    for (md in max_dist){
      message(glue('calculating panelapp similarity for {md} distance'))
      panelapp_comp<-panelapp_df%>%group_by(row)%>%filter(value<md)%>%filter(row_cluster==col_cluster)%>%ungroup()
      panelapp_comp<-panelapp_comp%>%group_by(row)%>%
        dplyr::summarize(num_o_comp_panels=n(),
                         compare_panel_vs_many_panels(row,unique(col),
                                                      gtr_with_expert_db,
                                                      min_confidence_level = 2))%>%
        mutate(max_dist=md)
      
      panels_without_similar_panels<-setdiff(panelapp_df%>%pull(row)%>%as.character()%>%unique(),panelapp_comp%>%pull(row))
      if (length(panels_without_similar_panels)>0){
        panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                     num_o_comp_panels=0,
                                                     genes_in_b=0,
                                                     common_genes=0,
                                                     B_not_in_A=0,
                                                     Agenes_not_in_B='',
                                                     max_dist=md)
        panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
          genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
          A_not_in_B=genes_in_a
        )
        panelapp_comp<-panelapp_comp%>%bind_rows(panels_without_similar_panels_df)
      }
      
      panelapp_vs_all_other_cluster<-panelapp_vs_all_other_cluster%>%
        bind_rows(data.frame(method=method,cluster_size=cluster_size,panelapp_comp))
    }
    
    
  }
}

panelapp_vs_all_other_cluster<-panelapp_vs_all_other_cluster%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)%>%rowwise()%>%mutate(f1= (2 * precision * sensitivity) / (precision + sensitivity))

cluster_analysis_res<-panelapp_vs_all_other_cluster %>%
  group_by(method, cluster_size, max_dist) %>%
  dplyr::summarize(mean_sens = mean(sensitivity),
                   mean_prec = mean(precision),
                   mean_f1 = mean(f1,na.rm=T),
                   num_o_panels_with_missing_genes=sum(A_not_in_B>0))

cluster_analysis_res%>%filter(method=='ward.D2')%>%select(-num_o_panels_with_missing_genes)%>%
  pivot_longer(-c(cluster_size,max_dist,method))%>%
  ggplot(aes(x=factor(cluster_size),y=value,fill=method))+
  geom_col(position='dodge')+facet_wrap(name~max_dist,scale='free')+
  theme_minimal()

cluster_analysis_res%>%filter(method=='ward.D')%>%select(-num_o_panels_with_missing_genes)%>%
  pivot_longer(-c(cluster_size,max_dist,method))%>%
  ggplot(aes(x=cluster_size,y=value,col=factor(max_dist)))+
  geom_line()+facet_wrap(name~.,scale='free')+
  theme_minimal()+
  ggsci::scale_color_nejm()+
  theme(legend.position = 'top')


```


```{r compare_clingen_vs_panelapp_plus_gtr}


#max_dist<-c(1,0.99,0.95,0.9,0.85)
max_dist<-c(1,0.99,0.98,0.975,0.97,0.95)
cluster_sizes<-c(1,50,80,100)
cluster_method<-'ward.D'
clingen_vs_all_other<-NULL

for(cluster_size in cluster_sizes){
  
  hc <- hclust(gtr_dm,method=cluster_method )
  # library(cluster)
  # z<-agnes(gtr_dm)
  # cut tree
  cut_avg <- cutree(hc,k = cluster_size)
  panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
  
  # for every clingen panel - build a panel and see how it covers it
  clingen_df<-gtr_dm_df%>%filter(grepl('clingen',row))%>%filter(!grepl('clingen',col))

  
  clingen_df<-clingen_df%>%
        left_join(panels_list%>%
                    mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
        left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
        left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)
  
  for (md in max_dist){
    message(glue('calculating clingen similarity for {md} distance'))
    clingen_comp<-clingen_df%>%group_by(row)%>%filter(value<md)%>%ungroup()
    clingen_comp<-clingen_df%>%group_by(row)%>%filter(value<md)%>%filter(row_cluster==col_cluster)%>%ungroup()
    clingen_comp<-clingen_comp%>%group_by(row)%>%
      summarize(num_o_comp_panels=n(),
                compare_panel_vs_many_panels(row,unique(col),gtr_with_expert_db,min_confidence_level = 3))%>%
      mutate(max_dist=md,cluster_size=cluster_size)
    panels_without_similar_panels<-setdiff(clingen_df%>%pull(row)%>%as.character()%>%unique(),clingen_comp%>%pull(row))
    if (length(panels_without_similar_panels)>0){
      panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                   num_o_comp_panels=0,
                                                   genes_in_b=0,
                                                   common_genes=0,
                                                   B_not_in_A=0,
                                                   Agenes_not_in_B='',
                                                   max_dist=md)
      panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
        genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
        A_not_in_B=genes_in_a
      )
      clingen_comp<-clingen_comp%>%bind_rows(panels_without_similar_panels_df)
    }
    clingen_vs_all_other<-clingen_vs_all_other%>%bind_rows(clingen_comp)
  }
}

clingen_vs_all_other<-clingen_vs_all_other%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)


clingen_test_res<-clingen_vs_all_other%>%group_by(cluster_size,max_dist)%>%summarize(mean_sens=mean(sensitivity),
                                              mean_prec=mean(precision),
                                              num_o_panels_with_missing_genes=sum(A_not_in_B>0))
z<-clingen_vs_all_other%>%filter(max_dist==0.9)
```

```{r compare_auto_vs_vendor}

hc <- hclust(gtr_dm,method='ward.D' )
# library(cluster)
# z<-agnes(gtr_dm)
# cut tree
cut_avg <- cutree(hc,k = 100)
panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))

auto_df<-gtr_dm_df%>%
      left_join(panels_list%>%
                  mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
      left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
      left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)

# start with a panelapp panel
all_panel_names<-auto_df%>%pull(row)%>%unique()
auto_df%>%select(row,row_cluster)%>%distinct()%>%filter(grepl('panelapp',row),grepl('hypertroph',row,ignore.case = T))%>%pull(row,row_cluster)
panelapp_panel<-"panelapp|49|Hypertrophic cardiomyopathy - teen and adult" 
cluster=35
same_cluster_panels<-auto_df%>%select(col,col_cluster)%>%distinct()%>%filter(col_cluster==35)%>%pull(col)

# build hypertrophy panel according to panelapp panel
auto_panel<-generate_panel_from_panels(joined_panel_names = panelapp_panel,
                                       gtr_with_expert_db%>%
                                         filter(panel_joined_name%in%c(panelapp_panel,same_cluster_panels)),
                                       auto_df,
                                       max_distance = 1,
                                       with_mitochondrial = T)

# blueprint
blueprint_panel<-source_db%>%filter(source=='blueprint',grepl('hypertrophic',panel_joined_name,ignore.case = T))
# invitae
invitae_panel<-source_db%>%filter(source=='invitae',grepl('hypertrophic',panel_joined_name,ignore.case = T))
# cegat
cegat_panel<-source_db%>%filter(source=='cegat',grepl('hypertrophic',panel_joined_name,ignore.case = T))

library(UpSetR)
panels_mat<-auto_panel%>%filter(adjusted_rank_percentile>0.5)%>%
  select(gene_symbol)%>%mutate(auto_panel=1)%>%
  pivot_wider(names_from = gene_symbol,values_from = auto_panel)%>%
  bind_rows(blueprint_panel%>%select(gene_symbol)%>%mutate(blueprint=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = blueprint,values_fill = NA))%>%
  bind_rows(invitae_panel%>%select(gene_symbol)%>%mutate(invitae=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = invitae,values_fill = NA))%>%
  bind_rows(cegat_panel%>%select(gene_symbol)%>%mutate(cegat=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = cegat,values_fill = NA))%>%
  t()%>%as.data.frame()
  
colnames(panels_mat)<-c('auto_panel','blueprint','invitae','cegat')
panels_mat[is.na(panels_mat)]<-0
upset(panels_mat)

# maybe at this point i will perform an automated pubmed search? and see how many genes have the association ?
```


```{r}

# now for hypertrophic cardiomyopathy - get the most associated panels
selected_panel_name<-find_panel_in_dm_df(dm_df,source='panelapp',panel_name='Hypertrophic cardio')
# get similar panels
similar_panels_df<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,selected_panel_name,top_most_similar = 0.005,maximal_distance = 0.99)

similar_panels_binary_gene_table<-subset_binary_gene_table(gtr_binary_panel_gene_table,similar_panels_df$col)
plot_panel_gene_heatmap(similar_panels_binary_gene_table,sort_by_similarity_to_selected_panel = T)


gbin_for_heatmap<-bin_for_heatmap%>%mutate(panel=row.names(bin_for_heatmap))%>%pivot_longer(-panel)%>%left_join(similar_panels_df,by=c('panel'='col'))
p <- ggplot(gbin_for_heatmap, aes(forcats::fct_reorder(panel, value.y), name, fill= value.x)) + 
  geom_tile()+coord_flip()+theme(axis.text.x = element_text(angle=90))
p

panels_to_plot<-as.character(unique(panelapp_hcm$col))
panels_num<-match(panels_to_plot,row.names(test_bin_data))
selected_bin_data<-test_bin_data[panels_num,]
selected_dm<-generate_distance_matrix(selected_bin_data)
plot_dm_as_dendogram(selected_dm,selected_panel)
# generate heatmap !!!

panels<-row.names(binary_panel_gene_table)[grepl('hypertroph',row.names(binary_panel_gene_table),ignore.case = T) & grepl('cardio',row.names(binary_panel_gene_table),ignore.case=T) & !(grepl('hpo',row.names(binary_panel_gene_table),ignore.case = T))]

```

```{r auto_generate_panel}

#joined_panel_names<-c('invitae|02261|Invitae Hypertrophic Cardiomyopathy Panel','health_in_code|1_1|Hypertrophic Cardiomyopathy  Basic Panel',"blueprint|1_16|Hypertrophic Cardiomyopathy (HCM) Panel",'cegat|3_2|Cardiomyopathy, hypertrophic','health_in_code|1_2|Hypertrophic Cardiomyopathy  Extended Panel')
# gen_source_db<-fixed_source_db%>%filter(!grepl('clingen|panelapp|hpo',source))
# gen_dm_df<-dm_df%>%filter(!grepl('clingen|panelapp',row))%>%filter(!grepl('clingen|panelapp|hpo',col))


joined_panel_names<-c('expert_panel|49|Hypertrophic cardiomyopathy - teen and adult')
joined_panel_names<-c('panelapp|49|Hypertrophic cardiomyopathy - teen and adult')
joined_panel_names<-c('panelapp|652|Dilated cardiomyopathy - adult and teen')
joined_panel_names<-c('Invitae|514924|Invitae Hypertrophic Cardiomyopathy Panel','Blueprint Genetics|552625|Hypertrophic Cardiomyopathy (HCM) Panel','Health in Code S.L.|530670|Hypertrophic cardiomyopathy extended panel')
joined_panel_names<-c('panelapp|555|Ichthyosis and erythrokeratoderma')

comp_panel<-expert_panels_source%>%filter(grepl('Hypertrophic',panel_joined_name))
comp_panel<-fixed_source_db%>%filter(grepl('panelapp',source))%>%filter(grepl('Hypertrophic',panel_joined_name))
comp_panel<-readr::read_csv('./data/accessory_files/hypertrophic_cardiomyopathy_basic.csv',col_names = 'gene_symbol')
comp_panel<-readr::read_csv('./data/accessory_files/hypertrophic_cardiomyopathy_extended.csv',col_names = 'gene_symbol')%>%mutate(gene_symbol=stringr::str_replace(gene_symbol,'\\*',''))
comp_panel<-readr::read_delim('./data/accessory_files/ichthyosis_erythrokeratodermi.csv')
comp_panel$gene_symbol<-comp_panel$gene_symbol%>%trimws()

gen_source_db<-gtr_with_expert_db
gen_dm_df<-gtr_dm_df%>%filter(!grepl('panelapp|clingen|expert',col))

gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db,gen_dm_df,max_distance = 0.9,with_mitochondrial = T)

# run panelapp analysis
panelapp_panels<-panels_list%>%filter(grepl('panelapp',source))%>%ungroup()%>%select(panel_joined_name)%>%mutate(panel_joined_name=panel_joined_name%>%as.character())

panelapp_generated_panels<-NULL
for (panelapp_panel in panelapp_panels$panel_joined_name){
  message(glue('Generating panel for {panelapp_panel}'))
  gene_rank<-generate_panel_from_panels(joined_panel_names = panelapp_panel,gen_source_db,gen_dm_df,max_distance = 0.9,with_mitochondrial = T)
  if (nrow(gene_rank)==0){next}
  # get the confidence level
  gene_rank<-gene_rank%>%left_join(fixed_source_db%>%filter(panel_joined_name==panelapp_panel)%>%select(gene_symbol,confidence_level))
  panelapp_generated_panels<-panelapp_generated_panels%>%bind_rows(data.frame(panelapp_panel,gene_rank))
}
g<-ggplot(panelapp_generated_panels,aes(y=scaled_adjusted_gene_score,x=cut(adjusted_rank_percentile,10),fill=factor(confidence_level)))+
  geom_boxplot()#+facet_wrap(panelapp_panel~.)
g


toPlot<-panelapp_generated_panels%>%mutate(adjusted_rank_percentile_cat=cut(adjusted_rank_percentile,10),
                                           confidence_level=ifelse(is.na(confidence_level),0,ifelse(confidence_level==0,1,confidence_level)))%>%
  group_by(adjusted_rank_percentile_cat,panelapp_panel)%>%
  summarize(num_o_genes=n(),
            num_o_high_conf=sum(ifelse(confidence_level==3,1,0)),
            num_o_med_conf=sum(ifelse(confidence_level==2,1,0)))%>%
  mutate(rate_o_high_conf=num_o_high_conf/num_o_genes,
         rate_o_med_conf=num_o_med_conf/num_o_genes)


g<-ggplot(toPlot,aes(x=adjusted_rank_percentile_cat,y=rate_o_med_conf))+
  geom_boxplot()
g



calculate_distance_by_thresh<-function(gene_rank,comp_panel,adjusted=F){
  dist_by_thresh<-NULL
  for (thresh in seq(0,1,0.001)){
    if (adjusted){
      genes_above_thresh<-gene_rank%>%filter(adjusted_rank_percentile>=thresh)%>%pull(gene_symbol)
    }else{
      genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
    }
    dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
    dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
  }
  dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)
  return(dist_by_thresh)
}

dist_by_thresh<-calculate_distance_by_thresh(gene_rank,comp_panel,adjusted=T)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g

g<-ggplot(dist_by_thresh,aes(x=thresh,y=sensitivity))+
  geom_point()+
  theme_minimal()
g

comp_panel<-comp_panel%>%left_join(gene_rank)


joined_panel_names<-c("Fulgent Genetics|566600|Comprehensive Glomerular Proteinuria NGS Panel")
comp_panel<-fixed_source_db%>%filter(panel_joined_name=='panelapp|106|Proteinuric renal disease')%>%filter(confidence_level>1)

gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db = gtr_db,gen_dm_df = gtr_dm_df,max_distance = 0.99,with_mitochondrial = T)

comp_panel<-comp_panel%>%left_join(gene_rank)

dist_by_thresh<-NULL
for (thresh in seq(0,1,0.001)){
  genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
  dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
  dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
}
dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g


g<-ggplot(dist_by_thresh,aes(x=thresh,y=sensitivity))+
  geom_point()+
  theme_minimal()
g
```


# Expert panel - panels combination

we will generate panel combinations for the expert panels (panelapp NHS and clingen) including only high condidence genes

```{r prepare_expert_panels}
expert_panels_panelapp<-panels_list%>%
  filter(source=='panelapp' & (high_conf_genes>1))

# get the panelapp NHS approved panels
panelapp_nhs_panels<-fixed_source_db%>%filter(panel_comment=='NHS panel')%>%pull(panel_name)%>%unique()
expert_panels_panelapp<-expert_panels_panelapp%>%filter(panel_name %in% panelapp_nhs_panels)

# now get the clingen panels
expert_panels_clingen<-panels_list%>%
  filter(source=='clingen' & (high_conf_genes>1))

clingen_panelapp_panels<-c(expert_panels_clingen%>%pull(panel_joined_name),
                           expert_panels_panelapp%>%pull(panel_joined_name))

```

```{r find_similar_clingen_panelapp_panels}

# now compare the two panels
expert_panels_dm_df<-dm_df%>%filter(grepl('clingen',row))%>%filter(col %in% clingen_panelapp_panels)
clingen_panels<-as.character(expert_panels_clingen%>%pull(panel_joined_name))
clingen_vs_panelapp<-get_top_similar_panels_table(joined_panel_names = clingen_panels,
                                                  source_db = fixed_source_db,
                                                  dm_df = expert_panels_dm_df,
                                                  number_of_top_panels = 5,
                                                  same_source = F)
write.table(file=glue('{analysis_dir}/clingen_vs_panelapp.csv'),clingen_vs_panelapp,sep='\t',row.names = F)
```

once the corresponding clingen / panelapp panels were found join them and collect only the high confidence genes

```{r generate_expert_panels}
clingen_vs_panelapp<-readr::read_delim('./data/accessory_files/clingen_vs_panelapp_manual_match.csv')
# use the NHS approved panelapp panels as reference and add the clingen high confidence genes 
expert_panels_list<-expert_panels_panelapp%>%mutate(source='expert_panels')
expert_panels_source<-fixed_source_db%>%
  mutate(source='expert_panel')%>%
  filter(panel_joined_name%in%expert_panels_list$panel_joined_name)%>% # take only the NHS panels
  filter(confidence_level==3) # take only the high confidence genes

# now add the matching clingen panel genes
for (clingen_panel_name in clingen_vs_panelapp$selected_panel_name){
  clingen_panel_source<-fixed_source_db%>%filter(panel_joined_name==clingen_panel_name)
  matching_panelapp_panel_joined_name<-clingen_vs_panelapp%>%filter(selected_panel_name==clingen_panel_name)%>%pull(similar_panel)
  matching_panelapp_panel<-expert_panels_source%>%filter(panel_joined_name==matching_panelapp_panel_joined_name)
  # now get the clingen genes that are not in the panelapp  panel
  clingen_genes_to_add<-clingen_panel_source%>%
    filter(confidence_level==3)%>%
    filter(!(gene_symbol %in% (matching_panelapp_panel%>%pull(gene_symbol))))
  message(glue('Will add {nrow(clingen_genes_to_add)} genes from {clingen_panel_name} to {matching_panelapp_panel_joined_name}'))
  expert_panels_source<-expert_panels_source%>%bind_rows(clingen_genes_to_add%>%mutate(source='expert_panels',
                                                                panel_joined_name=matching_panelapp_panel_joined_name,
                                                                panel_id=unique(matching_panelapp_panel%>%pull(panel_id)),
                                                                panel_name=unique(matching_panelapp_panel%>%pull(panel_name))))
}

# now add all the clingen panels without a similar panelapp panel
remaining_clingen_panels<-fixed_source_db%>%filter(source=='clingen')%>%
  filter(!(panel_joined_name%in%clingen_vs_panelapp$selected_panel_name))%>%
  filter(!(panel_name %in% c('General Gene Curation','UNC Biocuration Core','Syndromic Disorders','PTEN')))%>%
  filter(!(panel_name %in% (panels_list%>%filter(source=='clingen' & high_conf_genes==1)%>%pull(panel_name))))

expert_panels_source<-expert_panels_source%>%bind_rows(remaining_clingen_panels)
  
# now fix the source table 
joined_panel_names<-split_joined_panel_name(expert_panels_source$panel_joined_name)
joined_panel_names<-joined_panel_names%>%mutate(source='expert_panel',
                                                panel_joined_name=glue('{source}|{panel_id}|{panel_name}'))

expert_panels_source$source<-joined_panel_names$source
expert_panels_source$panel_joined_name<-joined_panel_names$panel_joined_name

```
