---
title: "Panel comparison analysis"
author: "Clalit Genomics"
date: "06/02/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

<style type="text/css">
  body{
  font-size: 14pt;
}
</style>

# TODO
- re-run gtr matrix builder (need to add the binary table to the save)
- calculate the average/median distance between the different panels in a given set - call it concordance
- show that concordance is stronger with the automatic search vs the naive search
- NEED TO SHOW THAT WE ARE BETTER THAN THE NAIVE SEARCH AND NOT PANELAPP


```{r render_report,eval=F,echo=F}
library(rmarkdown)
output_dir <- "/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/gene_panel_generator_project/output/reports"
render('/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/gene_panel_generator_project/src/prepare_gene_panel_data.Rmd', output_dir = output_dir,output_file = sprintf('gene_panel_prep_report.%s.html',Sys.Date()),quiet = F)
```

```{r setup, include=FALSE}
project_dir<-'/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/'
knitr::opts_knit$set(root.dir='/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/')
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()

analysis_name<-glue('panel_comparison_analysis_{Sys.Date()}')
analysis_dir<-glue('output/{analysis_name}')
if (!dir.exists(analysis_dir)){dir.create(analysis_dir)}

```


```{r definitions}
# Define the minimal number of genes in a panel - panels with less genes will be excluded from the analysis (under the assumption that they are not panels per-se)
minimal_number_of_genes_in_panel<-3
maximal_number_of_genes_in_panel<-4000
# first fix the manually collected panels
source_db_panels_with_enough_genes<-fixed_source_db%>%count(panel_joined_name)%>%filter(n>=minimal_number_of_genes_in_panel,n<=maximal_number_of_genes_in_panel)%>%pull(panel_joined_name)
fixed_source_db<-fixed_source_db%>%filter(panel_joined_name%in%source_db_panels_with_enough_genes)
# remove panelapp and clingen low conf genes
fixed_source_db<-fixed_source_db%>%filter(!(source%in%c('panelapp','clingen') & confidence_level<2))

gtr_db_panels_with_enough_genes<-gtr_db%>%count(panel_joined_name)%>%filter(n>=minimal_number_of_genes_in_panel)%>%pull(panel_joined_name)
gtr_db<-gtr_db%>%filter(panel_joined_name%in%gtr_db_panels_with_enough_genes)

```

# Source statistics

```{r panel_stats,include=T,eval=T,message=F}
source_stats<-get_source_stats(gtr_db)
# Create a gt table based on preprocessed
source_stats %>% arrange(desc(Panels))%>%
  gt() %>%
  tab_header(title = "Source comparison")%>%
  fmt_number(columns = c(`Genes per panel`),suffixing = TRUE)

gene_num_cat_table<-generate_gtr_gene_panel_cat_number_table(gtr_panels_list)
gene_num_cat_table%>%
  gt(rowname_col = 'name_of_laboratory') %>%
   tab_spanner(
    label = "Number of genes per panel",
    columns = c(2:7)
  )

number_of_genes_per_panel<-gtr_db%>%group_by(panel_joined_name)%>%summarize(n=n())
number_of_genes_per_panel_summary<-skimr::skim_without_charts(number_of_genes_per_panel$n)

unique_sources<-gtr_db%>%pull(source)%>%unique()
unique_panels<-gtr_db%>%pull(panel_joined_name)%>%unique()
unique_panelapp_panels<-fixed_source_db%>%filter(source=='panelapp')%>%pull(panel_joined_name)%>%unique()

```

Distance between panels was calculated using the Jaccard method. 

# Create a binary matrix of the manually selected gene panels

```{r gtr_binarize_and_calculate_distance,include=F,eval=T,cache=T}
# binary_panel_gene_table<-create_binary_gene_table(fixed_source_db)
# # filter out the hpo data
# binary_panel_gene_table_no_hpo<-binary_panel_gene_table%>%filter(!grepl('hpo',row.names(binary_panel_gene_table)))
# # generate distance matrix between all the panels
# dm<-generate_distance_matrix(binary_panel_gene_table_no_hpo)
# # convert dm to df
# dm_df<-convert_dm_to_df(dm)
# 
# save(dm,dm_df,file='./data/pre_generated_data/manual_panels_data.Rdata')

# load dm data
load('./data/pre_generated_data/manual_panels_data.Rdata')
```

# Create a binary matrix of the GTR gene panels

```{r gtr_binarize_and_calculate_distance,include=F,eval=T,cache=T}

# gtr_with_expert_db<-gtr_db%>%mutate(panel_id=as.character(panel_id))%>%
#   bind_rows(fixed_source_db%>%filter(source%in%c('panelapp','clingen')))#%>%
#   #bind_rows(expert_panels_source)
# 
# 
# gtr_binary_panel_gene_table<-create_binary_gene_table(gtr_with_expert_db)
# # generate distance matrix between all the panels
# gtr_dm<-generate_distance_matrix(gtr_binary_panel_gene_table)
# # convert dm to df
# gtr_dm_df<-convert_dm_to_df(gtr_dm)
# 
# # save gtr matrix and dataframe
# save(gtr_dm,gtr_binary_panel_gene_table,gtr_with_expert_db,gtr_dm_df,file='./data/pre_generated_data/gtr_data.Rdata')

# load gtr data
load('./data/pre_generated_data/gtr_data.Rdata')
```

# Define Selected panels

```{r selected_panels}
selected_phenotypes_panelapp<-c("panelapp|49|Hypertrophic cardiomyopathy - teen and adult",
                                'panelapp|544|Cholestasis',
                                'panelapp|149|Nephrocalcinosis or nephrolithiasis',
                                'panelapp|179|Hydrocephalus',
                                'panelapp|657|Autism',
                                'panelapp|126|Hearing loss',
                                'panelapp|294|Ocular coloboma',
                                'panelapp|309|Skeletal dysplasia',
                                'panelapp|386|Pancreatitis',
                                'panelapp|285|Intellectual disability',
                                'panelapp|943|Osteopetrosis',
                                'panelapp|553|Ectodermal dysplasia',
                                'panelapp|487|Cystic renal disease',
                                'panelapp|258|Arthrogryposis',
                                'panelapp|78|Holoprosencephaly',
                                #'panelapp|522|Familial melanoma', 
                                'panelapp|652|Dilated cardiomyopathy - adult and teen',
                                'panelapp|658|Corneal dystrophies',
                                'panelapp|555|Ichthyosis and erythrokeratoderma',
                                'panelapp|546|Lipodystrophy - childhood onset',
                                'panelapp|168|Craniosynostosis')
selected_phenotypes<-c('hypertrophic cardiomyopathy',
                       'cholestasis',
                       'nephrolithiasis',
                       'hydrocephalus',
                       'autism',
                       'hearing loss',
                       'coloboma',
                       'skeletal dysplasia',
                       'pancreatitis',
                       'intellectual disability',
                       'osteopetrosis',
                       'ectodermal dysplasia',
                       'polycystic kidney',
                       'Arthrogryposis',
                       'Holoprosencephaly',
                       #'melanoma',
                       'dilated cardiomyopathy',
                       'corneal dystrophy',
                       'ichthyosis',
                       'lipodystrophy',
                       'craniosynostosis')

phenotype_queries<-c('hypertrophic cardio|hcm|cardio.+hypertrophic',
                     'cholestasis',
                     'nephrolithiasis|monogenic kidney stone',
                     'hydrocephalus',
                     'autism',
                     'hearing loss|deafness',
                     'coloboma',
                     'skeletal dysplasia',
                     'pancreatitis',
                     'intellectual disability|mental retard',
                     'osteopetrosis',
                     'ectodermal dysplasia|ectoderm',
                     'Cystic renal|polycys.*kidn',
                     'Arthrogryposis',
                     'Holoprosencephaly',
                     #'melanoma',
                     'dilated cardio|dcm',
                     'corneal dyst',
                     'ichthyosis',
                     'lipodystrophy',
                     'craniosynostosis')


selected_phenotypes_df<-data.frame(phenotype_name=selected_phenotypes,
                                   panelapp_panel=selected_phenotypes_panelapp,
                                   query=phenotype_queries)


               
```

# Analyze the selected panels 

```{r run_comparison_for_selected}
all_panel_names<-gtr_dm_df%>%pull(row)%>%unique()
panelapp_panels<-grep('panelapp',all_panel_names,ignore.case = T,value = T)
all_genes<-gtr_with_expert_db%>%pull(gene_symbol)%>%unique()
max_dists_for_auto<-c(0.7,0.75,0.8,0.85,0.9)
# run once to pre-generate the per gene publications data
#pubmed_grab_all_gene_publications(all_genes)
# load the pregenerated all gene publications
load('./data/pre_generated_data/genes_pubmed_search.RData')

selected_phenotypes_analysis<-list()
for (selected_phenotype_name in selected_phenotypes){
  message(glue('Running comparison for {selected_phenotype_name}..'))
  panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
  naive_search<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(query)
  selected_phenotypes_analysis[[selected_phenotype_name]] <-
    compare_panelapp_naive_and_cluster(
      panelapp_panel = panelapp_panel,
      phenotype_name = selected_phenotype_name,
      naive_search =  naive_search,
      max_dists_for_cluster = NA,
      gtr_with_expert_db = gtr_with_expert_db,
      gtr_dm_df = gtr_dm_df,
      all_genes = all_genes,
      all_panel_names = all_panel_names,
      pre_generated_gene_publications_list = entrez_pubmed_res)
}

#save(selected_phenotypes_analysis,file='./data/pre_generated_data/selected_phenotypes_analysis.RData')
#load('./data/pre_generated_data/selected_phenotypes_analysis.RData')
# now check the ranks for each 

selected_phenotypes_analysis_df<-NULL
for (selected_phenotype in names(selected_phenotypes_analysis)){
  selected_phenotypes_analysis_df<-selected_phenotypes_analysis_df%>%
    bind_rows(data.frame(phenotype=selected_phenotype,
                         selected_phenotypes_analysis[[selected_phenotype]]$all_metrics%>%
                           mutate(rank_f1=rank(desc(f1),ties.method = 'min'),
                                  rank_recall=rank(desc(recall),ties.method = 'min'),
                                  rank_precision=rank(desc(precision),ties.method = 'min'))))
}

# Summarize the performance of panelapp vs naive
methods_performance_metrics<-selected_phenotypes_analysis_df%>%
  mutate(group=ifelse(grepl('^panelapp',group),'panelapp',group))%>%
  filter(grepl('naive|panelapp$|max_dist|^panelapp',group))%>%
  group_by(group)%>%
  select(-c(ngenes,phenotype))%>%
  pivot_longer(-c(group))%>%
  group_by(group,name)%>%
  summarize(skimr::skim_without_charts(value))
```

# Sanity check- review panelapp genes not found in search

```{r sanity_check}

panelapp_genes_not_in_search<-NULL
literature_genes_not_in_panelapp<-NULL
for (selected_phenotype in selected_phenotypes){
  panelapp_genes<-selected_phenotypes_analysis[[selected_phenotype]]$panelapp_genes
  all_genes_vs_phenotype<-pubmed_phenotype_vs_all_genes(phenotype_name = selected_phenotype,gtr_with_expert_db = gtr_with_expert_db)
  panelapp_genes_without_publications<-all_genes_vs_phenotype%>%filter(pub_num<2)%>%filter(gene_symbol%in%panelapp_genes)
  lit_genes_not_in_panelapp<-all_genes_vs_phenotype%>%filter(pub_num>1)%>%filter(!gene_symbol%in%panelapp_genes)
  panelapp_genes_not_in_search[[selected_phenotype]]<-panelapp_genes_without_publications
  literature_genes_not_in_panelapp[[selected_phenotype]]<-lit_genes_not_in_panelapp
}

panelapp_genes_not_in_search$`hypertrophic cardiomyopathy`
```


```{r generate auto panels}
auto_generator_panelapp_only<-list()
auto_generator_panelapp_with_similar<-list()
max_dists=c(0.75,0.8,0.85,0.9,0.95)
for (max_dist in max_dists){
  auto_generator_panelapp_only[[glue('max_dist={max_dist}')]]<-list()
  for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
    message(glue('Generating panel for {selected_phenotype_name} max dist = {max_dist}'))
    panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
    # similar_panels<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,
    #                                                       selected_panel_name = panelapp_panel,
    #                                                       top_most_similar = 1,maximal_distance = 0.7,
    #                                                       names_only = T)
    auto_generator_panelapp_only[[glue('max_dist={max_dist}')]][[selected_phenotype_name]] <-
      generate_panel_from_panels(
        joined_panel_names = panelapp_panel,
        gen_source_db = gtr_with_expert_db,
        gen_dm_df = gtr_dm_df,
        max_distance = max_dist,
        with_mitochondrial = F)
    # auto_generator_panelapp_with_similar[[glue('max_dist={max_dist}')]][[selected_phenotype_name]] <-
    #   generate_panel_from_panels(
    #     joined_panel_names = similar_panels,
    #     gen_source_db = gtr_with_expert_db,
    #     gen_dm_df = gtr_dm_df,
    #     max_distance = max_dist,
    #     with_mitochondrial = F)
  }
}

```

# ROC analysis

```{r roc_analysis}

# now add whether the gene is positive or not
auto_generator_rocs<-NULL
panelapp_roc<-NULL
naive_roc<-NULL
min_num_of_pub_to_consider_positive<-2

for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
  panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
  all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype_name,
                                                            gene_list = all_genes,
                                                            pre_generated_genes_pubs = entrez_pubmed_res)
  all_genes_vs_phenotype<-all_genes_vs_phenotype%>%
    mutate(has_pub=factor(ifelse(pub_num>=min_num_of_pub_to_consider_positive,1,0)))
  for (max_dist in max_dists){
    auto_generator_panel_res<-auto_generator_panelapp_only[[glue('max_dist={max_dist}')]][[selected_phenotype_name]]
    phenotype_auto_generator_pr_curve <-
      calculate_auto_generator_pr_for_phenotype(phenotype_name = selected_phenotype_name,
                                                auto_generator_panel_res = auto_generator_panel_res,
                                                all_genes_vs_phenotype = all_genes_vs_phenotype)
    auto_generator_rocs<-auto_generator_rocs%>%
      bind_rows(data.frame(phenotype_auto_generator_pr_curve,max_dist=max_dist)%>%
                  mutate(group=glue('ag_{max_dist}')))# MIND THAT IF THE AUTO GENERATOR ROC IS FOR MORE THAN ONE SCORE (INCLUDES THE MODIFIED) YOU NEED TO GIVE EACH SCORE A DIFFERENT GROUP NAME
    
  }
  
  # calculate the panelapp recall and precision
  panelapp_genes<-gtr_with_expert_db%>%filter(panel_joined_name==panelapp_panel)%>%pull(gene_symbol)
  panelapp_phenotype_pr<-calculate_sens_and_ppv_vs_pubmed(panelapp_genes,
                                         all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol),
                                         precision_recall_naming = T)
  
  panelapp_roc<-panelapp_roc%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         group='panelapp',
                         recall=panelapp_phenotype_pr$recall,
                         precision=panelapp_phenotype_pr$precision,
                         f1=panelapp_phenotype_pr$f1))
  
  # calculate the naive search (with panelapp) recall and precision
  naive_genes<-selected_phenotypes_analysis[[selected_phenotype_name]]$naive_with_panelapp_genes
  naive_phenotype_pr<-calculate_sens_and_ppv_vs_pubmed(naive_genes,
                                         all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol),
                                         precision_recall_naming = T)
  naive_roc<-naive_roc%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         group='naive search',
                         recall=naive_phenotype_pr$recall,
                         precision=naive_phenotype_pr$precision,
                         f1=naive_phenotype_pr$f1))
  
}

all_rocs<-auto_generator_rocs%>%
  bind_rows(panelapp_roc)%>%
  bind_rows(naive_roc)

# get the max f1 value per max dist per phenotype 
all_rocs%>%
  group_by(phenotype_name,group)%>%
  slice_max(n=1,order_by = f1,with_ties = F)%>%
  ungroup()%>%
  select(group,f1,precision,recall)%>%
  pivot_longer(-group)%>%
  ggplot(aes(x=group,y=value,fill=group))+
  geom_boxplot()+#coord_flip()+
  facet_wrap(name~.)+
  guides(fill=F)+
  ggsci::scale_fill_nejm()+
  theme_minimal()
```

# compare selected auto generated roc against panelapp and naive

```{r ag_vs_pa_vs_ns}

# add the max f1 stats for all the panels to the metrics df
selected_dist<-0.85
selected_ag_roc<-auto_generator_rocs%>%filter(!(!is.na(max_dist)&group!=glue('ag_{selected_dist}')))

# add number of genes
library(listr)
auto_panels_genes<-auto_generator_panelapp_only[[glue('max_dist={selected_dist}')]] %>%list_name_to_df()%>%list_bind(selected_phenotypes)
auto_panels_genes<-auto_panels_genes[[1]]%>%select(phenotype_name=.group,gene_symbol,scaled_gene_score)

# for each threshold calculate the number of genes
number_of_genes_above_thresh<-list()
for (selected_phenotype in selected_phenotypes){
  message(glue('Adding number of genes per threshold for {selected_phenotype}'))
  number_of_genes_above_thresh[[selected_phenotype]]<-NULL
  phenotype_auto_panels<-auto_panels_genes%>%filter(phenotype_name==selected_phenotype)
  threshs <- selected_ag_roc%>%filter(phenotype_name==selected_phenotype) %>% pull(.threshold)
  for (thresh in threshs){
    number_of_genes_above_thresh[[selected_phenotype]]<-
      number_of_genes_above_thresh[[selected_phenotype]]%>%
      bind_rows(data.frame(.threshold=thresh,
                           ngenes=phenotype_auto_panels%>%filter(scaled_gene_score>=thresh)%>%nrow()))
    
  }
}

# now flatten the list using listr
ngenes_to_add<-number_of_genes_above_thresh%>%list_name_to_df()%>%list_bind(selected_phenotypes)
ngenes_to_add<-ngenes_to_add[[1]]%>%rename(phenotype_name=.group)

selected_ag_roc<-selected_ag_roc%>%
  left_join(ngenes_to_add)


comparison_metrics<-selected_phenotypes_analysis_df%>%
  select(-contains('rank'))%>%
  filter(grepl('panelapp|naive',group),!grepl('no panelapp',group))%>%
  mutate(group=ifelse(group!='panelapp','naive',group))%>%
  rename(phenotype_name=phenotype)

panelapp_metrics<-comparison_metrics%>%filter(grepl('^panelapp',group))
auto_matched_to_panelapp_by_precision<-selected_ag_roc%>%left_join(panelapp_metrics,by='phenotype_name')%>%
  group_by(phenotype_name)%>%
  filter(abs(precision.x-precision.y)==min(abs(precision.x-precision.y)))%>%
  slice_max(n=1,order_by=recall.x)

# now do the same for the naive search, but this time match by recall (in order to show better precision)
naive_metrics<-comparison_metrics%>%filter(grepl('naive',group))
auto_matched_to_naive_by_recall<-selected_ag_roc%>%left_join(naive_metrics,by='phenotype_name')%>%
  group_by(phenotype_name)%>%
  filter(abs(recall.x-recall.y)==min(abs(recall.x-recall.y)))%>%
  slice_max(n=1,order_by=precision.x)


comparison_metrics<-
  comparison_metrics%>%
  # now add the max f1 for each phenotype
  bind_rows(selected_ag_roc%>%group_by(phenotype_name)%>%
              slice_max(n=1,order_by = f1)%>%
              mutate(group='distance_max_f1')%>%
              select(-c(.threshold,max_dist)))%>%
  # now add the max recall to the closest ppv
  bind_rows(auto_matched_to_panelapp_by_precision%>%
              select(phenotype_name,ngenes=ngenes.x,group=group.x,f1=f1.x,precision=precision.x,recall=recall.x)%>%
              mutate(group='distance_matched_to_panelapp_percision'))%>%
  # now add max precision to the closest recall
  bind_rows(auto_matched_to_naive_by_recall%>%
              select(phenotype_name,ngenes=ngenes.x,group=group.x,f1=f1.x,precision=precision.x,recall=recall.x)%>%
              mutate(group='distance_matched_to_naive_recall'))


excluded_phenos<-c('melanoma')
comparison_metrics<-comparison_metrics%>%filter(!phenotype_name%in%excluded_phenos)
# plot max f1 vs panelapp vs naive
comparison_metrics%>%filter(!grepl('distance_(naive|panel)',group))%>%ggplot(aes(x=group,y=f1,fill=group))+
  geom_col()+coord_flip()+
  facet_grid(phenotype_name~.)+
  guides(fill=F)+ggsci::scale_fill_nejm()+
  theme_minimal()

# Plot precision recall curves and mark the pr for panelapp and the naive search
selected_ag_roc%>%ggplot(aes(x=recall,y=precision,col=factor(group)))+
  geom_line()+facet_wrap(phenotype_name~.)+
  geom_abline(slope = -1,intercept = 1,linetype=2,alpha=0.5)+
  geom_point(data=panelapp_roc,inherit.aes = F,aes(x=recall,y=precision))+
  geom_point(data=naive_roc,inherit.aes = F,aes(x=recall,y=precision),col='red')+
  geom_point(data=comparison_metrics%>%filter(group=='distance_max_f1'),inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  #geom_point(data=dist_roc,inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  theme_minimal()
```

# Compare vs clingen

```{r clingen_comp}

clingen_panels<-gtr_with_expert_db%>%filter(source=='clingen')%>%pull(panel_joined_name)%>%unique()

clingen_vs_selected<-data.frame(
  phenotype_name=c('dilated cardiomyopathy',
                       'hypertrophic cardiomyopathy',
                       'hearing loss'),
  clingen_panel=c("clingen|5|Dilated Cardiomyopathy",
                  "clingen|11|Hypertrophic Cardiomyopathy",
                  'clingen|7|Hearing Loss')
)

for (selected_phenotype_name in clingen_vs_selected$phenotype_name){
  clingen_panel<-clingen_vs_selected%>%filter(phenotype_name==selected_phenotype_name)%>%pull(clingen_panel)
  clingen_genes<-gtr_with_expert_db%>%filter(panel_joined_name==clingen_panel,confidence_level>1)%>%pull(gene_symbol)
  all_genes_vs_clingen<-data.frame(gene_symbol=all_genes)%>%mutate(has_pub=factor(ifelse(gene_symbol%in%clingen_genes,1,0)))
  selected_ag<-auto_generator_panelapp_only[[glue('max_dist={selected_dist}')]][[selected_phenotype_name]]
  ag_roc <- calculate_auto_generator_pr_for_phenotype(phenotype_name = selected_phenotype_name,
                                                      auto_generator_panel_res = selected_ag,
                                                      all_genes_vs_phenotype = all_genes_vs_clingen)
  g<-ag_roc%>%ggplot(aes(x=recall,y=precision,col=factor(group)))+
  geom_line()+facet_wrap(phenotype_name~.)+
  geom_abline(slope = -1,intercept = 1,linetype=2,alpha=0.5)+
  #geom_point(data=dist_roc,inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  theme_minimal()
  print(g)
}

```

# Statistical analysis

```{r statistical_analysis}
f1_comp<-comparison_metrics%>%ungroup()%>%select(phenotype_name,group,f1)%>%pivot_wider(names_from = group,values_from = f1)
recall_comp<-comparison_metrics%>%ungroup()%>%select(phenotype_name,group,recall)%>%pivot_wider(names_from = group,values_from = recall)
precision_comp<-comparison_metrics%>%ungroup()%>%select(phenotype_name,group,precision)%>%pivot_wider(names_from = group,values_from = precision)

# in cases where the precision was matched, check how many times the distance method achieved higher recall than panelapp
better_recall_vs_panelapp <-
  recall_comp %>% filter(distance_panelapp_percision > panelapp) %>%
  filter(phenotype_name %in% (precision_comp %>% filter(abs(distance_panelapp_percision -
                                                             panelapp) < 0.05) %>% pull(phenotype_name)))
# in cases where the recall was matched, check how many times the distance method achieved higher precision than naive
better_precision_vs_naive <-
  precision_comp %>% filter(distance_naive_recall > naive) %>%
  filter(phenotype_name %in% (recall_comp %>% filter(abs(distance_naive_recall -
                                                             naive) < 0.05) %>% pull(phenotype_name)))

comparison_text<-paste0(glue('Compared to a literature search, PanelApp panels demonstrated '),
                        paste0(comparison_metrics_text%>%filter(group=='panelapp')%>%pull(metric_text), 
                        collapse=', '),
                        glue('. The unified panels based on naive search demonstrated '),
                        paste0(comparison_metrics_text%>%filter(group=='naive')%>%pull(metric_text), 
                        collapse=', '),
                        glue('. Distance based generated panels demonstrated '),
                        paste0(comparison_metrics_text%>%filter(group=='distance_max_f1')%>%pull(metric_text), 
                        collapse=', '),
                        glue('. For {nrow(f1_comp%>%filter(distance_max_f1>panelapp))}/{nrow(f1_comp)} ({round(100*nrow(f1_comp%>%filter(distance_max_f1>panelapp))/nrow(f1_comp),1)}%) of the phenotypes, the distance based panel achieved a higher F1 score compared to the PanelApp panel'),
                        glue('. For {nrow(better_recall_vs_panelapp)}/{nrow(recall_comp)} ({round(100*nrow(better_recall_vs_panelapp)/nrow(recall_comp),1)}%) of the phenotypes, the distance based panel achieved a higher sensitivity compared to the PanelApp panel while maintaining a comparable PPV'),
                        glue('. For {nrow(f1_comp%>%filter(distance_max_f1>naive))}/{nrow(f1_comp)} ({round(100*nrow(f1_comp%>%filter(distance_max_f1>naive))/nrow(f1_comp),1)}%) of the phenotypes, the distance based panel achieved a higher F1 score compared to the naive search panel'),
                        glue('. For {nrow(better_precision_vs_naive)}/{nrow(recall_comp)} ({round(100*nrow(better_precision_vs_naive)/nrow(recall_comp),1)}%) of the phenotypes, the distance based panel achieved a higher PPV compared to the naive search panel while maintaining a comparable sensitivity'),
                        coollapse=' ') 
comparison_text
# calculate between group comparisons
params_list <- combn(unique(comparison_metrics$group), 2, FUN = list)

# for some reason this doesnt work
# # perform t test for each combination
# model_t <- map(.x = params_list,
#                .f = ~ tidy(t.test(
#                  formula = recall ~ group,
#                  paired=F,
#                  data    = subset(comparison_metrics%>%filter(!(phenotype_name %in% c('cleft lip palate'))), group %in% .x),
#                  conf.int=T
#                )))
# names(model_t)<-params_list
# model_t

# but this works and seems more legit - you can do this for these comparisons
# distance f1 max vs panelapp f1
# distance precision matched recall vs panelapp recall - to show that for a given PPV level, sensitivity is better
# distance recall matched precision vs naive precision - to show that for a given sensitivity level, PPV is better
library(report)
precision_distance_vs_naive_recall_matched<-t.test(precision_comp%>%pull(distance_naive_recall),precision_comp%>%pull(naive),paired = T)
report(precision_distance_vs_naive_recall_matched)
precision_distance_vs_naive_recall_matched_df<-tidy(precision_naive_vs_distance_recall_matched)
# make sure the recall is indeed matched
t.test(recall_comp%>%pull(distance_naive_recall),recall_comp%>%pull(naive),paired = T)

precision_matched_recall_comp<-precision_comp%>%filter(panelapp-distance_panelapp_percision<0.01)
recall_distance_vs_panelapp_precision_matched<-t.test(recall_comp%>%pull(distance_panelapp_percision),recall_comp%>%pull(panelapp),paired = T)
report(recall_distance_vs_panelapp_precision_matched)
# make sure the precision is indeed matched
precision_comp%>%select(distance_panelapp_percision,panelapp)%>%colMeans()
t.test(precision_comp%>%pull(distance_panelapp_percision),precision_comp%>%pull(panelapp),paired = T)


f1_distance_vs_panelapp_max<-t.test(f1_comp%>%pull(distance_max_f1),f1_comp%>%pull(panelapp),paired = T)
report(f1_distance_vs_panelapp_max)

f1_distance_vs_naive_max<-t.test(f1_comp%>%pull(distance_max_f1),f1_comp%>%pull(naive),paired = T)

report(f1_distance_vs_naive_max)

```

# Summary text

```{r summary_text}
# Source summary text
source_summary_text<-paste0(glue('After exclusion, a total of {length(unique_panels)} panels from {length(unique_sources)} different vendors were collected from the GTR and {length(unique_panelapp_panels)} panels were collected from PanelApp.'),
                            glue('The median number of genes per panel was {number_of_genes_per_panel_summary$numeric.p25} (IQR[{number_of_genes_per_panel_summary$numeric.p50},{number_of_genes_per_panel_summary$numeric.p75}])'),
                            collapse=' ')

source_summary_text

# Summarize the selected phenotypes
selected_phenotypes_summary_text<-paste0(glue('In order to demonstrate the discrepancy between available panels, and the utility of our method, {nrow(selected_phenotypes_df)} phenotypes were selected: {paste0("(",seq(1,20),") ",selected_phenotypes,collapse=", ")}.'),
                                         collapse=' ')
selected_phenotypes_summary_text   

# Discrepancy throughout selected panels
percent_of_genes_in_less_than_10pct_of_panels_df<-NULL
for (selected_phenotype_name in selected_phenotypes){
  relatedness_df<-selected_phenotypes_analysis[[selected_phenotype_name]]$naive_discrepancy$relatedness
  percent_of_genes_in_less_than_10pct_of_panels<-relatedness_df$percent_of_panels%>%filter(percent_of_panels_cat=='0-10%')%>%pull(rate)
  if (length(percent_of_genes_in_less_than_10pct_of_panels)==0){percent_of_genes_in_less_than_10pct_of_panels<-0}
  percent_of_genes_in_less_than_10pct_of_panels_df<-percent_of_genes_in_less_than_10pct_of_panels_df%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         num_of_panels=relatedness_df$num_of_panels,
                         percent_of_genes_in_less_than_10pct_of_panels))
}
num_o_panels_summary<-skimr::skim_without_charts(percent_of_genes_in_less_than_10pct_of_panels_df$num_of_panels)

percent_less_than_10_summary<-skimr::skim_without_charts(percent_of_genes_in_less_than_10pct_of_panels_df$percent_of_genes_in_less_than_10pct_of_panels*100)

discrepancy_text<-paste0(glue('For the selected phenotypes, a naive search identified, on average, {round(num_o_panels_summary$numeric.mean,1)} panels per phenotype (IQR [{num_o_panels_summary$numeric.p25},{num_o_panels_summary$numeric.p75}]).'),
                         glue('For most phenotypes, the majority of the genes ({round(percent_less_than_10_summary$numeric.mean,1)}% of genes) IQR[{round(percent_less_than_10_summary$numeric.p25,1)}%,{round(percent_less_than_10_summary$numeric.p75,1)}%] were found in less than 10% of the collected panels.'),
                         collapse=' ')
discrepancy_text

# Literature text
comparison_metrics_summary<-comparison_metrics%>%
  group_by(group)%>%select(precision,recall,f1)%>%skimr::skim_without_charts()

comparison_metrics_text<-comparison_metrics_summary%>%group_by(group,skim_variable)%>%
  summarize(metric_text=glue('a mean {skim_variable} of {round(numeric.mean,2)} (IQR[{round(numeric.p25,2)},{round(numeric.p75,2)}])'))
```



```{r pca}
# Add PCA calculation
test_binary<-gtr_binary_panel_gene_table[1:1000,1:1000]
library(logisticPCA)
logsvd_model = logisticSVD(test_binary, k = 2)
plot(logsvd_model)

panels = rownames(test_binary)
plot(logsvd_model, type = "scores") + geom_point(aes()) + 
  ggtitle("Exponential Family PCA") + scale_colour_manual(values = c("blue", "red"))

library(FactoMineR)
test_binary<-gtr_binary_panel_gene_table[,]
res.mca <- FactoMineR::PCA(test_binary, 
              ncp=2,graph=F)
coords<-data.frame(panel=rownames(res.mca$ind$coord),res.mca$ind$coord)%>%
  mutate(source=stringr::str_replace(panel,'\\|.+',''))

to_plot<-coords%>%filter(panel%in%naive_panels)%>%mutate(group='naive_panels')%>%
  bind_rows(coords%>%filter(panel%in%dist_panels)%>%mutate(group='dist_panels'))

to_plot%>%
  #filter(Dim.1<50,Dim.2<20)%>%
  ggplot(aes(x=Dim.1,y=Dim.2,col=factor(source=='panelapp')))+
  geom_point(alpha=0.4)#+facet_wrap(group~.)+
  theme_minimal()

```


```{r check_dist_distribution_for_panelapp_panels}
panelapp_panels_dist<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%group_by(row)%>%
  filter(col!=row)%>%
  summarize(mean_dist=mean(value),
            median_dist=median(value),
            min_dist=min(value))%>%
  left_join(gtr_with_expert_db%>%
              filter(source=='panelapp')%>%
              mutate(panel_joined_name=as.character(panel_joined_name))%>%
              group_by(panel_joined_name)%>%
              summarize(n_genes=n()),
            by=c('row'='panel_joined_name'))

panelapp_panels_dist%>%filter(n_genes<1000)%>%ggplot(aes(x=n_genes,y=min_dist))+
  geom_point()+
  theme_minimal()

```



```{r cluster_panels}
# hierarchal clustering
hc <- hclust(gtr_dm,method='ward.D' )
# cut tree
cut_avg <- cutree(hc,k = 100)
panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
# this is not bad - it actually groups similar panels together.. but what is the optimal number of clusters??
# basically the purpose is to reduce false positives while maintaining high sensitivity - so lets see what happens when i divide the panels into X clusters and go from there
panel_clusters%>%dplyr::count(cluster)%>%arrange(n)

```



```{r analyze_distance_matrix}
# Analyze the ratio between the number of genes in a panel and the distance it has with all other panels with the same approximate size

dm_to_analyze<-gtr_dm_df%>%
  left_join(gtr_panels_list%>%select(panel_joined_name,number_of_genes_row=number_of_genes),by=c('row'='panel_joined_name'))%>%
  left_join(gtr_panels_list%>%select(panel_joined_name,number_of_genes_col=number_of_genes),by=c('col'='panel_joined_name'))

dm_to_analyze<-dm_to_analyze%>%mutate(ngenes_cat_row=cut(number_of_genes_row,breaks=c(0,10,30,50,100,200,500,1000,10000)),
                                      ngenes_cat_col=cut(number_of_genes_col,breaks=c(0,10,30,50,100,200,500,1000,10000)))


head(dm_to_analyze)
dm_to_analyze%>%slice_sample(n=100000)%>%ggplot(aes(x=number_of_genes_row,y=value))+
  geom_point()+facet_wrap(ngenes_cat_row~.,scales='free')
  

```

```{r compare_panelapp_against_the_rest}
# This code goes through all of panelapps panels and compares them against all the panels with at most max_dist distance. 
# it then calculates sensitivity and precision for all the panels and outputs how many panels had missing genes in them

panelapp_df<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%filter(!grepl('panelapp',col))
panelapp_df<-panelapp_df%>%
  left_join(panels_list%>%
              mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))

# Check whether how well the different panelapp panels are represented by looking for the sensitivity 
max_dist<-c(1,0.99,0.95,0.9,0.85)
panelapp_vs_all_other<-NULL
for (md in max_dist){
  message(glue('calculating panelapp similarity for {md} distance'))
  panelapp_comp<-panelapp_df%>%group_by(row)%>%filter(value<md)%>%ungroup()
  panelapp_comp<-panelapp_comp%>%group_by(row)%>%dplyr::summarize(num_o_comp_panels=n(),compare_panel_vs_many_panels(row,unique(col),gtr_with_expert_db,min_confidence_level = 2))%>%mutate(max_dist=md)
  
  panels_without_similar_panels<-setdiff(panelapp_df%>%pull(row)%>%as.character()%>%unique(),panelapp_comp%>%pull(row))
  if (length(panels_without_similar_panels)>0){
    panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                 num_o_comp_panels=0,
                                                 genes_in_b=0,
                                                 common_genes=0,
                                                 B_not_in_A=0,
                                                 Agenes_not_in_B='',
                                                 max_dist=md)
    panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
      genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
      A_not_in_B=genes_in_a
    )
    panelapp_comp<-panelapp_comp%>%bind_rows(panels_without_similar_panels_df)
  }
  
  panelapp_vs_all_other<-panelapp_vs_all_other%>%bind_rows(panelapp_comp)
}

panelapp_vs_all_other<-panelapp_vs_all_other%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)

panelapp_vs_all_other%>%group_by(max_dist)%>%dplyr::summarize(mean_sens=mean(sensitivity),
                                              mean_prec=mean(precision),
                                              num_o_panels_with_missing_genes=sum(A_not_in_B>0))



```


```{r compare_panelapp_against_the_rest_cluster_analysis}
# This code goes through all of panelapps panels and compares them against all the panels with at most max_dist distance. 
# it then calculates sensitivity and precision for all the panels and outputs how many panels had missing genes in them



panelapp_vs_all_other_cluster<-NULL
for (method in c('ward.D2','ward.D')){
  print(method)
  for (cluster_size in c(1,75,150)){
    print(cluster_size)
    # hierarchal clustering
    hc <- hclust(gtr_dm,method=method )
    
    # library(cluster)
    # z<-agnes(gtr_dm)
    # cut tree
    cut_avg <- cutree(hc,k = cluster_size)
    panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
    # this is not bad - it actually groups similar panels together.. but what is the optimal number of clusters??
    # basically the purpose is to reduce false positives while maintaining high sensitivity - so i lets see what happens when i divide the panels into X clusters and go from there
    panel_clusters%>%dplyr::count(cluster)%>%arrange(n)
    
    
    panelapp_df<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%filter(!grepl('panelapp',col))
    panelapp_df<-panelapp_df%>%
      left_join(panels_list%>%
                  mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
      left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
      left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)
    
    # now do the same but only within the panel cluster
    # Check whether how well the different panelapp panels are represented by looking for the sensitivity 
    max_dist<-c(1,0.99,0.95,0.9)
    
    for (md in max_dist){
      message(glue('calculating panelapp similarity for {md} distance'))
      panelapp_comp<-panelapp_df%>%group_by(row)%>%filter(value<md)%>%filter(row_cluster==col_cluster)%>%ungroup()
      panelapp_comp<-panelapp_comp%>%group_by(row)%>%
        dplyr::summarize(num_o_comp_panels=n(),
                         compare_panel_vs_many_panels(row,unique(col),
                                                      gtr_with_expert_db,
                                                      min_confidence_level = 2))%>%
        mutate(max_dist=md)
      
      panels_without_similar_panels<-setdiff(panelapp_df%>%pull(row)%>%as.character()%>%unique(),panelapp_comp%>%pull(row))
      if (length(panels_without_similar_panels)>0){
        panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                     num_o_comp_panels=0,
                                                     genes_in_b=0,
                                                     common_genes=0,
                                                     B_not_in_A=0,
                                                     Agenes_not_in_B='',
                                                     max_dist=md)
        panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
          genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
          A_not_in_B=genes_in_a
        )
        panelapp_comp<-panelapp_comp%>%bind_rows(panels_without_similar_panels_df)
      }
      
      panelapp_vs_all_other_cluster<-panelapp_vs_all_other_cluster%>%
        bind_rows(data.frame(method=method,cluster_size=cluster_size,panelapp_comp))
    }
    
    
  }
}

panelapp_vs_all_other_cluster<-panelapp_vs_all_other_cluster%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)%>%rowwise()%>%mutate(f1= (2 * precision * sensitivity) / (precision + sensitivity))

cluster_analysis_res<-panelapp_vs_all_other_cluster %>%
  group_by(method, cluster_size, max_dist) %>%
  dplyr::summarize(mean_sens = mean(sensitivity),
                   mean_prec = mean(precision),
                   mean_f1 = mean(f1,na.rm=T),
                   num_o_panels_with_missing_genes=sum(A_not_in_B>0))

cluster_analysis_res%>%filter(method=='ward.D2')%>%select(-num_o_panels_with_missing_genes)%>%
  pivot_longer(-c(cluster_size,max_dist,method))%>%
  ggplot(aes(x=factor(cluster_size),y=value,fill=method))+
  geom_col(position='dodge')+facet_wrap(name~max_dist,scale='free')+
  theme_minimal()

cluster_analysis_re(2 * precision * sensitivity) / (precision + sensitivity)s%>%filter(method=='ward.D')%>%select(-num_o_panels_with_missing_genes)%>%
  pivot_longer(-c(cluster_size,max_dist,method))%>%
  ggplot(aes(x=cluster_size,y=value,col=factor(max_dist)))+
  geom_line()+facet_wrap(name~.,scale='free')+
  theme_minimal()+
  ggsci::scale_color_nejm()+
  theme(legend.position = 'top')


```


```{r compare_clingen_vs_panelapp_plus_gtr}


#max_dist<-c(1,0.99,0.95,0.9,0.85)
max_dist<-c(1,0.99,0.98,0.975,0.97,0.95)
cluster_sizes<-c(1,50,80,100)
cluster_method<-'ward.D'
clingen_vs_all_other<-NULL

for(cluster_size in cluster_sizes){
  
  hc <- hclust(gtr_dm,method=cluster_method )
  # library(cluster)
  # z<-agnes(gtr_dm)
  # cut tree
  cut_avg <- cutree(hc,k = cluster_size)
  panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
  
  # for every clingen panel - build a panel and see how it covers it
  clingen_df<-gtr_dm_df%>%filter(grepl('clingen',row))%>%filter(!grepl('clingen',col))

  
  clingen_df<-clingen_df%>%
        left_join(pan(2 * precision * sensitivity) / (precision + sensitivity)els_list%>%
                    mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
        left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
        left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)
  
  for (md in max_dist){
    message(glue('calculating clingen similarity for {md} distance'))
    clingen_comp<-clingen_df%>%group_by(row)%>%filter(value<md)%>%ungroup()
    clingen_comp<-clingen_df%>%group_by(row)%>%filter(value<md)%>%filter(row_cluster==col_cluster)%>%ungroup()
    clingen_comp<-clingen_comp%>%group_by(row)%>%
      summarize(num_o_comp_panels=n(),
                compare_panel_vs_many_panels(row,unique(col),gtr_with_expert_db,min_confidence_level = 3))%>%
      mutate(max_dist=md,cluster_size=cluster_size)
    panels_without_similar_panels<-setdiff(clingen_df%>%pull(row)%>%as.character()%>%unique(),clingen_comp%>%pull(row))
    if (length(panels_without_similar_panels)>0){
      panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                   num_o_comp_panels=0,
                                                   genes_in_b=0,
                                                   common_genes=0,
                                                   B_not_in_A=0,
                                                   Agenes_not_in_B='',
                                                   max_dist=md)
      panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
        genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
        A_not_in_B=genes_in_a
      )
      clingen_comp<-clingen_comp%>%bind_rows(panels_without_similar_panels_df)
    }
    clingen_vs_all_other<-clingen_vs_all_other%>%bind_rows(clingen_comp)
  }
}

clingen_vs_all_other<-clingen_vs_all_other%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)


clingen_test_res<-clingen_vs_all_other%>%group_by(cluster_size,max_dist)%>%summarize(mean_sens=mean(sensitivity),
                                              mean_prec=mean(precision),
                                              num_o_panels_with_missing_genes=sum(A_not_in_B>0))
z<-clingen_vs_all_other%>%filter(max_dist==0.9)
```

```{r compare_auto_vs_vendor}

hc <- hclust(gtr_dm,method='ward.D' )
# library(cluster)
# z<-agnes(gtr_dm)
# cut tree
cut_avg <- cutree(hc,k = 100)
panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))

auto_df<-gtr_dm_df%>%
      left_join(panels_list%>%
                  mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
      left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
      left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)

# start with a panelapp panel
all_panel_names<-auto_df%>%pull(row)%>%unique()
auto_df%>%select(row,row_cluster)%>%distinct()%>%filter(grepl('panelapp',row),grepl('hypertroph',row,ignore.case = T))%>%pull(row,row_cluster)
panelapp_panel<-"panelapp|49|Hypertrophic cardiomyopathy - teen and adult" 
cluster=35
same_cluster_panels<-auto_df%>%select(col,col_cluster)%>%distinct()%>%filter(col_cluster==35)%>%pull(col)

# build hypertrophy panel according to panelapp panel
auto_panel<-generate_panel_from_panels(joined_panel_names = panelapp_panel,
                                       gtr_with_expert_db%>%
                                         filter(panel_joined_name%in%c(panelapp_panel,same_cluster_panels)),
                                       auto_df,
                                       max_distance = 1,
                                       with_mitochondrial = T)

# blueprint
blueprint_panel<-source_db%>%filter(source=='blueprint',grepl('hypertrophic',panel_joined_name,ignore.case = T))
# invitae
invitae_panel<-source_db%>%filter(source=='invitae',grepl('hypertrophic',panel_joined_name,ignore.case = T))
# cegat
cegat_panel<-source_db%>%filter(source=='cegat',grepl('hypertrophic',panel_joined_name,ignore.case = T))

library(UpSetR)
panels_mat<-auto_panel%>%filter(adjusted_rank_percentile>0.5)%>%
  select(gene_symbol)%>%mutate(auto_panel=1)%>%
  pivot_wider(names_from = gene_symbol,values_from = auto_panel)%>%
  bind_rows(blueprint_panel%>%select(gene_symbol)%>%mutate(blueprint=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = blueprint,values_fill = NA))%>%
  bind_rows(invitae_panel%>%select(gene_symbol)%>%mutate(invitae=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = invitae,values_fill = NA))%>%
  bind_rows(cegat_panel%>%select(gene_symbol)%>%mutate(cegat=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = cegat,values_fill = NA))%>%
  t()%>%as.data.frame()
  
colnames(panels_mat)<-c('auto_panel','blueprint','invitae','cegat')
panels_mat[is.na(panels_mat)]<-0
upset(panels_mat)

# maybe at this point i will perform an automated pubmed search? and see how many genes have the association ?
```


```{r}

# now for hypertrophic cardiomyopathy - get the most associated panels
selected_panel_name<-find_panel_in_dm_df(dm_df,source='panelapp',panel_name='Hypertrophic cardio')
# get similar panels
similar_panels_df<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,selected_panel_name,top_most_similar = 0.005,maximal_distance = 0.99)

similar_panels_binary_gene_table<-subset_binary_gene_table(gtr_binary_panel_gene_table,similar_panels_df$col)
plot_panel_gene_heatmap(similar_panels_binary_gene_table,sort_by_similarity_to_selected_panel = T)


gbin_for_heatmap<-bin_for_heatmap%>%mutate(panel=row.names(bin_for_heatmap))%>%pivot_longer(-panel)%>%left_join(similar_panels_df,by=c('panel'='col'))
p <- ggplot(gbin_for_heatmap, aes(forcats::fct_reorder(panel, value.y), name, fill= value.x)) + 
  geom_tile()+coord_flip()+theme(axis.text.x = element_text(angle=90))
p

panels_to_plot<-as.character(unique(panelapp_hcm$col))
panels_num<-match(panels_to_plot,row.names(test_bin_data))
selected_bin_data<-test_bin_data[panels_num,]
selected_dm<-generate_distance_matrix(selected_bin_data)
plot_dm_as_dendogram(selected_dm,selected_panel)
# generate heatmap !!!

panels<-row.names(binary_panel_gene_table)[grepl('hypertroph',row.names(binary_panel_gene_table),ignore.case = T) & grepl('cardio',row.names(binary_panel_gene_table),ignore.case=T) & !(grepl('hpo',row.names(binary_panel_gene_table),ignore.case = T))]

```

```{r auto_generate_panel}

#joined_panel_names<-c('invitae|02261|Invitae Hypertrophic Cardiomyopathy Panel','health_in_code|1_1|Hypertrophic Cardiomyopathy – Basic Panel',"blueprint|1_16|Hypertrophic Cardiomyopathy (HCM) Panel",'cegat|3_2|Cardiomyopathy, hypertrophic','health_in_code|1_2|Hypertrophic Cardiomyopathy – Extended Panel')
# gen_source_db<-fixed_source_db%>%filter(!grepl('clingen|panelapp|hpo',source))
# gen_dm_df<-dm_df%>%filter(!grepl('clingen|panelapp',row))%>%filter(!grepl('clingen|panelapp|hpo',col))


joined_panel_names<-c('expert_panel|49|Hypertrophic cardiomyopathy - teen and adult')
joined_panel_names<-c('panelapp|49|Hypertrophic cardiomyopathy - teen and adult')
joined_panel_names<-c('panelapp|652|Dilated cardiomyopathy - adult and teen')
joined_panel_names<-c('Invitae|514924|Invitae Hypertrophic Cardiomyopathy Panel','Blueprint Genetics|552625|Hypertrophic Cardiomyopathy (HCM) Panel','Health in Code S.L.|530670|Hypertrophic cardiomyopathy extended panel')
joined_panel_names<-c('panelapp|555|Ichthyosis and erythrokeratoderma')

comp_panel<-expert_panels_source%>%filter(grepl('Hypertrophic',panel_joined_name))
comp_panel<-fixed_source_db%>%filter(grepl('panelapp',source))%>%filter(grepl('Hypertrophic',panel_joined_name))
comp_panel<-readr::read_csv('./data/accessory_files/hypertrophic_cardiomyopathy_basic.csv',col_names = 'gene_symbol')
comp_panel<-readr::read_csv('./data/accessory_files/hypertrophic_cardiomyopathy_extended.csv',col_names = 'gene_symbol')%>%mutate(gene_symbol=stringr::str_replace(gene_symbol,'\\*',''))
comp_panel<-readr::read_delim('./data/accessory_files/ichthyosis_erythrokeratodermi.csv')
comp_panel$gene_symbol<-comp_panel$gene_symbol%>%trimws()

gen_source_db<-gtr_with_expert_db
gen_dm_df<-gtr_dm_df%>%filter(!grepl('panelapp|clingen|expert',col))

gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db,gen_dm_df,max_distance = 0.9,with_mitochondrial = T)

# run panelapp analysis
panelapp_panels<-panels_list%>%filter(grepl('panelapp',source))%>%ungroup()%>%select(panel_joined_name)%>%mutate(panel_joined_name=panel_joined_name%>%as.character())

panelapp_generated_panels<-NULL
for (panelapp_panel in panelapp_panels$panel_joined_name){
  message(glue('Generating panel for {panelapp_panel}'))
  gene_rank<-generate_panel_from_panels(joined_panel_names = panelapp_panel,gen_source_db,gen_dm_df,max_distance = 0.9,with_mitochondrial = T)
  if (nrow(gene_rank)==0){next}
  # get the confidence level
  gene_rank<-gene_rank%>%left_join(fixed_source_db%>%filter(panel_joined_name==panelapp_panel)%>%select(gene_symbol,confidence_level))
  panelapp_generated_panels<-panelapp_generated_panels%>%bind_rows(data.frame(panelapp_panel,gene_rank))
}
g<-ggplot(panelapp_generated_panels,aes(y=scaled_adjusted_gene_score,x=cut(adjusted_rank_percentile,10),fill=factor(confidence_level)))+
  geom_boxplot()#+facet_wrap(panelapp_panel~.)
g


toPlot<-panelapp_generated_panels%>%mutate(adjusted_rank_percentile_cat=cut(adjusted_rank_percentile,10),
                                           confidence_level=ifelse(is.na(confidence_level),0,ifelse(confidence_level==0,1,confidence_level)))%>%
  group_by(adjusted_rank_percentile_cat,panelapp_panel)%>%
  summarize(num_o_genes=n(),
            num_o_high_conf=sum(ifelse(confidence_level==3,1,0)),
            num_o_med_conf=sum(ifelse(confidence_level==2,1,0)))%>%
  mutate(rate_o_high_conf=num_o_high_conf/num_o_genes,
         rate_o_med_conf=num_o_med_conf/num_o_genes)


g<-ggplot(toPlot,aes(x=adjusted_rank_percentile_cat,y=rate_o_med_conf))+
  geom_boxplot()
g



calculate_distance_by_thresh<-function(gene_rank,comp_panel,adjusted=F){
  dist_by_thresh<-NULL
  for (thresh in seq(0,1,0.001)){
    if (adjusted){
      genes_above_thresh<-gene_rank%>%filter(adjusted_rank_percentile>=thresh)%>%pull(gene_symbol)
    }else{
      genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
    }
    dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
    dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
  }
  dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)
  return(dist_by_thresh)
}

dist_by_thresh<-calculate_distance_by_thresh(gene_rank,comp_panel,adjusted=T)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g

g<-ggplot(dist_by_thresh,aes(x=thresh,y=sensitivity))+
  geom_point()+
  theme_minimal()
g

comp_panel<-comp_panel%>%left_join(gene_rank)


joined_panel_names<-c("Fulgent Genetics|566600|Comprehensive Glomerular Proteinuria NGS Panel")
comp_panel<-fixed_source_db%>%filter(panel_joined_name=='panelapp|106|Proteinuric renal disease')%>%filter(confidence_level>1)

gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db = gtr_db,gen_dm_df = gtr_dm_df,max_distance = 0.99,with_mitochondrial = T)

comp_panel<-comp_panel%>%left_join(gene_rank)

dist_by_thresh<-NULL
for (thresh in seq(0,1,0.001)){
  genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
  dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
  dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
}
dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g


g<-ggplot(dist_by_thresh,aes(x=thresh,y=sensitivity))+
  geom_point()+
  theme_minimal()
g
```


# Expert panel - panels combination

we will generate panel combinations for the expert panels (panelapp NHS and clingen) including only high condidence genes

```{r prepare_expert_panels}
expert_panels_panelapp<-panels_list%>%
  filter(source=='panelapp' & (high_conf_genes>1))

# get the panelapp NHS approved panels
panelapp_nhs_panels<-fixed_source_db%>%filter(panel_comment=='NHS panel')%>%pull(panel_name)%>%unique()
expert_panels_panelapp<-expert_panels_panelapp%>%filter(panel_name %in% panelapp_nhs_panels)

# now get the clingen panels
expert_panels_clingen<-panels_list%>%
  filter(source=='clingen' & (high_conf_genes>1))

clingen_panelapp_panels<-c(expert_panels_clingen%>%pull(panel_joined_name),
                           expert_panels_panelapp%>%pull(panel_joined_name))

```

```{r find_similar_clingen_panelapp_panels}

# now compare the two panels
expert_panels_dm_df<-dm_df%>%filter(grepl('clingen',row))%>%filter(col %in% clingen_panelapp_panels)
clingen_panels<-as.character(expert_panels_clingen%>%pull(panel_joined_name))
clingen_vs_panelapp<-get_top_similar_panels_table(joined_panel_names = clingen_panels,
                                                  source_db = fixed_source_db,
                                                  dm_df = expert_panels_dm_df,
                                                  number_of_top_panels = 5,
                                                  same_source = F)
write.table(file=glue('{analysis_dir}/clingen_vs_panelapp.csv'),clingen_vs_panelapp,sep='\t',row.names = F)
```

once the corresponding clingen / panelapp panels were found join them and collect only the high confidence genes

```{r generate_expert_panels}
clingen_vs_panelapp<-readr::read_delim('./data/accessory_files/clingen_vs_panelapp_manual_match.csv')
# use the NHS approved panelapp panels as reference and add the clingen high confidence genes 
expert_panels_list<-expert_panels_panelapp%>%mutate(source='expert_panels')
expert_panels_source<-fixed_source_db%>%
  mutate(source='expert_panel')%>%
  filter(panel_joined_name%in%expert_panels_list$panel_joined_name)%>% # take only the NHS panels
  filter(confidence_level==3) # take only the high confidence genes

# now add the matching clingen panel genes
for (clingen_panel_name in clingen_vs_panelapp$selected_panel_name){
  clingen_panel_source<-fixed_source_db%>%filter(panel_joined_name==clingen_panel_name)
  matching_panelapp_panel_joined_name<-clingen_vs_panelapp%>%filter(selected_panel_name==clingen_panel_name)%>%pull(similar_panel)
  matching_panelapp_panel<-expert_panels_source%>%filter(panel_joined_name==matching_panelapp_panel_joined_name)
  # now get the clingen genes that are not in the panelapp  panel
  clingen_genes_to_add<-clingen_panel_source%>%
    filter(confidence_level==3)%>%
    filter(!(gene_symbol %in% (matching_panelapp_panel%>%pull(gene_symbol))))
  message(glue('Will add {nrow(clingen_genes_to_add)} genes from {clingen_panel_name} to {matching_panelapp_panel_joined_name}'))
  expert_panels_source<-expert_panels_source%>%bind_rows(clingen_genes_to_add%>%mutate(source='expert_panels',
                                                                panel_joined_name=matching_panelapp_panel_joined_name,
                                                                panel_id=unique(matching_panelapp_panel%>%pull(panel_id)),
                                                                panel_name=unique(matching_panelapp_panel%>%pull(panel_name))))
}

# now add all the clingen panels without a similar panelapp panel
remaining_clingen_panels<-fixed_source_db%>%filter(source=='clingen')%>%
  filter(!(panel_joined_name%in%clingen_vs_panelapp$selected_panel_name))%>%
  filter(!(panel_name %in% c('General Gene Curation','UNC Biocuration Core','Syndromic Disorders','PTEN')))%>%
  filter(!(panel_name %in% (panels_list%>%filter(source=='clingen' & high_conf_genes==1)%>%pull(panel_name))))

expert_panels_source<-expert_panels_source%>%bind_rows(remaining_clingen_panels)
  
# now fix the source table 
joined_panel_names<-split_joined_panel_name(expert_panels_source$panel_joined_name)
joined_panel_names<-joined_panel_names%>%mutate(source='expert_panel',
                                                panel_joined_name=glue('{source}|{panel_id}|{panel_name}'))

expert_panels_source$source<-joined_panel_names$source
expert_panels_source$panel_joined_name<-joined_panel_names$panel_joined_name

```

```{r check_dist_distribution_for_panelapp_panels}


panelapp_panels_dist<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%group_by(row)%>%
  filter(col!=row)%>%
  summarize(mean_dist=mean(value),
            median_dist=median(value),
            min_dist=min(value))%>%
  left_join(gtr_with_expert_db%>%
              filter(source=='panelapp')%>%
              mutate(panel_joined_name=as.character(panel_joined_name))%>%
              group_by(panel_joined_name)%>%
              summarize(n_genes=n()),
            by=c('row'='panel_joined_name'))

panelapp_panels_dist%>%filter(n_genes<1000)%>%ggplot(aes(x=n_genes,y=min_dist))+
  geom_point()+
  theme_minimal()



```
