---
title: "Panel comparison analysis"
author: "Clalit Genomics"
date: "06/02/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

<style type="text/css">
  body{
  font-size: 14pt;
}
</style>


```{r analysis_setup,include=T,eval=T}
library(glue)
analysis_name<-glue('panel_comparison_analysis_{Sys.Date()}')
analysis_dir<-glue('output/{analysis_name}')
if (!dir.exists(analysis_dir)){dir.create(analysis_dir)}
```

```{r render_report,eval=F,echo=F}
library(rmarkdown)
output_dir <- "/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/output/"
render('/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/src/panel_comparison_analysis.Rmd', output_dir = output_dir,output_file = sprintf('panel_comparison_analysis.%s.html',Sys.Date()),quiet = F)
```

```{r setup, include=T,eval=T}
project_dir<-'/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/'
knitr::opts_knit$set(root.dir='/media/SSD/Bioinformatics/Projects/gene_panel_generator_202108/panel_comparison_project/')
knitr::opts_chunk$set(echo = T,include=F,eval=F,message=F,warning=F)
library(ProjectTemplate)
setwd(project_dir)
load.project()
```


```{r definitions,include=T,eval=T}
# Define the minimal number of genes in a panel - panels with less genes will be excluded from the analysis (under the assumption that they are not panels per-se)
minimal_number_of_genes_in_panel<-3
maximal_number_of_genes_in_panel<-4000
# first fix the manually collected panels
source_db_panels_with_enough_genes<-fixed_source_db%>%count(panel_joined_name)%>%filter(n>=minimal_number_of_genes_in_panel,n<=maximal_number_of_genes_in_panel)%>%pull(panel_joined_name)
fixed_source_db<-fixed_source_db%>%filter(panel_joined_name%in%source_db_panels_with_enough_genes)
# remove panelapp and clingen low conf genes
fixed_source_db<-fixed_source_db%>%filter(!(source%in%c('panelapp','clingen') & confidence_level<2))

gtr_db_panels_with_enough_genes<-gtr_db%>%count(panel_joined_name)%>%filter(n>=minimal_number_of_genes_in_panel)%>%pull(panel_joined_name)
gtr_db<-gtr_db%>%filter(panel_joined_name%in%gtr_db_panels_with_enough_genes)
# filter the tests list to match the filtered gtrdb file
gtr_panels_list<-gtr_panels_list%>%filter(lab_test_id %in% gtr_db$panel_id)

# define the minimal number of publications required to define gene-disease association
min_num_of_pub_to_consider_positive<-2
# define the score to use for the auto generator
ag_score<-'scaled_gene_score'
```

Distance between panels was calculated using the Jaccard method. 

# Create a binary matrix of the manually selected gene panels

```{r gtr_binarize_and_calculate_distance_X,include=T,eval=T}
# binary_panel_gene_table<-create_binary_gene_table(fixed_source_db)
# # filter out the hpo data
# binary_panel_gene_table_no_hpo<-binary_panel_gene_table%>%filter(!grepl('hpo',row.names(binary_panel_gene_table)))
# # generate distance matrix between all the panels
# dm<-generate_distance_matrix(binary_panel_gene_table_no_hpo)
# # convert dm to df
# dm_df<-convert_dm_to_df(dm)
# 
# save(dm,dm_df,file='./data/pre_generated_data/manual_panels_data.Rdata')

# load dm data
load('./data/pre_generated_data/manual_panels_data.Rdata')

```

# Create a binary matrix of the GTR gene panels

```{r gtr_binarize_and_calculate_distance,include=T,eval=T}

# gtr_with_expert_db<-gtr_db%>%mutate(panel_id=as.character(panel_id))%>%
#   bind_rows(fixed_source_db%>%filter(source%in%c('panelapp','clingen')))#%>%
#   #bind_rows(expert_panels_source)
# 
# 
# gtr_binary_panel_gene_table<-create_binary_gene_table(gtr_with_expert_db)
# # generate distance matrix between all the panels
# gtr_dm<-generate_distance_matrix(gtr_binary_panel_gene_table)
# # convert dm to df
# gtr_dm_df<-convert_dm_to_df(gtr_dm)
# 
# # save gtr matrix and dataframe
# save(gtr_dm,gtr_binary_panel_gene_table,gtr_with_expert_db,gtr_dm_df,file='./data/pre_generated_data/gtr_data.Rdata')

# load gtr data
load('./data/pre_generated_data/gtr_data.Rdata')
```

# Source statistics

```{r panel_stats,include=T,eval=T}

#source_stats<-get_source_stats(gtr_db)

# Get the years of the different tests
## year created
gtr_panels_list%>%count(year(test_create_date))
## year updated
gtr_panels_list%>%count(year(last_touch_date))
## Country of origin
gtr_panels_list%>%count(facility_country)%>%arrange(desc(n))

# Total panels
total_panels<-nrow(gtr_panels_list)
# Number of countries
ncountries<-gtr_panels_list%>%pull(facility_country)%>%unique()%>%length()
# Number of panels updated after 2019
npanels_after_2019<-gtr_panels_list%>%filter(year(last_touch_date)>=2019)%>%nrow()

# number of genes per panel
number_of_genes_per_panel<-gtr_db%>%group_by(panel_joined_name)%>%summarize(n=n())
number_of_genes_per_panel_summary<-skimr::skim_without_charts(number_of_genes_per_panel$n)

unique_sources<-gtr_db%>%pull(source)%>%unique()
# number of unique sources
n_unique_sources<-length(unique_sources)
unique_panels<-gtr_db%>%pull(panel_joined_name)%>%unique()
n_unique_panels<-length(unique_panels)
unique_panelapp_panels<-fixed_source_db%>%filter(source=='panelapp')%>%pull(panel_joined_name)%>%unique()

```

# Define Selected panels

```{r selected_panels,include=T,eval=T}
selected_phenotypes_panelapp<-c("panelapp|49|Hypertrophic cardiomyopathy - teen and adult",
                                'panelapp|544|Cholestasis',
                                'panelapp|149|Nephrocalcinosis or nephrolithiasis',
                                'panelapp|179|Hydrocephalus',
                                'panelapp|657|Autism',
                                'panelapp|126|Hearing loss',
                                'panelapp|294|Ocular coloboma',
                                #'panelapp|309|Skeletal dysplasia',
                                'panelapp|386|Pancreatitis',
                                'panelapp|285|Intellectual disability',
                                'panelapp|943|Osteopetrosis',
                                'panelapp|553|Ectodermal dysplasia',
                                # #'panelapp|487|Cystic renal disease',
                                'panelapp|258|Arthrogryposis',
                                'panelapp|78|Holoprosencephaly',
                                # #'panelapp|522|Familial melanoma', 
                                'panelapp|652|Dilated cardiomyopathy - adult and teen',
                                'panelapp|658|Corneal dystrophies',
                                'panelapp|555|Ichthyosis and erythrokeratoderma',
                                'panelapp|546|Lipodystrophy - childhood onset',
                                'panelapp|168|Craniosynostosis',
                                'panelapp|842|Cardiac arrhythmias',
                                'panelapp|65|Primary lymphoedema',
                                'panelapp|92|Hypogonadotropic hypogonadism',
                                'panelapp|1|Thoracic aortic aneurysm or dissection',
                                'panelapp|526|Neuronal ceroid lipofuscinosis'
                                #'panelapp|551|Surfactant deficiency'
                                )
selected_phenotypes<-c('hypertrophic cardiomyopathy',
                       'cholestasis',
                       'nephrolithiasis',
                       'hydrocephalus',
                       'autism',
                       'hearing loss',
                       'coloboma',
                       #'skeletal dysplasia',
                       'pancreatitis',
                       'intellectual disability',
                       'osteopetrosis',
                       'ectodermal dysplasia',
                       #'polycystic kidney',
                       'arthrogryposis',
                       'holoprosencephaly',
                       #'melanoma',
                       'dilated cardiomyopathy',
                       'corneal dystrophy',
                       'ichthyosis',
                       'lipodystrophy',
                       'craniosynostosis',
                       'arrhythmia',
                       'lymphedema',
                       'hypogonadotropic hypogonadism',
                       'aortic aneurysm',
                       'neuronal ceroid lipofuscinosis'
                       #'surfactant deficiency'
                       )

phenotype_queries<-c('hypertrophic.*cardiomyopathy|hcm|cardiomyopathy.*hypertrophic|ICD10:I421|ICD10:I422',
                     'cholestasis|ICD10:K831',
                     'nephrolithiasis|monogenic kidney stone|ICD10:N200',
                     'hydrocephalus|ICD10:G91',
                     'autism|autistic|ICD10:F840',
                     'hearing loss|deafness|ICD10:H93',
                     'coloboma|ICD10:Q130',
                     #'skeletal.* dysplasia|dysplasia.*skeletal|osteo.*dysplas|dysplas*.osteo|ICD10:Q78',
                     'pancreatitis|ICD10:K85|ICD10:K861',
                     'intellectual.*disability|mental.*retard|cognitive.*impair|ICD10:F7[012389]',
                     'osteopetrosis|ICD10:782',
                     'ectodermal.*dysplasia|ectoderm|dysplasia.*ectoderm|ICD10:Q824',
                     #'cystic.*renal|polycys.*kidn|renal.*cyst|renal*.ciliop|ciliop.*renal',
                     'arthrogryposis|ICD10:Q743',
                     'holoprosencephaly|ICD10:Q042',
                     #'melanoma',
                     'dilated.*cardio|dcm|cardiomyo.*dilated|ICD10:I420',
                     'cornea.*dyst|dystr.*cornea|ICD10:H185',
                     'ichthyosis|ICD10:Q80',
                     'lipodystrophy|ICD10:E881',
                     'craniosynostosis|ICD10:Q750',
                     'arrhythmia|ICD10:I49|qt syndrome|brugada|fibrillation',
                     'lymphedema|ICD10:I890',
                     'hypogonadism',
                     'aort.*aneurysm|aneurysm.*aort|ICD10:I71',
                     'lipofuscinosis|ICD10:E754'
                     #'surfactant'
                     )


selected_phenotypes_df<-data.frame(phenotype_name=selected_phenotypes,
                                   panelapp_panel=selected_phenotypes_panelapp,
                                   query=phenotype_queries)


selected_phenotypes_df<-selected_phenotypes_df
```

# Analyze the selected panels 

```{r run_comparison_for_selected,include=T,eval=T}
all_panel_names<-gtr_dm_df%>%pull(row)%>%unique()
panelapp_panels<-grep('panelapp',all_panel_names,ignore.case = T,value = T)
all_genes<-gtr_with_expert_db%>%pull(gene_symbol)%>%unique()
max_dists_for_auto<-c(0.7,0.75,0.8,0.85,0.9)
# run once to pre-generate the per gene publications data
#pubmed_grab_all_gene_publications(all_genes)
# load the pregenerated all gene publications
load('./data/pre_generated_data/genes_pubmed_search.RData')

selected_phenotypes_analysis<-list()
for (selected_phenotype_name in selected_phenotypes){
  message(glue('Running comparison for {selected_phenotype_name}..'))
  panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
  naive_search<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(query)

  selected_phenotypes_analysis[[selected_phenotype_name]] <-
    compare_panelapp_naive_and_cluster(
      panelapp_panel = panelapp_panel,
      phenotype_name = selected_phenotype_name,
      naive_search =  naive_search,
      max_dists_for_cluster = NA,
      gtr_with_expert_db = gtr_with_expert_db,
      gtr_dm_df = gtr_dm_df,
      all_genes = all_genes,
      all_panel_names = all_panel_names,
      pre_generated_gene_publications_list = entrez_pubmed_res,
      min_num_of_pub_to_consider_positive = min_num_of_pub_to_consider_positive)
}

# collect all text
all_text<-c()
all_text_df<-NULL
for (selected_phenotype_name in selected_phenotypes){
  phenotype_text<-selected_phenotypes_analysis[[selected_phenotype_name]]$relatedness_summary$summary_text$relatedness_text
  phenotype_text_row<-selected_phenotypes_analysis[[selected_phenotype_name]]$relatedness_summary$summary_text$relatedness_table
  all_text<-c(all_text,phenotype_text)
  all_text_df<-all_text_df%>%bind_rows(phenotype_text_row)
}




write.table(data.frame(selected_phenotypes,discrepancy_text=all_text),file=glue('{analysis_dir}/discrepancy_text.tsv'),sep='\t',row.names = F,quote = F)
all_text<-all_text
all_text_df<-all_text_df

# HCM discrepancy
phenotype_name<-'hypertrophic cardiomyopathy'
all_genes_vs_phenotype<-
  pubmed_gene_list_vs_phenotype(
    phenotype_name = phenotype_name,
    gene_list = all_genes,
    pre_generated_genes_pubs = pre_generated_gene_publications_list)
  
panels_discrepancy_analysis(phenotype_name = phenotype_name,
                            phenotype_panels = selected_phenotypes_analysis[[phenotype_name]][['naive_panels']],
                            gtr_with_expert_db = gtr_with_expert_db,
                            panel_to_compare = selected_phenotypes_analysis[[phenotype_name]][['panelapp_panel']],
                            gtr_dm_df = gtr_dm_df,
                            all_genes_vs_pubmed = all_genes_vs_phenotype,
                            min_num_of_pub_to_consider_positive=min_num_of_pub_to_consider_positive)

# Autism discrepancy

phenotype_name<-'autism'
all_genes_vs_phenotype<-
  pubmed_gene_list_vs_phenotype(
    phenotype_name = phenotype_name,
    gene_list = all_genes,
    pre_generated_genes_pubs = pre_generated_gene_publications_list)
panels_discrepancy_analysis(phenotype_name = phenotype_name,
                            phenotype_panels = selected_phenotypes_analysis[[phenotype_name]][['naive_panels']],
                            gtr_with_expert_db = gtr_with_expert_db,
                            panel_to_compare = selected_phenotypes_analysis[[phenotype_name]][['panelapp_panel']],
                            gtr_dm_df = gtr_dm_df,
                            all_genes_vs_pubmed = all_genes_vs_phenotype,
                            min_num_of_pub_to_consider_positive=min_num_of_pub_to_consider_positive)

#save(selected_phenotypes_analysis,file='./data/pre_generated_data/selected_phenotypes_analysis.RData')
#load('./data/pre_generated_data/selected_phenotypes_analysis.RData')
# now check the ranks for each 

```

# Sanity check- review panelapp genes not found in search

```{r sanity_check}

panelapp_genes_not_in_search<-NULL
literature_genes_not_in_panelapp<-NULL
for (selected_phenotype in selected_phenotypes){
  panelapp_genes<-selected_phenotypes_analysis[[selected_phenotype]]$panelapp_genes
  all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype,
                                                        gene_list = all_genes,
                                                        pre_generated_genes_pubs = pre_generated_gene_publications_list)
  
  panelapp_genes_without_publications<-all_genes_vs_phenotype%>%filter(pub_num<2)%>%filter(gene_symbol%in%panelapp_genes)
  lit_genes_not_in_panelapp<-all_genes_vs_phenotype%>%filter(pub_num>1)%>%filter(!gene_symbol%in%panelapp_genes)
  panelapp_genes_not_in_search[[selected_phenotype]]<-panelapp_genes_without_publications
  literature_genes_not_in_panelapp[[selected_phenotype]]<-lit_genes_not_in_panelapp
}

panelapp_genes_not_in_search$`hydrocephalus`

phenotype_search_res<- entrez_search(db="pubmed", term='cholestasis',retmax=999999)
phenotype_search_res_mesh<- entrez_search(db="pubmed", term='hypertrophic cardiomyopathy[mesh]',retmax=999999)


entrez_to_hgnc<-hgnc%>%dplyr::select(symbol,entrez_id)
names_to_hgnc<-data.frame(entrez_id=names(entrez_pubmed_res))%>%left_join(entrez_to_hgnc%>%mutate(entrez_id=as.character(entrez_id)))
'33414089' %in% phenotype_search_res$ids
'33414089' %in% entrez_pubmed_res[[entrez_to_hgnc%>%filter(symbol=='CYP27A1')%>%pull(entrez_id)]]
```

# Diseases db
The files contain all links in the DISEASES database. All files start with the following four columns: gene identifier, gene name, disease identifier, and disease name. The knowledge files further contain the source database, the evidence type, and the confidence score. The experiments files instead contain the source database, the source score, and the confidence score. Finally, the textmining files contain the z-score, the confidence score, and a URL to a viewer of the underlying abstracts.

```{r diseases_db,include=T,eval=T}
#diseases_knowledge_db<-readr::read_delim('/media/SSD/Bioinformatics/Databases/diseases/human_disease_knowledge_full.tsv',col_names = c('gene_id','gene_symbol','disease_id','disease_name','source','evidence_type','confidence_score'))

diseases_textmining_db<-readr::read_delim('/media/SSD/Bioinformatics/Databases/diseases/human_disease_textmining_full.tsv',col_names = c('gene_id','gene_symbol','disease_id','disease_name','z_score', 'confidence_score','url'))

diseases_knowledge_db<-
  readr::read_delim('/media/SSD/Bioinformatics/Databases/diseases/human_disease_knowledge_full.tsv',                             col_names = c('gene_id','gene_symbol','disease_id','disease_name','source','type','confidence_level'))


# join the textmining and the knowledge db
disease_count<-diseases_textmining_db%>%count(disease_name)
diseases_db<-diseases_textmining_db%>%
  filter(!disease_name%in%(disease_count%>%filter(n>10000)%>%pull(disease_name)))%>%
  select(gene_symbol,disease_name,confidence_score,url)%>%
  bind_rows(diseases_knowledge_db%>%
              #filter(source%in%c('MedlinePlus','UniProtKB-KW'))%>%
              mutate(confidence_score=confidence_level,
                     url=source)%>%
              select(gene_symbol,disease_name,confidence_score,url))
  #filter(!grepl('ICD10:',disease_name))


all_diseases<-diseases_db%>%pull(disease_name)%>%unique()

ddb_vs_phenotype<-list()
for (selected_phenotype in selected_phenotypes){
  message(glue('Retrieving {selected_phenotype} genes from diseases_db'))
  phenotype_query<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype)%>%pull(query)
  ddb_pheno_diseases<-grep(phenotype_query,diseases_db$disease_name,ignore.case = T,value=T)%>%unique()
  
  ddb_genes<-diseases_db%>%
    filter(disease_name %in% ddb_pheno_diseases)%>%
    group_by(gene_symbol)%>%
    slice_max(n=1,order_by=confidence_score,with_ties = F)
  message(glue('found {nrow(ddb_genes)} genes with a confidence score associated with {selected_phenotype}'))
  ddb_vs_phenotype[[selected_phenotype]]<-ddb_genes
}

# for each phenotype, check if there is a confidence score cutoff that includes all the panelapp panels
pa_vs_ddb<-NULL
all_lit<-NULL
all_ddb<-NULL
pa_genes<-NULL
for (selected_phenotype in selected_phenotypes){
  panelapp_genes<-selected_phenotypes_analysis[[selected_phenotype]]$panelapp_genes
  pa_genes<-pa_genes%>%bind_rows(data.frame(phenotype_name=selected_phenotype,
                                            gene_symbol=panelapp_genes))
  all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype,
                                                        gene_list = all_genes,
                                                        pre_generated_genes_pubs = pre_generated_gene_publications_list)
  
  
  # panelapp vs diseases vs literature
  pa_vs_ddb<-pa_vs_ddb%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype,gene_symbol=panelapp_genes)%>%
                left_join(ddb_vs_phenotype[[selected_phenotype]])%>%
                left_join(all_genes_vs_phenotype)
      )
  
  # literature strat
  all_lit<-all_lit%>%bind_rows(all_genes_vs_phenotype%>%filter(pub_num>0)%>%
                                 mutate(phenotype_name=selected_phenotype,
                                        is_panelapp=ifelse(gene_symbol%in%panelapp_genes,1,0)))
  
  # literature strat
  all_ddb<-all_ddb%>%bind_rows(ddb_vs_phenotype[[selected_phenotype]]%>%
                                 mutate(phenotype_name=selected_phenotype,
                                        is_panelapp=ifelse(gene_symbol%in%panelapp_genes,1,0)))

}

# check literature search
lit_vs_pa<-all_lit%>%
  left_join(pa_genes%>%group_by(phenotype_name)%>%summarize(n_panelapp_genes=n()))%>%
  group_by(phenotype_name,n_panelapp_genes)%>%
  summarize(n=n(),
            above_0_ppv=sum(ifelse(pub_num>=1 & is_panelapp==1,1,0))/sum(ifelse(pub_num>=1,1,0)),
            above_1_ppv=sum(ifelse(pub_num>=2 & is_panelapp==1,1,0))/sum(ifelse(pub_num>=2,1,0)),
            above_2_ppv=sum(ifelse(pub_num>=3 & is_panelapp==1,1,0))/sum(ifelse(pub_num>=3,1,0)),
            above_0_sens=sum(ifelse(pub_num>=1 & is_panelapp==1,1,0)),
            above_1_sens=sum(ifelse(pub_num>=2 & is_panelapp==1,1,0)),
            above_2_sens=sum(ifelse(pub_num>=3 & is_panelapp==1,1,0)))%>%
  mutate(above_0_sens=above_0_sens/n_panelapp_genes,
         above_1_sens=above_1_sens/n_panelapp_genes,
         above_2_sens=above_2_sens/n_panelapp_genes)
  
# ddb
ddv_vs_pa<-all_ddb%>%
  left_join(pa_genes%>%group_by(phenotype_name)%>%summarize(n_panelapp_genes=n()))%>%
  group_by(phenotype_name,n_panelapp_genes)%>%
  summarize(n=n(),
            n_above_1=sum(ifelse(confidence_score>1,1,0)),
            n_above_2=sum(ifelse(confidence_score>2,1,0)),
            n_above_3=sum(ifelse(confidence_score>3,1,0)),
            above_0_ppv=sum(ifelse(confidence_score>0 & is_panelapp==1,1,0))/sum(ifelse(confidence_score>0,1,0)),
            above_1_ppv=sum(ifelse(confidence_score>=1 & is_panelapp==1,1,0))/sum(ifelse(confidence_score>=1,1,0)),
            above_2_ppv=sum(ifelse(confidence_score>=2 & is_panelapp==1,1,0))/sum(ifelse(confidence_score>=2,1,0)),
            above_3_ppv=sum(ifelse(confidence_score>=3 & is_panelapp==1,1,0))/sum(ifelse(confidence_score>=3,1,0)),
            above_0_sens=sum(ifelse(confidence_score>0 & is_panelapp==1,1,0)),
            above_1_sens=sum(ifelse(confidence_score>=1 & is_panelapp==1,1,0)),
            above_2_sens=sum(ifelse(confidence_score>=2 & is_panelapp==1,1,0)),
            above_3_sens=sum(ifelse(confidence_score>=3 & is_panelapp==1,1,0)))%>%
  mutate(above_0_sens=above_0_sens/n_panelapp_genes,
         above_1_sens=above_1_sens/n_panelapp_genes,
         above_2_sens=above_2_sens/n_panelapp_genes,
         above_3_sens=above_3_sens/n_panelapp_genes)

# check for arthogryposis missing genes
ag_pa_genes<-pa_genes%>%filter(phenotype_name=='ectodermal dysplasia')
ag_pa_genes_in_diseases<-diseases_db%>%filter(gene_symbol%in%ag_pa_genes$gene_symbol)%>%
  filter(grepl('ecto',disease_name,ignore.case = T))

genes_not_in_ddb<-ag_pa_genes%>%filter(!(gene_symbol%in%ag_pa_genes_in_diseases$gene_symbol))
their_relations<-diseases_db%>%filter(gene_symbol%in%genes_not_in_ddb$gene_symbol)
diseases_db%>%filter(gene_symbol=='APCDD1')

```

# Generate auto panels

```{r generate auto panels,include=T,eval=T}
auto_generator_panelapp_only<-list()
auto_generator_panelapp_with_similar<-list()
max_dists=c(0.85,0.95,0.99)
for (max_dist in max_dists){
  auto_generator_panelapp_only[[glue('max_dist={max_dist}')]]<-list()
  for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
    message(glue('Generating panel for {selected_phenotype_name} max dist = {max_dist}'))
    panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
    # similar_panels<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,
    #                                                       selected_panel_name = panelapp_panel,
    #                                                       top_most_similar = 1,maximal_distance = 0.7,
    #                                                       names_only = T)
    auto_generator_panelapp_only[[glue('max_dist={max_dist}')]][[selected_phenotype_name]] <-
      generate_panel_from_panels(
        joined_panel_names = panelapp_panel,
        gen_source_db = gtr_with_expert_db,
        gen_dm_df = gtr_dm_df,
        max_distance = max_dist,
        with_mitochondrial = F)
    # auto_generator_panelapp_with_similar[[glue('max_dist={max_dist}')]][[selected_phenotype_name]] <-
    #   generate_panel_from_panels(
    #     joined_panel_names = similar_panels,
    #     gen_source_db = gtr_with_expert_db,
    #     gen_dm_df = gtr_dm_df,
    #     max_distance = max_dist,
    #     with_mitochondrial = F)
  }
}

# check the distribution of scores
autopanel_scores<-NULL
for (selected_phenotype in selected_phenotypes){
  all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype,
                                                          gene_list = all_genes,
                                                          pre_generated_genes_pubs = entrez_pubmed_res)
  for (max_dist in max_dists){
    all_genes_vs_phenotype<-all_genes_vs_phenotype%>%
      mutate(has_pub=factor(ifelse(pub_num>=min_num_of_pub_to_consider_positive,1,0)))
    pheno_scores<-all_genes_vs_phenotype%>%select(-pubmed_ids)%>%
      left_join(data.frame(max_dist=max_dist,phenotype_name=selected_phenotype,auto_generator_panelapp_only[[max_dist]][[selected_phenotype]]))
    autopanel_scores<-autopanel_scores%>%bind_rows(pheno_scores)
  }
}

autopanel_scores%>%
  filter(max_dist==0.85)%>%
  select(phenotype_name,has_pub,num_o_panels,scaled_gene_score)%>%
  pivot_longer(-c(phenotype_name,has_pub))%>%
  ggplot(aes(x=value,fill=has_pub))+
    geom_histogram(position='dodge')+
  facet_grid(phenotype_name~name,scales='free')+
  theme_minimal()

```

# ROC analysis - vs pubmed

```{r roc_analysis_pubmed,include=T,eval=T}

# now add whether the gene is positive or not
auto_generator_rocs_pubmed<-NULL
panelapp_roc_pubmed<-NULL
naive_roc_pubmed<-NULL

for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
  panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
  all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype_name,
                                                            gene_list = all_genes,
                                                            pre_generated_genes_pubs = entrez_pubmed_res)
  all_genes_vs_phenotype<-all_genes_vs_phenotype%>%
    mutate(has_pub=factor(ifelse(pub_num>=min_num_of_pub_to_consider_positive,1,0)))
  
  
  # autogenerator rocs
  for (max_dist in max_dists){
    auto_generator_panel_res<-auto_generator_panelapp_only[[glue('max_dist={max_dist}')]][[selected_phenotype_name]]
    phenotype_auto_generator_pr_curve <-
      calculate_auto_generator_pr_for_phenotype(phenotype_name = selected_phenotype_name,
                                                auto_generator_panel_res = auto_generator_panel_res,
                                                all_genes_vs_phenotype = all_genes_vs_phenotype,
                                                gen_source_db = gtr_with_expert_db,
                                                score_column = 'scaled_gene_score')
    auto_generator_rocs_pubmed<-auto_generator_rocs_pubmed%>%
      bind_rows(data.frame(phenotype_auto_generator_pr_curve,max_dist=max_dist)%>%
                  mutate(group=glue('ag_{max_dist}')))
    
    # phenotype_auto_generator_pr_curve <-
    #   calculate_auto_generator_pr_for_phenotype(phenotype_name = selected_phenotype_name,
    #                                             auto_generator_panel_res = auto_generator_panel_res,
    #                                             all_genes_vs_phenotype = all_genes_vs_phenotype,
    #                                             gen_source_db = gtr_with_expert_db,
    #                                             score_column = 'scaled_gene_score_exp')
    # auto_generator_rocs<-auto_generator_rocs%>%
    #   bind_rows(data.frame(phenotype_auto_generator_pr_curve,max_dist=max_dist)%>%
    #               mutate(group=glue('ag_{max_dist}_modified')))
    
  }
  
  # calculate the panelapp recall and precision
  panelapp_genes<-gtr_with_expert_db%>%filter(panel_joined_name==panelapp_panel)%>%pull(gene_symbol)
  panelapp_phenotype_pr<-calculate_sens_and_ppv_vs_pubmed(panelapp_genes,
                                         all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol),
                                         precision_recall_naming = T)
  
  panelapp_roc_pubmed<-panelapp_roc_pubmed%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         group='panelapp',
                         recall=panelapp_phenotype_pr$recall,
                         precision=panelapp_phenotype_pr$precision,
                         f1=panelapp_phenotype_pr$f1))
  
  # calculate the naive search (with panelapp) recall and precision
  naive_genes<-selected_phenotypes_analysis[[selected_phenotype_name]]$naive_with_panelapp_genes
  naive_phenotype_pr<-calculate_sens_and_ppv_vs_pubmed(naive_genes,
                                         all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol),
                                         precision_recall_naming = T)
  naive_roc_pubmed<-naive_roc_pubmed%>%
    bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                         group='naive search',
                         recall=naive_phenotype_pr$recall,
                         precision=naive_phenotype_pr$precision,
                         f1=naive_phenotype_pr$f1))
  
}

all_aucs_pubmed<-auto_generator_rocs_pubmed%>%group_by(phenotype_name,group)%>%distinct(pr_auc)%>%
  ungroup()%>%group_by(group)%>%
  summarize(pr_auc=mean(pr_auc))

all_rocs_pubmed<-auto_generator_rocs_pubmed%>%
  bind_rows(panelapp_roc_pubmed)%>%
  bind_rows(naive_roc_pubmed)

# get the max f1 value per max dist per phenotype 
all_rocs_pubmed%>%
  group_by(phenotype_name,group)%>%
  slice_max(n=1,order_by = f1,with_ties = F)%>%
  ungroup()%>%
  select(group,f1,precision,recall)%>%
  pivot_longer(-group)%>%
  ggplot(aes(x=group,y=value,fill=group,col=group))+
  geom_boxplot(col='gray',alpha=0.2)+#coord_flip()+
  geom_jitter(width = 0.1)+
  #coord_flip()+
  facet_wrap(name~.)+
  guides(fill=F)+
  ggsci::scale_fill_nejm()+ggsci::scale_color_nejm()+
  theme_minimal()+
  theme(legend.position = 'top')

# compare rocs auto generator
all_rocs_pubmed%>%
  filter(grepl('ag_0.95',group))%>%
  filter(!((.threshold==Inf) | (.threshold==0))) %>%
  ggplot(aes(x=recall,y=precision,col=factor(group)))+
  geom_line()+facet_wrap(phenotype_name~.)+
  geom_abline(slope = -1,intercept = 1,linetype=2,alpha=0.5)+
  geom_point(data=panelapp_roc_pubmed,inherit.aes = F,aes(x=recall,y=precision))+
  geom_point(data=naive_roc_pubmed,inherit.aes = F,aes(x=recall,y=precision),col='red')+
  #geom_point(data=comparison_metrics%>%filter(group=='distance_max_f1'),inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  #geom_point(data=dist_roc,inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  theme_minimal()+
  scale_color_nejm()+
  theme(legend.position = 'top')


```

# ROC analysis - vs textmining

```{r roc_analysis,include=T,eval=T}
all_ddb_genes<-diseases_db%>%pull(gene_symbol)%>%unique()
# now add whether the gene is positive or not
#confidence_score_thresh<-2
confidence_score_threshs<-c(1,2,3,4)
all_rocs_ddb<-NULL

for (confidence_score_thresh in confidence_score_threshs){
  auto_generator_rocs_ddb<-NULL
  panelapp_roc_ddb<-NULL
  naive_roc_ddb<-NULL
  for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
    panelapp_panel<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype_name)%>%pull(panelapp_panel)
    all_genes_vs_phenotype<-data.frame(gene_symbol=all_genes)%>%
      left_join(ddb_vs_phenotype[[selected_phenotype_name]])%>%
      mutate(confidence_score=ifelse(is.na(confidence_score),0,confidence_score),
             has_pub=factor(ifelse(confidence_score>=confidence_score_thresh,1,0),levels=c(0,1)))
    
  
    # autogenerator rocs
    for (max_dist in max_dists){
      message(glue('Analyzing {selected_phenotype_name}: {confidence_score_thresh}: {max_dist}'))
      auto_generator_panel_res<-auto_generator_panelapp_only[[glue('max_dist={max_dist}')]][[selected_phenotype_name]]
      phenotype_auto_generator_pr_curve <-
        calculate_auto_generator_pr_for_phenotype(phenotype_name = selected_phenotype_name,
                                                  auto_generator_panel_res = auto_generator_panel_res,
                                                  all_genes_vs_phenotype = all_genes_vs_phenotype,
                                                  gen_source_db = gtr_with_expert_db,
                                                  score_column = 'scaled_gene_score')
      auto_generator_rocs_ddb<-auto_generator_rocs_ddb%>%
        bind_rows(data.frame(phenotype_auto_generator_pr_curve,max_dist=max_dist)%>%
                    mutate(group=glue('ag_{max_dist}')))# MIND THAT IF THE AUTO GENERATOR ROC IS FOR MORE THAN ONE SCORE (INCLUDES THE MODIFIED) YOU NEED TO GIVE EACH SCORE A DIFFERENT GROUP NAME
      
    }
    
    # calculate the panelapp recall and precision
    panelapp_genes<-gtr_with_expert_db%>%filter(panel_joined_name==panelapp_panel)%>%pull(gene_symbol)
    panelapp_phenotype_pr<-calculate_sens_and_ppv_vs_pubmed(panelapp_genes,
                                           all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol),
                                           precision_recall_naming = T)
    
    panelapp_roc_ddb<-panelapp_roc_ddb%>%
      bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                           group='panelapp',
                           recall=panelapp_phenotype_pr$recall,
                           precision=panelapp_phenotype_pr$precision,
                           f1=panelapp_phenotype_pr$f1))
    
    # calculate the naive search (with panelapp) recall and precision
    naive_genes<-selected_phenotypes_analysis[[selected_phenotype_name]]$naive_with_panelapp_genes
    naive_phenotype_pr<-calculate_sens_and_ppv_vs_pubmed(naive_genes,
                                           all_genes_vs_phenotype%>%filter(has_pub==1)%>%pull(gene_symbol),
                                           precision_recall_naming = T)
    naive_roc_ddb<-naive_roc_ddb%>%
      bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                           group='naive search',
                           recall=naive_phenotype_pr$recall,
                           precision=naive_phenotype_pr$precision,
                           f1=naive_phenotype_pr$f1))
    
    
  }
  all_rocs_ddb<-all_rocs_ddb%>%bind_rows(
    auto_generator_rocs_ddb%>%
        bind_rows(panelapp_roc_ddb)%>%
        bind_rows(naive_roc_ddb)%>%
    mutate(confidence_score_thresh=confidence_score_thresh)
  )
}



library(gghalves)
# get the max f1 value per max dist per phenotype 
all_rocs_ddb%>%
  #filter(confidence_score_thresh==2)%>%
  group_by(phenotype_name,group,confidence_score_thresh)%>%
  slice_max(n=1,order_by = f1,with_ties = F)%>%
  ungroup()%>%
  select(group,confidence_score_thresh,f1,precision,recall)%>%
  pivot_longer(-c(group,confidence_score_thresh))%>%
  ggplot(aes(x=group,y=value,fill=group,col=group))+
  #geom_boxplot(alpha=0.7)+#coord_flip()+
  geom_half_boxplot(alpha=0.7,col='black') +
  geom_half_point(alpha=0.5)+
  facet_grid(name~confidence_score_thresh)+
  guides(fill=F)+coord_flip()+
  ggsci::scale_fill_nejm()+ggsci::scale_color_nejm()+
  theme_minimal()

# compare rocs auto generator 
to_plot<-all_rocs_ddb%>%
  filter(confidence_score_thresh==2)
  

to_plot%>%
  filter(group%in%c('ag_0.85'))%>%
  filter(!((.threshold==Inf) | (.threshold==0))) %>%
  ggplot(aes(x=recall,y=precision,col=factor(group)))+
  geom_line()+facet_wrap(phenotype_name~.)+
  geom_abline(slope = -1,intercept = 1,linetype=2,alpha=0.5)+
  geom_point(data=to_plot%>%filter(group%in%c('panelapp')),inherit.aes = F,aes(x=recall,y=precision))+
  geom_point(data=to_plot%>%filter(group%in%c('naive search')),inherit.aes = F,aes(x=recall,y=precision),col='red')+
  # geom_point(data=comparison_metrics%>%filter(group=='distance_max_f1'),inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  #geom_point(data=dist_roc,inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  theme_minimal()+
  scale_color_nejm()+
  theme(legend.position = 'top',legend.text = element_text(size=12))


```

# Categorizing gene-score

```{r score_cat}
ag_panels<-auto_generator_panelapp_only$`max_dist=0.95`
# flatten the panels
ag_panels<-ag_panels%>%listr::list_name_to_df()%>%listr::list_bind_all()
# join with has pub
all_genes_has_pub<-NULL
for (selected_phenotype in selected_phenotypes){
  all_genes_vs_phenotype<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype,
                                                            gene_list = all_genes,
                                                            pre_generated_genes_pubs = entrez_pubmed_res)
  all_genes_vs_phenotype<-all_genes_vs_phenotype%>%
    mutate(has_pub=factor(ifelse(pub_num>=min_num_of_pub_to_consider_positive,1,0)))
  all_genes_has_pub<-all_genes_has_pub%>%
    bind_rows(data.frame(all_genes_vs_phenotype%>%select(-pubmed_ids),
                         .group=selected_phenotype))
}

# now join them
genes_with_score_and_pub<-all_genes_has_pub%>%
  left_join(ag_panels%>%select(gene_symbol,.group,scaled_gene_score,rank_percentile))%>%
  mutate(scaled_gene_score=ifelse(is.na(scaled_gene_score),0,scaled_gene_score),
         rank_percentile=ifelse(is.na(rank_percentile),0,rank_percentile))

genes_with_score_and_pub%>%
  filter(rank_percentile!=0)%>%
  ggplot(aes(x=rank_percentile,fill=has_pub))+
  geom_histogram(position='dodge')+facet_wrap(.group~.,scales='free_y')

```

# Statistical analysis

```{r ag_vs_pa_vs_ns,include=T,eval=T}


statistical_tests_data<-list()
for (reference_source in c('pubmed')){# cannot use this part in the ddb comparison, because the results are divided per confidence thresh, need to create another list just for it
  if (reference_source=='pubmed'){
    auto_generator_rocs<-auto_generator_rocs_pubmed
    panelapp_roc<-panelapp_roc_pubmed
    naive_roc<-naive_roc_pubmed
  }
  statistical_tests_data[[reference_source]]<-list()
  for (selected_dist in max_dists){
    statistical_tests_data[[reference_source]][[as.character(selected_dist)]]<-list()
    message(glue('Running stats test for {reference_source} {selected_dist}'))
    selected_ag_roc<-auto_generator_rocs%>%filter(!(!is.na(max_dist)&group!=glue('ag_{selected_dist}')))
    
    # add number of genes
    library(listr)
    auto_panels_genes<-auto_generator_panelapp_only[[glue('max_dist={selected_dist}')]] %>%list_name_to_df()%>%list_bind(selected_phenotypes)
    auto_panels_genes<-auto_panels_genes[[1]]%>%select(phenotype_name=.group,gene_symbol,scaled_gene_score)
    
    # for each threshold calculate the number of genes
    number_of_genes_above_thresh<-list()
    # for (selected_phenotype in selected_phenotypes){
    #   message(glue('Adding number of genes per threshold for {selected_phenotype}'))
    #   number_of_genes_above_thresh[[selected_phenotype]]<-NULL
    #   phenotype_auto_panels<-auto_panels_genes%>%filter(phenotype_name==selected_phenotype)
    #   threshs <- selected_ag_roc%>%filter(phenotype_name==selected_phenotype) %>% pull(.threshold)
    #   for (thresh in threshs){
    #     number_of_genes_above_thresh[[selected_phenotype]]<-
    #       number_of_genes_above_thresh[[selected_phenotype]]%>%
    #       bind_rows(data.frame(.threshold=thresh,
    #                            ngenes=phenotype_auto_panels%>%filter(scaled_gene_score>=thresh)%>%nrow()))
    #     
    #   }
    # }
    # 
    # # now flatten the list using listr
    # ngenes_to_add<-number_of_genes_above_thresh%>%list_name_to_df()%>%list_bind(selected_phenotypes)
    # ngenes_to_add<-ngenes_to_add[[1]]%>%rename(phenotype_name=.group)
    # 
    # selected_ag_roc<-selected_ag_roc%>%
    #   left_join(ngenes_to_add)
    # 
      
    # flatten metrics for selected phenotypes
    comparison_metrics<-panelapp_roc%>%bind_rows(naive_roc)
    
    panelapp_metrics<-panelapp_roc
    
    auto_matched_to_panelapp_by_precision<-selected_ag_roc%>%
      filter(!is.infinite(.threshold))%>%
      left_join(panelapp_metrics,by='phenotype_name')%>%
      group_by(phenotype_name)%>%
      # find the maximal recall value with precision equal or larger than panelapp
      filter(precision.x>=(precision.y-0.05))%>%
      #filter(abs(precision.x-precision.y)==min(abs(precision.x-precision.y)))%>%
      slice_max(n=1,order_by=recall.x)%>%slice_max(n=1,order_by=precision.x)
    
    # now do the same for the naive search, but this time match by recall (in order to show better precision)
    naive_metrics<-naive_roc
    auto_matched_to_naive_by_recall<-selected_ag_roc%>%left_join(naive_metrics,by='phenotype_name')%>%
      filter(.threshold>0)%>%
      group_by(phenotype_name)%>%
      #filter(abs(recall.x-recall.y)==min(abs(recall.x-recall.y)))%>%
      filter(recall.x>=(recall.y-0.05))%>%
      slice_max(n=1,order_by=precision.x)%>%slice_max(n=1,order_by=recall.x)
    
    
    comparison_metrics<-
      comparison_metrics%>%
      # now add the max f1 for each phenotype
      bind_rows(selected_ag_roc%>%group_by(phenotype_name)%>%
                  slice_max(n=1,order_by = f1)%>%
                  mutate(group='distance_max_f1')%>%
                  select(-c(.threshold,max_dist)))%>%
      # now add the max recall to the closest ppv
      bind_rows(auto_matched_to_panelapp_by_precision%>%
                  select(phenotype_name,group=group.x,f1=f1.x,precision=precision.x,recall=recall.x)%>%
                  mutate(group='distance_matched_to_panelapp_percision'))%>%
      # now add max precision to the closest recall
      bind_rows(auto_matched_to_naive_by_recall%>%
                  select(phenotype_name,group=group.x,f1=f1.x,precision=precision.x,recall=recall.x)%>%
                  mutate(group='distance_matched_to_naive_recall'))
    
      
      excluded_phenos<-c('melanoma')
      comparison_metrics<-comparison_metrics%>%filter(!phenotype_name%in%excluded_phenos)
      # plot max f1 vs panelapp vs naive
      comparison_metrics%>%filter(!grepl('distance_(naive|panel)',group))%>%ggplot(aes(x=group,y=f1,fill=group))+
        geom_col()+coord_flip()+
        facet_grid(phenotype_name~.)+
        guides(fill=F)+ggsci::scale_fill_nejm()+
        theme_minimal()
      
      # Plot precision recall curves and mark the pr for panelapp and the naive search
      selected_ag_roc%>%ggplot(aes(x=recall,y=precision,col=factor(group)))+
        geom_line()+facet_wrap(phenotype_name~.)+
        geom_abline(slope = -1,intercept = 1,linetype=2,alpha=0.5)+
        geom_point(data=panelapp_roc,inherit.aes = F,aes(x=recall,y=precision))+
        geom_point(data=naive_roc,inherit.aes = F,aes(x=recall,y=precision),col='red')+
        geom_point(data=comparison_metrics%>%filter(group=='distance_max_f1'),inherit.aes = F,aes(x=recall,y=precision),col='blue')+
        #geom_point(data=dist_roc,inherit.aes = F,aes(x=recall,y=precision),col='blue')+
        theme_minimal()
      
      f1_comp<-comparison_metrics%>%ungroup()%>%select(phenotype_name,group,f1)%>%pivot_wider(names_from = group,values_from = f1)
      recall_comp<-comparison_metrics%>%ungroup()%>%select(phenotype_name,group,recall)%>%pivot_wider(names_from = group,values_from = recall)
      precision_comp<-comparison_metrics%>%ungroup()%>%select(phenotype_name,group,precision)%>%pivot_wider(names_from = group,values_from = precision)
      
      
      
      # calculate between group comparisons
    
      # but this works and seems more legit - you can do this for these comparisons
      # distance f1 max vs panelapp f1
      # distance precision matched recall vs panelapp recall - to show that for a given PPV level, sensitivity is better
      # distance recall matched precision vs naive precision - to show that for a given sensitivity level, PPV is better
      library(report)
      precision_distance_vs_naive_recall_matched<-t.test(precision_comp%>%pull(distance_matched_to_naive_recall),precision_comp%>%pull(`naive search`),paired = T)
      report(precision_distance_vs_naive_recall_matched)
      precision_distance_vs_naive_recall_matched_df<-tidy(precision_distance_vs_naive_recall_matched)
      # make sure the recall is indeed matched
      distance_vs_naive_recall_diff_after_matching<-t.test(recall_comp%>%pull(distance_matched_to_naive_recall),recall_comp%>%pull(`naive search`),paired = T)
      
      precision_matched_recall_comp<-precision_comp%>%filter(panelapp-distance_matched_to_panelapp_percision<0.01)
      recall_distance_vs_panelapp_precision_matched<-t.test(recall_comp%>%pull(distance_matched_to_panelapp_percision),recall_comp%>%pull(panelapp),paired = T)
      report(recall_distance_vs_panelapp_precision_matched)
      # make sure the precision is indeed matched
      precision_comp%>%select(distance_matched_to_panelapp_percision,panelapp)%>%colMeans(na.rm = T)
      distance_vs_panelapp_precision_diff_after_matching<-t.test(precision_comp%>%pull(distance_matched_to_panelapp_percision),precision_comp%>%pull(panelapp),paired = T)
      
      
      f1_distance_vs_panelapp_max<-t.test(f1_comp%>%pull(distance_max_f1),f1_comp%>%pull(panelapp),paired = T)
      report(f1_distance_vs_panelapp_max)
      
      f1_distance_vs_naive_max<-t.test(f1_comp%>%pull(distance_max_f1),f1_comp%>%pull(`naive search`),paired = T)
      
      report(f1_distance_vs_naive_max)
      
      # in cases where the precision was matched, check how many times the distance method achieved higher recall than panelapp
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['better_recall_vs_panelapp']] <-
        recall_comp %>% filter(distance_matched_to_panelapp_percision > panelapp)
      # in cases where the recall was matched, check how many times the distance method achieved higher precision than naive
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['better_precision_vs_naive']] <-
        precision_comp %>% filter(distance_matched_to_naive_recall > `naive search`) 
      
      
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['f1_comp']]<-f1_comp
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['recall_comp']]<-recall_comp
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['precision_comp']]<-precision_comp
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['comparison_metrics']]<-comparison_metrics
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['precision_distance_vs_naive_recall_matched']]<-precision_distance_vs_naive_recall_matched
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['recall_distance_vs_panelapp_precision_matched']]<-recall_distance_vs_panelapp_precision_matched
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['f1_distance_vs_panelapp_max']]<-f1_distance_vs_panelapp_max
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['f1_distance_vs_naive_max']]<-f1_distance_vs_naive_max
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['distance_vs_panelapp_precision_diff_after_matching']]<-distance_vs_panelapp_precision_diff_after_matching
      statistical_tests_data[[reference_source]][[as.character(selected_dist)]][['distance_vs_naive_recall_diff_after_matching']]<-distance_vs_naive_recall_diff_after_matching
    }
}


```

# Save results

```{r save_results,include=T,eval=T}
elements_size<-data.frame(size=sort( sapply(ls(),function(x){object.size(get(x))})))
elements_to_save<-elements_size%>%filter(size<20000000)%>%row.names()
save(list=elements_to_save,file=glue('{analysis_dir}/workspace.RData'))
```

# Machine learning method

```{r ml_method}


library(listr)
auto_generator_panels_set<-auto_generator_panelapp_only[['max_dist=0.99']]
# add has_pub
ml_data<-NULL
panels_per_gene<-gen_source_db%>%group_by(gene_symbol)%>%summarize(number_of_panels_with_gene=n())%>%ungroup()
for (selected_phenotype_name in selected_phenotypes_df$phenotype_name){
  
  phenotype_ml_table<-pubmed_gene_list_vs_phenotype(phenotype_name = selected_phenotype_name,
                                                            gene_list = all_genes,
                                                            pre_generated_genes_pubs = entrez_pubmed_res)
  phenotype_ml_table<-phenotype_ml_table%>%
    mutate(has_pub=factor(ifelse(pub_num>=min_num_of_pub_to_consider_positive,1,0)))
  # add the auto_generator gene information
  phenotype_ml_table<-phenotype_ml_table%>%
    left_join(auto_generator_panels_set[[selected_phenotype_name]])%>%
    mutate(scaled_gene_score=ifelse(is.na(scaled_gene_score),0,scaled_gene_score),
           in_original_panels=ifelse(is.na(in_original_panels),0,in_original_panels),
           scaled_adjusted_gene_score=ifelse(is.na(scaled_adjusted_gene_score),0,scaled_adjusted_gene_score),
           rank_percentile=ifelse(is.na(rank_percentile),0,rank_percentile),
           num_o_panels=ifelse(is.na(num_o_panels),0,num_o_panels),
           min_panel_dist=ifelse(is.na(min_panel_dist),1,min_panel_dist),
           max_panel_dist=ifelse(is.na(max_panel_dist),1,max_panel_dist),
           median_panel_dist=ifelse(is.na(median_panel_dist),1,median_panel_dist),
           mean_panel_dist=ifelse(is.na(mean_panel_dist),1,mean_panel_dist),
           )
  
  # select relevant columns
  phenotype_ml_table<-phenotype_ml_table%>%select(-contains(c('pubmed_ids','pub_num')))
  ml_data<-ml_data%>%bind_rows(data.frame(phenotype_name=selected_phenotype_name,
                                          phenotype_ml_table))
}

# add total number of panels per gene
ml_data<-ml_data%>%select(-number_of_panels_with_gene)%>%left_join(panels_per_gene)

library(tidymodels)
ml_data<-ml_data%>%mutate(has_pub=factor(ifelse(has_pub==1,'Y','N')))

# some EDA
eda_ml_data<-ml_data%>%pivot_longer(-c(phenotype_name,gene_symbol,has_pub,num_o_panels,number_of_panels_with_gene))
eda_ml_data%>%
  filter(name%in%c('scaled_gene_score'))%>%
  ggplot(aes(x=phenotype_name,y=value,fill=has_pub))+
  geom_boxplot(outlier.size = 0.4)+facet_wrap(name~.,scales='free',nrow=2)+
  coord_flip()+
  theme_minimal()

eda_ml_data%>%
  filter(name%in%c('scaled_gene_score'))%>%
  ggplot(aes(x=number_of_panels_with_gene,y=value,col=has_pub))+
  geom_point(alpha=0.5)+facet_wrap(phenotype_name~.,scales='free',nrow=2)+
  theme_minimal()

eda_ml_data%>%
  filter(name%in%c('scaled_gene_score'))%>%
  ggplot(aes(x=phenotype_name,y=num_o_panels,fill=has_pub))+
  geom_boxplot()+facet_wrap(name~.,scales='free',nrow=2)+
  theme_minimal()

features<-setdiff(colnames(ml_data),c('phenotype_name','gene_symbol','has_pub',grep('^gene|rank$|^adjusted|size',colnames(ml_data),value=T)))
features<-c('scaled_adjusted_gene_score','min_panel_dist','max_panel_dist','number_of_panels_with_gene')

form<-as.formula(glue('has_pub~{paste0(features,collapse="+")}'))
# divide training and test data using the phenotypes
set.seed(1234)
train_phenotypes<-sample(selected_phenotypes,15)
test_phenotypes<-setdiff(selected_phenotypes,train_phenotypes)
train_data<-ml_data%>%filter(phenotype_name%in%train_phenotypes,gene_score>0)

# train down sampled
train_downsample<-downsample_training_data(train_data,'has_pub',under_ratio = 5)
test_data<-ml_data%>%filter(phenotype_name%in%test_phenotypes)

model_def<-xgb_def()
model_def<-glmnet_def()

model_res<-train_model(train_data = train_downsample,
                       model_def = model_def,
                       formula = form,
                       group = 'phenotype_name',
                       grid_size = 20)

model_res$train_metrics
library(vip)
model_vip<-model_res$model_fit%>%extract_fit_parsnip() %>%
  vip(geom = "point")
model_vip

rf_testing_pred <- 
  predict(model_res$model_fit, test_data) %>% 
  bind_cols(predict(model_res$model_fit, test_data, type = "prob")) %>% 
  bind_cols(test_data %>% select(phenotype_name,has_pub,scaled_gene_score))

rf_testing_pred%>%
  ggplot(aes(x=.pred_Y,fill=has_pub))+
  geom_density(alpha=0.4)+
  theme_minimal()

rf_testing_pred%>%pr_auc(truth=has_pub,estimate=.pred_Y,event_level='second')
test_data%>%pr_auc(truth=has_pub,estimate=scaled_gene_score,event_level='second')


rf_testing_pred %>%
  select(phenotype_name,has_pub,scaled_gene_score,.pred_Y)%>%
  pivot_longer(-c(phenotype_name,has_pub))%>%
  group_by(phenotype_name,name)%>%
  ggplot(aes(x=value,fill=has_pub))+
  facet_wrap(name~.,scales='free')+
  geom_density(alpha=0.4)+
  theme_minimal()

rf_testing_pred %>%
  select(phenotype_name,has_pub,scaled_gene_score,.pred_Y)%>%
  pivot_longer(-c(phenotype_name,has_pub))%>%
  group_by(phenotype_name,name)%>%
  pr_curve(has_pub, value,event_level='second') %>%
  ggplot(aes(x = recall, y = precision,col=name)) +
  geom_line() +
  facet_wrap(phenotype_name~.)+
  geom_abline(
    lty = 2, alpha = -0.5,
    color = "gray50",
    size = 1.2
  )
```

# Compare vs clingen

```{r clingen_comp}

clingen_panels<-gtr_with_expert_db%>%filter(source=='clingen')%>%pull(panel_joined_name)%>%unique()

clingen_vs_selected<-data.frame(
  phenotype_name=c('dilated cardiomyopathy',
                       'hypertrophic cardiomyopathy',
                       'hearing loss'),
  clingen_panel=c("clingen|5|Dilated Cardiomyopathy",
                  "clingen|11|Hypertrophic Cardiomyopathy",
                  'clingen|7|Hearing Loss')
)

for (selected_phenotype_name in clingen_vs_selected$phenotype_name){
  clingen_panel<-clingen_vs_selected%>%filter(phenotype_name==selected_phenotype_name)%>%pull(clingen_panel)
  clingen_genes<-gtr_with_expert_db%>%filter(panel_joined_name==clingen_panel,confidence_level>1)%>%pull(gene_symbol)
  all_genes_vs_clingen<-data.frame(gene_symbol=all_genes)%>%mutate(has_pub=factor(ifelse(gene_symbol%in%clingen_genes,1,0)))
  selected_ag<-auto_generator_panelapp_only[[glue('max_dist={selected_dist}')]][[selected_phenotype_name]]
  ag_roc <- calculate_auto_generator_pr_for_phenotype(phenotype_name = selected_phenotype_name,
                                                      auto_generator_panel_res = selected_ag,
                                                      all_genes_vs_phenotype = all_genes_vs_clingen)
  g<-ag_roc%>%ggplot(aes(x=recall,y=precision,col=factor(group)))+
  geom_line()+facet_wrap(phenotype_name~.)+
  geom_abline(slope = -1,intercept = 1,linetype=2,alpha=0.5)+
  #geom_point(data=dist_roc,inherit.aes = F,aes(x=recall,y=precision),col='blue')+
  theme_minimal()
  print(g)
}

# Literature text
comparison_metrics_summary<-comparison_metrics%>%
  group_by(group)%>%select(precision,recall,f1)%>%skimr::skim_without_charts()

comparison_metrics_text<-comparison_metrics_summary%>%group_by(group,skim_variable)%>%
  summarize(metric_text=glue('a mean {skim_variable} of {round(numeric.mean,2)} (IQR[{round(numeric.p25,2)},{round(numeric.p75,2)}])'))

```






```{r pca}
# Add PCA calculation
test_binary<-gtr_binary_panel_gene_table[1:1000,1:1000]
library(logisticPCA)
logsvd_model = logisticSVD(test_binary, k = 2)
plot(logsvd_model)

panels = rownames(test_binary)
plot(logsvd_model, type = "scores") + geom_point(aes()) + 
  ggtitle("Exponential Family PCA") + scale_colour_manual(values = c("blue", "red"))

library(FactoMineR)
test_binary<-gtr_binary_panel_gene_table[,]
res.mca <- FactoMineR::PCA(test_binary, 
              ncp=2,graph=F)
coords<-data.frame(panel=rownames(res.mca$ind$coord),res.mca$ind$coord)%>%
  mutate(source=stringr::str_replace(panel,'\\|.+',''))

to_plot<-coords%>%filter(panel%in%naive_panels)%>%mutate(group='naive_panels')%>%
  bind_rows(coords%>%filter(panel%in%dist_panels)%>%mutate(group='dist_panels'))

to_plot%>%
  #filter(Dim.1<50,Dim.2<20)%>%
  ggplot(aes(x=Dim.1,y=Dim.2,col=factor(source=='panelapp')))+
  geom_point(alpha=0.4)#+facet_wrap(group~.)+
  theme_minimal()

```


```{r check_dist_distribution_for_panelapp_panels_1}
panelapp_panels_dist<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%group_by(row)%>%
  filter(col!=row)%>%
  summarize(mean_dist=mean(value),
            median_dist=median(value),
            min_dist=min(value))%>%
  left_join(gtr_with_expert_db%>%
              filter(source=='panelapp')%>%
              mutate(panel_joined_name=as.character(panel_joined_name))%>%
              group_by(panel_joined_name)%>%
              summarize(n_genes=n()),
            by=c('row'='panel_joined_name'))

panelapp_panels_dist%>%filter(n_genes<1000)%>%ggplot(aes(x=n_genes,y=min_dist))+
  geom_point()+
  theme_minimal()

```



```{r cluster_panels}
# hierarchal clustering
hc <- hclust(gtr_dm,method='ward.D' )
# cut tree
cut_avg <- cutree(hc,k = 100)
panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
# this is not bad - it actually groups similar panels together.. but what is the optimal number of clusters??
# basically the purpose is to reduce false positives while maintaining high sensitivity - so lets see what happens when i divide the panels into X clusters and go from there
panel_clusters%>%dplyr::count(cluster)%>%arrange(n)

```



```{r analyze_distance_matrix}
# Analyze the ratio between the number of genes in a panel and the distance it has with all other panels with the same approximate size

dm_to_analyze<-gtr_dm_df%>%
  left_join(gtr_panels_list%>%select(panel_joined_name,number_of_genes_row=number_of_genes),by=c('row'='panel_joined_name'))%>%
  left_join(gtr_panels_list%>%select(panel_joined_name,number_of_genes_col=number_of_genes),by=c('col'='panel_joined_name'))

dm_to_analyze<-dm_to_analyze%>%mutate(ngenes_cat_row=cut(number_of_genes_row,breaks=c(0,10,30,50,100,200,500,1000,10000)),
                                      ngenes_cat_col=cut(number_of_genes_col,breaks=c(0,10,30,50,100,200,500,1000,10000)))


head(dm_to_analyze)
dm_to_analyze%>%slice_sample(n=100000)%>%ggplot(aes(x=number_of_genes_row,y=value))+
  geom_point()+facet_wrap(ngenes_cat_row~.,scales='free')
  

```

```{r compare_panelapp_against_the_rest}
# This code goes through all of panelapps panels and compares them against all the panels with at most max_dist distance. 
# it then calculates sensitivity and precision for all the panels and outputs how many panels had missing genes in them

panelapp_df<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%filter(!grepl('panelapp',col))
panelapp_df<-panelapp_df%>%
  left_join(panels_list%>%
              mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))

# Check whether how well the different panelapp panels are represented by looking for the sensitivity 
max_dist<-c(1,0.99,0.95,0.9,0.85)
panelapp_vs_all_other<-NULL
for (md in max_dist){
  message(glue('calculating panelapp similarity for {md} distance'))
  panelapp_comp<-panelapp_df%>%group_by(row)%>%filter(value<md)%>%ungroup()
  panelapp_comp<-panelapp_comp%>%group_by(row)%>%dplyr::summarize(num_o_comp_panels=n(),compare_panel_vs_many_panels(row,unique(col),gtr_with_expert_db,min_confidence_level = 2))%>%mutate(max_dist=md)
  
  panels_without_similar_panels<-setdiff(panelapp_df%>%pull(row)%>%as.character()%>%unique(),panelapp_comp%>%pull(row))
  if (length(panels_without_similar_panels)>0){
    panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                 num_o_comp_panels=0,
                                                 genes_in_b=0,
                                                 common_genes=0,
                                                 B_not_in_A=0,
                                                 Agenes_not_in_B='',
                                                 max_dist=md)
    panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
      genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
      A_not_in_B=genes_in_a
    )
    panelapp_comp<-panelapp_comp%>%bind_rows(panels_without_similar_panels_df)
  }
  
  panelapp_vs_all_other<-panelapp_vs_all_other%>%bind_rows(panelapp_comp)
}

panelapp_vs_all_other<-panelapp_vs_all_other%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)

panelapp_vs_all_other%>%group_by(max_dist)%>%dplyr::summarize(mean_sens=mean(sensitivity),
                                              mean_prec=mean(precision),
                                              num_o_panels_with_missing_genes=sum(A_not_in_B>0))



```


```{r compare_panelapp_against_the_rest_cluster_analysis}
# This code goes through all of panelapps panels and compares them against all the panels with at most max_dist distance. 
# it then calculates sensitivity and precision for all the panels and outputs how many panels had missing genes in them



panelapp_vs_all_other_cluster<-NULL
for (method in c('ward.D2','ward.D')){
  print(method)
  for (cluster_size in c(1,75,150)){
    print(cluster_size)
    # hierarchal clustering
    hc <- hclust(gtr_dm,method=method )
    
    # library(cluster)
    # z<-agnes(gtr_dm)
    # cut tree
    cut_avg <- cutree(hc,k = cluster_size)
    panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
    # this is not bad - it actually groups similar panels together.. but what is the optimal number of clusters??
    # basically the purpose is to reduce false positives while maintaining high sensitivity - so i lets see what happens when i divide the panels into X clusters and go from there
    panel_clusters%>%dplyr::count(cluster)%>%arrange(n)
    
    
    panelapp_df<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%filter(!grepl('panelapp',col))
    panelapp_df<-panelapp_df%>%
      left_join(panels_list%>%
                  mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
      left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
      left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)
    
    # now do the same but only within the panel cluster
    # Check whether how well the different panelapp panels are represented by looking for the sensitivity 
    max_dist<-c(1,0.99,0.95,0.9)
    
    for (md in max_dist){
      message(glue('calculating panelapp similarity for {md} distance'))
      panelapp_comp<-panelapp_df%>%group_by(row)%>%filter(value<md)%>%filter(row_cluster==col_cluster)%>%ungroup()
      panelapp_comp<-panelapp_comp%>%group_by(row)%>%
        dplyr::summarize(num_o_comp_panels=n(),
                         compare_panel_vs_many_panels(row,unique(col),
                                                      gtr_with_expert_db,
                                                      min_confidence_level = 2))%>%
        mutate(max_dist=md)
      
      panels_without_similar_panels<-setdiff(panelapp_df%>%pull(row)%>%as.character()%>%unique(),panelapp_comp%>%pull(row))
      if (length(panels_without_similar_panels)>0){
        panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                     num_o_comp_panels=0,
                                                     genes_in_b=0,
                                                     common_genes=0,
                                                     B_not_in_A=0,
                                                     Agenes_not_in_B='',
                                                     max_dist=md)
        panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
          genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
          A_not_in_B=genes_in_a
        )
        panelapp_comp<-panelapp_comp%>%bind_rows(panels_without_similar_panels_df)
      }
      
      panelapp_vs_all_other_cluster<-panelapp_vs_all_other_cluster%>%
        bind_rows(data.frame(method=method,cluster_size=cluster_size,panelapp_comp))
    }
    
    
  }
}

panelapp_vs_all_other_cluster<-panelapp_vs_all_other_cluster%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)%>%rowwise()%>%mutate(f1= (2 * precision * sensitivity) / (precision + sensitivity))

cluster_analysis_res<-panelapp_vs_all_other_cluster %>%
  group_by(method, cluster_size, max_dist) %>%
  dplyr::summarize(mean_sens = mean(sensitivity),
                   mean_prec = mean(precision),
                   mean_f1 = mean(f1,na.rm=T),
                   num_o_panels_with_missing_genes=sum(A_not_in_B>0))

cluster_analysis_res%>%filter(method=='ward.D2')%>%select(-num_o_panels_with_missing_genes)%>%
  pivot_longer(-c(cluster_size,max_dist,method))%>%
  ggplot(aes(x=factor(cluster_size),y=value,fill=method))+
  geom_col(position='dodge')+facet_wrap(name~max_dist,scale='free')+
  theme_minimal()

cluster_analysis_re(2 * precision * sensitivity) / (precision + sensitivity)s%>%filter(method=='ward.D')%>%select(-num_o_panels_with_missing_genes)%>%
  pivot_longer(-c(cluster_size,max_dist,method))%>%
  ggplot(aes(x=cluster_size,y=value,col=factor(max_dist)))+
  geom_line()+facet_wrap(name~.,scale='free')+
  theme_minimal()+
  ggsci::scale_color_nejm()+
  theme(legend.position = 'top')


```


```{r compare_clingen_vs_panelapp_plus_gtr}


#max_dist<-c(1,0.99,0.95,0.9,0.85)
max_dist<-c(1,0.99,0.98,0.975,0.97,0.95)
cluster_sizes<-c(1,50,80,100)
cluster_method<-'ward.D'
clingen_vs_all_other<-NULL

for(cluster_size in cluster_sizes){
  
  hc <- hclust(gtr_dm,method=cluster_method )
  # library(cluster)
  # z<-agnes(gtr_dm)
  # cut tree
  cut_avg <- cutree(hc,k = cluster_size)
  panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))
  
  # for every clingen panel - build a panel and see how it covers it
  clingen_df<-gtr_dm_df%>%filter(grepl('clingen',row))%>%filter(!grepl('clingen',col))

  
  clingen_df<-clingen_df%>%
        left_join(pan(2 * precision * sensitivity) / (precision + sensitivity)els_list%>%
                    mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
        left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
        left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)
  
  for (md in max_dist){
    message(glue('calculating clingen similarity for {md} distance'))
    clingen_comp<-clingen_df%>%group_by(row)%>%filter(value<md)%>%ungroup()
    clingen_comp<-clingen_df%>%group_by(row)%>%filter(value<md)%>%filter(row_cluster==col_cluster)%>%ungroup()
    clingen_comp<-clingen_comp%>%group_by(row)%>%
      summarize(num_o_comp_panels=n(),
                compare_panel_vs_many_panels(row,unique(col),gtr_with_expert_db,min_confidence_level = 3))%>%
      mutate(max_dist=md,cluster_size=cluster_size)
    panels_without_similar_panels<-setdiff(clingen_df%>%pull(row)%>%as.character()%>%unique(),clingen_comp%>%pull(row))
    if (length(panels_without_similar_panels)>0){
      panels_without_similar_panels_df<-data.frame(row=panels_without_similar_panels,
                                                   num_o_comp_panels=0,
                                                   genes_in_b=0,
                                                   common_genes=0,
                                                   B_not_in_A=0,
                                                   Agenes_not_in_B='',
                                                   max_dist=md)
      panels_without_similar_panels_df<-panels_without_similar_panels_df%>%rowwise%>%mutate(
        genes_in_a=length(gtr_with_expert_db%>%filter(panel_joined_name==row)%>%pull(gene_symbol)),
        A_not_in_B=genes_in_a
      )
      clingen_comp<-clingen_comp%>%bind_rows(panels_without_similar_panels_df)
    }
    clingen_vs_all_other<-clingen_vs_all_other%>%bind_rows(clingen_comp)
  }
}

clingen_vs_all_other<-clingen_vs_all_other%>%mutate(
  TP=common_genes,
  FP=B_not_in_A,
  FN=A_not_in_B,
  sensitivity=TP/(TP+FN),
  precision=ifelse(TP>0,TP/(TP+FP),0)
)


clingen_test_res<-clingen_vs_all_other%>%group_by(cluster_size,max_dist)%>%summarize(mean_sens=mean(sensitivity),
                                              mean_prec=mean(precision),
                                              num_o_panels_with_missing_genes=sum(A_not_in_B>0))
z<-clingen_vs_all_other%>%filter(max_dist==0.9)
```

```{r compare_auto_vs_vendor}

hc <- hclust(gtr_dm,method='ward.D' )
# library(cluster)
# z<-agnes(gtr_dm)
# cut tree
cut_avg <- cutree(hc,k = 100)
panel_clusters<-data.frame(panel_name=names(cut_avg),cluster=as.numeric(cut_avg))

auto_df<-gtr_dm_df%>%
      left_join(panels_list%>%
                  mutate(panel_joined_name=as.character(panel_joined_name)),by=c('row'='panel_joined_name'))%>%
      left_join(panel_clusters,by=c('row'='panel_name'))%>%dplyr::rename('row_cluster'=cluster)%>%
      left_join(panel_clusters,by=c('col'='panel_name'))%>%dplyr::rename('col_cluster'=cluster)

# start with a panelapp panel
all_panel_names<-auto_df%>%pull(row)%>%unique()
auto_df%>%select(row,row_cluster)%>%distinct()%>%filter(grepl('panelapp',row),grepl('hypertroph',row,ignore.case = T))%>%pull(row,row_cluster)
panelapp_panel<-"panelapp|49|Hypertrophic cardiomyopathy - teen and adult" 
cluster=35
same_cluster_panels<-auto_df%>%select(col,col_cluster)%>%distinct()%>%filter(col_cluster==35)%>%pull(col)

# build hypertrophy panel according to panelapp panel
auto_panel<-generate_panel_from_panels(joined_panel_names = panelapp_panel,
                                       gtr_with_expert_db%>%
                                         filter(panel_joined_name%in%c(panelapp_panel,same_cluster_panels)),
                                       auto_df,
                                       max_distance = 1,
                                       with_mitochondrial = T)

# blueprint
blueprint_panel<-source_db%>%filter(source=='blueprint',grepl('hypertrophic',panel_joined_name,ignore.case = T))
# invitae
invitae_panel<-source_db%>%filter(source=='invitae',grepl('hypertrophic',panel_joined_name,ignore.case = T))
# cegat
cegat_panel<-source_db%>%filter(source=='cegat',grepl('hypertrophic',panel_joined_name,ignore.case = T))

library(UpSetR)
panels_mat<-auto_panel%>%filter(adjusted_rank_percentile>0.5)%>%
  select(gene_symbol)%>%mutate(auto_panel=1)%>%
  pivot_wider(names_from = gene_symbol,values_from = auto_panel)%>%
  bind_rows(blueprint_panel%>%select(gene_symbol)%>%mutate(blueprint=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = blueprint,values_fill = NA))%>%
  bind_rows(invitae_panel%>%select(gene_symbol)%>%mutate(invitae=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = invitae,values_fill = NA))%>%
  bind_rows(cegat_panel%>%select(gene_symbol)%>%mutate(cegat=1)%>%
              pivot_wider(names_from = gene_symbol,values_from = cegat,values_fill = NA))%>%
  t()%>%as.data.frame()
  
colnames(panels_mat)<-c('auto_panel','blueprint','invitae','cegat')
panels_mat[is.na(panels_mat)]<-0
upset(panels_mat)

# maybe at this point i will perform an automated pubmed search? and see how many genes have the association ?
```


```{r}

# now for hypertrophic cardiomyopathy - get the most associated panels
selected_panel_name<-find_panel_in_dm_df(dm_df,source='panelapp',panel_name='Hypertrophic cardio')
# get similar panels
similar_panels_df<-get_similar_panels_from_dm_df_by_prop(gtr_dm_df,selected_panel_name,top_most_similar = 0.005,maximal_distance = 0.99)

similar_panels_binary_gene_table<-subset_binary_gene_table(gtr_binary_panel_gene_table,similar_panels_df$col)
plot_panel_gene_heatmap(similar_panels_binary_gene_table,sort_by_similarity_to_selected_panel = T)


gbin_for_heatmap<-bin_for_heatmap%>%mutate(panel=row.names(bin_for_heatmap))%>%pivot_longer(-panel)%>%left_join(similar_panels_df,by=c('panel'='col'))
p <- ggplot(gbin_for_heatmap, aes(forcats::fct_reorder(panel, value.y), name, fill= value.x)) + 
  geom_tile()+coord_flip()+theme(axis.text.x = element_text(angle=90))
p

panels_to_plot<-as.character(unique(panelapp_hcm$col))
panels_num<-match(panels_to_plot,row.names(test_bin_data))
selected_bin_data<-test_bin_data[panels_num,]
selected_dm<-generate_distance_matrix(selected_bin_data)
plot_dm_as_dendogram(selected_dm,selected_panel)
# generate heatmap !!!

panels<-row.names(binary_panel_gene_table)[grepl('hypertroph',row.names(binary_panel_gene_table),ignore.case = T) & grepl('cardio',row.names(binary_panel_gene_table),ignore.case=T) & !(grepl('hpo',row.names(binary_panel_gene_table),ignore.case = T))]

```

```{r auto_generate_panel}

#joined_panel_names<-c('invitae|02261|Invitae Hypertrophic Cardiomyopathy Panel','health_in_code|1_1|Hypertrophic Cardiomyopathy – Basic Panel',"blueprint|1_16|Hypertrophic Cardiomyopathy (HCM) Panel",'cegat|3_2|Cardiomyopathy, hypertrophic','health_in_code|1_2|Hypertrophic Cardiomyopathy – Extended Panel')
# gen_source_db<-fixed_source_db%>%filter(!grepl('clingen|panelapp|hpo',source))
# gen_dm_df<-dm_df%>%filter(!grepl('clingen|panelapp',row))%>%filter(!grepl('clingen|panelapp|hpo',col))


joined_panel_names<-c('expert_panel|49|Hypertrophic cardiomyopathy - teen and adult')
joined_panel_names<-c('panelapp|49|Hypertrophic cardiomyopathy - teen and adult')
joined_panel_names<-c('panelapp|652|Dilated cardiomyopathy - adult and teen')
joined_panel_names<-c('Invitae|514924|Invitae Hypertrophic Cardiomyopathy Panel','Blueprint Genetics|552625|Hypertrophic Cardiomyopathy (HCM) Panel','Health in Code S.L.|530670|Hypertrophic cardiomyopathy extended panel')
joined_panel_names<-c('panelapp|555|Ichthyosis and erythrokeratoderma')

comp_panel<-expert_panels_source%>%filter(grepl('Hypertrophic',panel_joined_name))
comp_panel<-fixed_source_db%>%filter(grepl('panelapp',source))%>%filter(grepl('Hypertrophic',panel_joined_name))
comp_panel<-readr::read_csv('./data/accessory_files/hypertrophic_cardiomyopathy_basic.csv',col_names = 'gene_symbol')
comp_panel<-readr::read_csv('./data/accessory_files/hypertrophic_cardiomyopathy_extended.csv',col_names = 'gene_symbol')%>%mutate(gene_symbol=stringr::str_replace(gene_symbol,'\\*',''))
comp_panel<-readr::read_delim('./data/accessory_files/ichthyosis_erythrokeratodermi.csv')
comp_panel$gene_symbol<-comp_panel$gene_symbol%>%trimws()

gen_source_db<-gtr_with_expert_db
gen_dm_df<-gtr_dm_df%>%filter(!grepl('panelapp|clingen|expert',col))

gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db,gen_dm_df,max_distance = 0.9,with_mitochondrial = T)

# run panelapp analysis
panelapp_panels<-panels_list%>%filter(grepl('panelapp',source))%>%ungroup()%>%select(panel_joined_name)%>%mutate(panel_joined_name=panel_joined_name%>%as.character())

panelapp_generated_panels<-NULL
for (panelapp_panel in panelapp_panels$panel_joined_name){
  message(glue('Generating panel for {panelapp_panel}'))
  gene_rank<-generate_panel_from_panels(joined_panel_names = panelapp_panel,gen_source_db,gen_dm_df,max_distance = 0.9,with_mitochondrial = T)
  if (nrow(gene_rank)==0){next}
  # get the confidence level
  gene_rank<-gene_rank%>%left_join(fixed_source_db%>%filter(panel_joined_name==panelapp_panel)%>%select(gene_symbol,confidence_level))
  panelapp_generated_panels<-panelapp_generated_panels%>%bind_rows(data.frame(panelapp_panel,gene_rank))
}
g<-ggplot(panelapp_generated_panels,aes(y=scaled_adjusted_gene_score,x=cut(adjusted_rank_percentile,10),fill=factor(confidence_level)))+
  geom_boxplot()#+facet_wrap(panelapp_panel~.)
g


toPlot<-panelapp_generated_panels%>%mutate(adjusted_rank_percentile_cat=cut(adjusted_rank_percentile,10),
                                           confidence_level=ifelse(is.na(confidence_level),0,ifelse(confidence_level==0,1,confidence_level)))%>%
  group_by(adjusted_rank_percentile_cat,panelapp_panel)%>%
  summarize(num_o_genes=n(),
            num_o_high_conf=sum(ifelse(confidence_level==3,1,0)),
            num_o_med_conf=sum(ifelse(confidence_level==2,1,0)))%>%
  mutate(rate_o_high_conf=num_o_high_conf/num_o_genes,
         rate_o_med_conf=num_o_med_conf/num_o_genes)


g<-ggplot(toPlot,aes(x=adjusted_rank_percentile_cat,y=rate_o_med_conf))+
  geom_boxplot()
g



calculate_distance_by_thresh<-function(gene_rank,comp_panel,adjusted=F){
  dist_by_thresh<-NULL
  for (thresh in seq(0,1,0.001)){
    if (adjusted){
      genes_above_thresh<-gene_rank%>%filter(adjusted_rank_percentile>=thresh)%>%pull(gene_symbol)
    }else{
      genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
    }
    dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
    dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
  }
  dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)
  return(dist_by_thresh)
}

dist_by_thresh<-calculate_distance_by_thresh(gene_rank,comp_panel,adjusted=T)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g

g<-ggplot(dist_by_thresh,aes(x=thresh,y=sensitivity))+
  geom_point()+
  theme_minimal()
g

comp_panel<-comp_panel%>%left_join(gene_rank)


joined_panel_names<-c("Fulgent Genetics|566600|Comprehensive Glomerular Proteinuria NGS Panel")
comp_panel<-fixed_source_db%>%filter(panel_joined_name=='panelapp|106|Proteinuric renal disease')%>%filter(confidence_level>1)

gene_rank<-generate_panel_from_panels(joined_panel_names = joined_panel_names,gen_source_db = gtr_db,gen_dm_df = gtr_dm_df,max_distance = 0.99,with_mitochondrial = T)

comp_panel<-comp_panel%>%left_join(gene_rank)

dist_by_thresh<-NULL
for (thresh in seq(0,1,0.001)){
  genes_above_thresh<-gene_rank%>%filter(rank_percentile>=thresh)%>%pull(gene_symbol)
  dist<-calculate_panel_similarity_as_df(genes_above_thresh,comp_panel%>%pull(gene_symbol))
  dist_by_thresh<-dist_by_thresh%>%bind_rows(data.frame(thresh,num_o_genes=length(genes_above_thresh),dist))
}
dist_by_thresh<-dist_by_thresh%>%mutate(sensitivity=common_genes/comparison_panel_genes,
                                          ppv=common_genes/selected_panel_genes)


g<-ggplot(dist_by_thresh%>%group_by(sensitivity)%>%slice_max(order_by = ppv,n=1),aes(x=sensitivity,y=ppv))+
  geom_line()+
  ylim(c(0,1))+xlim(c(0,1))+
  theme_minimal()
g


g<-ggplot(dist_by_thresh,aes(x=thresh,y=sensitivity))+
  geom_point()+
  theme_minimal()
g
```


# Expert panel - panels combination

we will generate panel combinations for the expert panels (panelapp NHS and clingen) including only high condidence genes

```{r prepare_expert_panels}
expert_panels_panelapp<-panels_list%>%
  filter(source=='panelapp' & (high_conf_genes>1))

# get the panelapp NHS approved panels
panelapp_nhs_panels<-fixed_source_db%>%filter(panel_comment=='NHS panel')%>%pull(panel_name)%>%unique()
expert_panels_panelapp<-expert_panels_panelapp%>%filter(panel_name %in% panelapp_nhs_panels)

# now get the clingen panels
expert_panels_clingen<-panels_list%>%
  filter(source=='clingen' & (high_conf_genes>1))

clingen_panelapp_panels<-c(expert_panels_clingen%>%pull(panel_joined_name),
                           expert_panels_panelapp%>%pull(panel_joined_name))

```

```{r find_similar_clingen_panelapp_panels}

# now compare the two panels
expert_panels_dm_df<-dm_df%>%filter(grepl('clingen',row))%>%filter(col %in% clingen_panelapp_panels)
clingen_panels<-as.character(expert_panels_clingen%>%pull(panel_joined_name))
clingen_vs_panelapp<-get_top_similar_panels_table(joined_panel_names = clingen_panels,
                                                  source_db = fixed_source_db,
                                                  dm_df = expert_panels_dm_df,
                                                  number_of_top_panels = 5,
                                                  same_source = F)
write.table(file=glue('{analysis_dir}/clingen_vs_panelapp.csv'),clingen_vs_panelapp,sep='\t',row.names = F)
```

once the corresponding clingen / panelapp panels were found join them and collect only the high confidence genes

```{r generate_expert_panels}
clingen_vs_panelapp<-readr::read_delim('./data/accessory_files/clingen_vs_panelapp_manual_match.csv')
# use the NHS approved panelapp panels as reference and add the clingen high confidence genes 
expert_panels_list<-expert_panels_panelapp%>%mutate(source='expert_panels')
expert_panels_source<-fixed_source_db%>%
  mutate(source='expert_panel')%>%
  filter(panel_joined_name%in%expert_panels_list$panel_joined_name)%>% # take only the NHS panels
  filter(confidence_level==3) # take only the high confidence genes

# now add the matching clingen panel genes
for (clingen_panel_name in clingen_vs_panelapp$selected_panel_name){
  clingen_panel_source<-fixed_source_db%>%filter(panel_joined_name==clingen_panel_name)
  matching_panelapp_panel_joined_name<-clingen_vs_panelapp%>%filter(selected_panel_name==clingen_panel_name)%>%pull(similar_panel)
  matching_panelapp_panel<-expert_panels_source%>%filter(panel_joined_name==matching_panelapp_panel_joined_name)
  # now get the clingen genes that are not in the panelapp  panel
  clingen_genes_to_add<-clingen_panel_source%>%
    filter(confidence_level==3)%>%
    filter(!(gene_symbol %in% (matching_panelapp_panel%>%pull(gene_symbol))))
  message(glue('Will add {nrow(clingen_genes_to_add)} genes from {clingen_panel_name} to {matching_panelapp_panel_joined_name}'))
  expert_panels_source<-expert_panels_source%>%bind_rows(clingen_genes_to_add%>%mutate(source='expert_panels',
                                                                panel_joined_name=matching_panelapp_panel_joined_name,
                                                                panel_id=unique(matching_panelapp_panel%>%pull(panel_id)),
                                                                panel_name=unique(matching_panelapp_panel%>%pull(panel_name))))
}

# now add all the clingen panels without a similar panelapp panel
remaining_clingen_panels<-fixed_source_db%>%filter(source=='clingen')%>%
  filter(!(panel_joined_name%in%clingen_vs_panelapp$selected_panel_name))%>%
  filter(!(panel_name %in% c('General Gene Curation','UNC Biocuration Core','Syndromic Disorders','PTEN')))%>%
  filter(!(panel_name %in% (panels_list%>%filter(source=='clingen' & high_conf_genes==1)%>%pull(panel_name))))

expert_panels_source<-expert_panels_source%>%bind_rows(remaining_clingen_panels)
  
# now fix the source table 
joined_panel_names<-split_joined_panel_name(expert_panels_source$panel_joined_name)
joined_panel_names<-joined_panel_names%>%mutate(source='expert_panel',
                                                panel_joined_name=glue('{source}|{panel_id}|{panel_name}'))

expert_panels_source$source<-joined_panel_names$source
expert_panels_source$panel_joined_name<-joined_panel_names$panel_joined_name

```

```{r check_dist_distribution_for_panelapp_panels}


panelapp_panels_dist<-gtr_dm_df%>%filter(grepl('panelapp',row))%>%group_by(row)%>%
  filter(col!=row)%>%
  summarize(mean_dist=mean(value),
            median_dist=median(value),
            min_dist=min(value))%>%
  left_join(gtr_with_expert_db%>%
              filter(source=='panelapp')%>%
              mutate(panel_joined_name=as.character(panel_joined_name))%>%
              group_by(panel_joined_name)%>%
              summarize(n_genes=n()),
            by=c('row'='panel_joined_name'))

panelapp_panels_dist%>%filter(n_genes<1000)%>%ggplot(aes(x=n_genes,y=min_dist))+
  geom_point()+
  theme_minimal()



```


# Disgenet

it appears that disgenet db has stopped updating since 2020 and now has a paid version called disgenet plus
i will therefore try and download other sources of gene-disease associations. 
specifically - this looks promising:
https://diseases.jensenlab.org/Downloads

```{r disgenet}

#save(disgenet_df,file = './data/pre_generated_data/disgenet_vs_all_genes.RData')
load('./data/pre_generated_data/disgenet_vs_all_genes.RData')

api_key<-get_dg_api_key()

disgenet_vs_phenotype<-list()
# now for each phenotype get the disgenet genes
for (selected_phenotype in selected_phenotypes){
  message(glue('Retrieving {selected_phenotype} genes from disgenet'))
  phenotype_query<-selected_phenotypes_df%>%filter(phenotype_name==selected_phenotype)%>%pull(query)
  disgenet_diseases<-grep(phenotype_query,disgenet_df$disease_name,ignore.case = T,value=T)
  # dsi is a score for gene-disease specificity
  disgenet_genes<-disgenet_df%>%filter(disease_name %in% disgenet_diseases)%>%
    select(gene_symbol,score,ei,el)%>%
    group_by(gene_symbol)%>%
    slice_max(n=1,order_by=score)%>%
    ungroup()
  disgenet_vs_phenotype[[selected_phenotype]]<-disgenet_genes
}


```